---
title: "JZ_GED_PHD_C1 Analysis Setup 8"
author: "Johan Zvrskovec"
date: "03/05/2022"
output:
  html_document:
    fig_height: 6
    fig_width: 9
  pdf_document: default
  word_document:
    fig_height: 6
    fig_width: 9
---

```{r clean, include=FALSE, purl=FALSE}
rm(list = ls(all.names = TRUE)) #will clear all objects includes hidden objects.
#gc() #free up memory and report the memory usage.
```

# Title
**The joint phenotypic and molecular genetic architecture of Major Depressive Disorder (MDD) and anxiety disorders**

This analysis script is meant to contain most of the analysis steps performed in the project. It is run either as Rmd chunks in order, or as an R-script produced by the knitr::purl command further down. Run from the command line it takes command line arguments as seen in the command line setup below. It can be instructed (through command line arguments) to run in two run modes configured for running either locally on my laptop or on the HPC cluster. Some parts of the script can be run separately as a specific task (controlled by the command line arguments). So far is the Multivariate LD block recommended to be run on a HPC cluster and this block is also subdivided into different tasks intended to be run separately. The results of these separate tasks are to be loaded by a general run of the script to be included in the later analysis. When this has been done once, the final LD results will be saved to a file for subsequent runs of the script to rely on without having to run the LD step again. Remove (or rename) the LD results for the procedure to start anew.

# Packages

```{r package setup, echo=FALSE, warning=F}
#install.packages("disk.frame")  
#install.packages("skimr")
#install.packages("psych")
#install.packages("Matrix")
#install.packages("tidyverse")
#install.packages("ggrepel")
#install.packages("gt")
#install.packages("kableExtra")
#remove.packages("corrplot")
#devtools::install_github("taiyun/corrplot", build_vignettes = TRUE) #this is the most updated
#devtools::install_github("mkanai/corrplot")
#install.packages("corrplot")
#devtools::install_github("caijun/ggcorrplot2")
#remove.packages("GenomicSEM")
#devtools::install_github("MichelNivard/GenomicSEM",ref = 'v2.1') #specify branch for better stability
#devtools::install_github("MichelNivard/GenomicSEM") #master branch as default
#remove.packages("GenomicSEM")
#devtools::install_github("johanzvrskovec/GenomicSEM",ref = 'mod-jz') #specify branch
#devtools::install_github("GenomicSEM/GenomicSEM")
#remove.packages("HDL")
#devtools::install_github("zhenin/HDL/HDL")
#devtools::install_github("zhenin/HDL/HDL@77cb9d0984d1302e40bfd871491e292f8f09f49d") #specify exact commit
#remove.packages("BiocManager")
#install.packages("BiocManager")
#remove.packages("MungeSumstats")
#devtools::install_github("neurogenomics/MungeSumstats")
#remove.packages("SNPlocs.Hsapiens.dbSNP144.GRCh38")
#BiocManager::install("SNPlocs.Hsapiens.dbSNP144.GRCh38")
#remove.packages("BSgenome.Hsapiens.NCBI.GRCh38")
#BiocManager::install("BSgenome.Hsapiens.NCBI.GRCh38")
#install.packages("optparse")
#remove.packages("shru")
#devtools::install_github("johanzvrskovec/shru")
#remotes::install_github("johanzvrskovec/shru") #install without devtools and without compiling c++ parts
#install.packages("reticulate")
#install.packages("readr")  
#install.packages("GGally")
#install.packages("FDRestimation")
#devtools::install_github("tnggroup/genetic_correlations",ref = 'dev_jz')
#devtools::install_github("tnggroup/genetic_correlations",ref = 'dev_jz', auth_token = "YOUR_PAT")
#install.packages("MASS")

library(R.utils) #for debugging supermunge
library(stringr) #for debugging Genomic SEM ldsc
library(readr) #for debugging Genomic SEM ldsc
library(corrplot) #do not run on cluster, does not work?
#library(ggcorrplot2) #this does not print coefficients on the boxes
#library(disk.frame)
#library(GGally)
library(data.table)
library(optparse)
library(skimr)
library(ggrepel)
library(gt)
library(psych)
library(Matrix)
library(stats)
#library(tidyverse)
#library(HDL)
#library(MungeSumstats)
library(shru)
library(GenomicSEM)
require(msm) #for the deltamethod function used for GWIS
#require(FDRestimation)
library(MASS)


#library(reticulate)

```

```{r command line setup}
clParser <- OptionParser()
clParser <- add_option(clParser, c("-t", "--task"), type="character", default="0",
                help="Index of the explicit task to run separately:\n0: No task\nmvLD.mvLDSC:multivariate LDSC\nmvLD.HDL.piecewise:HDL Piecewise\nmvLD.HDL.jackknife:HDL Jackknife\nmvLD.origHDL:original HDL(jackknife)\nmvLD.origHDL.liabilityScale:original HDL with applied liability scale [default %default]")
clParser <- add_option(clParser, c("-l", "--location"), type="character", default="local",
                help="The place where the code is run [local,cluster] [default %default]")

clParser <- add_option(clParser, c("-a", "--task_argument"), type="character", default=NA,
                help="General purpose argument for tasks [default %default]")

```

# Script settings
```{r settings}
p<-c() #create project metadata object
p$clOptions<-parse_args(clParser)
p$date.run<-Sys.Date()
p$setup.version<-8
p$setup.code<-paste0("setup",p$setup.version)
p$setup.code.date<-paste0(p$setup.code,"_",p$date.run)
p$filename.rmd<-paste0(p$setup.code,".Rmd")
p$filename.r<-paste0(p$setup.code,".R")
p$functions<-c()
p$nThread=6
setDTthreads(p$nThreads) #configure DT threads accordingly


p$host<-p$clOptions$location #this is the place where the code is run [local,cluster] read from command line - default local
#p$host<-"cluster" #hard coded setting
p$seting.refreshPrepareSummaryStatistics<-FALSE
p$setting.refreshLatentFactorGWAS<-FALSE

#color theme settings
theme.color<-c()
theme.color$contrastDark1<-"#2D2D2D"
theme.color$contrastDark2<-"#CC99CC"
theme.color$contrastDark3<-"#6699CC"
theme.color$contrastDark4<-"#99CC99"
theme.color$contrastLight1<-"#66CCCC"
theme.color$contrastLight2<-"#FFCC66"
theme.color$contrastLight3<-"#F99157"
theme.color$contrastLight4<-"#F2777A"

#file path settings
##set project shared working directory **change if you have other settings**
if(p$host=="local") {
p$folderpath<-normalizePath("/Users/jakz/project/JZ_GED_PHD_C1")
} else if (p$host=="cluster") {
p$folderpath<-normalizePath("/scratch/users/k19049801/project/JZ_GED_PHD_C1")
}
##project working directory subfolders
p$folderpath.workingDirectory<-normalizePath(file.path(p$folderpath,"working_directory"))

p$folderpath.scripts<-normalizePath(file.path(p$folderpath,"scripts"))
#p$folderpath.includedSoftware<-normalizePath(file.path(p$folderpath,"included_software"))
p$folderpath.plots<-normalizePath(file.path(p$folderpath,"plots"))

#general data folder
p$folderpath.data<-normalizePath(file.path(p$folderpath,"data"))

##cleaned sumstats folder
p$folderpath.data.sumstats.cleaned<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","cleaned"))

##munged sumstats folder
p$folderpath.data.sumstats.munged<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_eur_supermunge"))

p$folderpath.data.sumstats.munged.supermunge.1kg.orig.unfiltered<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_orig_eur_supermunge_unfiltered"))

#special ldsc.orig munge folders
p$folderpath.data.sumstats.munged.ldsc.hm3.plainN<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged"))
p$folderpath.data.sumstats.munged.ldsc.hm3<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_hm3_eur_supermunge_ldsc_maf_0.01")) #we need newly munged files to use effective N
p$folderpath.data.sumstats.munged.supermunge.hm3.unfiltered<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_hm3_eur_supermunge_unfiltered")) #we need newly munged files to use effective N
p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_01<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_orig_eur_supermunge_ldsc_maf_0.01"))
p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_05<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_orig_eur_supermunge_ldsc_maf_0.05"))
p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_05.info0<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_orig_eur_supermunge_ldsc_maf_0.05_info_0"))
p$folderpath.data.sumstats.munged.ldsc.1kg.maf0_05<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_eur_supermunge_ldsc_maf_0.05"))
p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_1<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_orig_eur_supermunge_ldsc_maf_0.1"))
p$folderpath.data.sumstats.munged.ldsc.1kg.maf0_1<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_eur_supermunge_ldsc_maf_0.1"))

##munged sumstats folder
p$folderpath.data.sumstats.munged.4ssimp<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","munged_1kg_eur_supermunge.4ssimp"))

##imputed sumstats folder
#p$folderpath.data.sumstats.imputed<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","imputed_1kg_eur_supermunge"))
p$folderpath.data.sumstats.imputed<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","imputed_1kg_eur_supermunge.500Kbwin"))

p$folderpath.data.sumstats.imputed.05cm<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","imputed_1kg_eur_supermunge.0.5cMwin"))

# p$folderpath.data.sumstats.imputed.10<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","imputed_1kg_eur_supermunge.10Kwin"))
# 
# p$folderpath.data.sumstats.imputed.500<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","imputed_1kg_eur_supermunge.500Kwin.validated"))

##imputed sumstats folder, SSIMP
p$folderpath.data.sumstats.imputed.ssimp<-normalizePath(file.path(p$folderpath.data,"gwas_sumstats","imputed_1kg_eur_ssimp"))



p$filename.suffix.data.sumstats.munged<-".gz"
#p$filename.suffix.data.sumstats.munged<-"_noMHC.sumstats.gz"


##imputed sumstats folder
#p$folderpath.data.sumstats.imputed<-normalizePath(file.path(p$folderpath.data,"gwas_sumstat","imputed"))

##export sumstats folders
p$folderpath.data.sumstats.export<-normalizePath(file.path(p$folderpath.data,"export","gwas"))
p$folderpath.data.sumstats.export.cleaned.nohighld<-normalizePath(file.path(p$folderpath.data,"export","gwas.cleaned.nohighld"))
p$folderpath.data.sumstats.export.500.cleaned.nohighld<-normalizePath(file.path(p$folderpath.data,"export","gwas.500.cleaned.nohighld"))
p$folderpath.data.sumstats.export.ssimp.cleaned.nohighld<-normalizePath(file.path(p$folderpath.data,"export","gwas.ssimp.cleaned.nohighld"))


#python virtual environment folder
if(p$host=="local") {
  p$folderpath.pythonVenv<-normalizePath("~/Documents/local_db/JZ_GED_PHD_C1/python-venv")
} else if (p$host=="cluster") {
  p$folderpath.pythonVenv<-normalizePath(file.path(p$folderpath,"python-venv"))
}

#variant lists folder
p$folderpath.data.variantLists<-normalizePath(file.path(p$folderpath.data,"variant_lists"))


##Reference SNP-list (HapMap3 SNPs for example). Used for munging sumstat SNP data.
p$filepath.SNPReference.hm3<-normalizePath(file.path(p$folderpath.data.variantLists,"w_hm3.snplist.flaskapp2018")) #HapMap3 SNPs
## Used in the preparation step for performing latent factor GWAS as reference for calculating SNP variance across traits.
#p$filepath.SNPReference<-normalizePath(paste0(p$folderpath.data,"/","reference.1000G.maf.0.005.txt")) #1000 genomes phase 3
p$filepath.SNPReference.1kg<-normalizePath(file.path(p$folderpath.data.variantLists,"1kgp3.bX.eur.l2.jz2023.gz")) 
#p$filepath.SNPReference.1kg<-normalizePath(file.path("/scratch/prj/gwas_sumstats/1kg_bed","1KG_Phase3.WG.CLEANED.EUR_MAF001.CM23.bim"))

p$filepath.SNPReference.hc1kg<-normalizePath(file.path(p$folderpath.data.variantLists,"hc1kgp3.b38.eur.l2.jz2024.gz")) #Based on the high coverage 1kgp3 panel in b38

#p$filepath.SNPReference.hc1kg<-normalizePath(file.path(p$folderpath.data.variantLists,"combined.hm3_1kg.snplist.vanilla.jz2020.txt")) #custom hm3 + 1kg SNPs, based on individual level data in /users/k1204688/brc_scratch/Public/1KG_Phase3


p$filepath.rsSynonyms.dbSNP151<-normalizePath(file.path(p$folderpath.data.variantLists,"hc1kgp3.b38.jz2024.synonyms.gz"))
#p$filepath.rsSynonyms.dbSNP151<-normalizePath(file.path(p$folderpath.data.variantLists,"dbsnp151.synonyms.gz"))

#reference panels
p$folderpath.data.referencePanels<-normalizePath(file.path(p$folderpath.data,"reference_panel"))

##Reference panel folder containing individual level data reference panel. Used for GWAS sumstat imputation tasks.
p$folderpath.data.sumstatImp.genomeReference<-file.path(p$folderpath.data.referencePanels,"hc1kgp3.b38.plink")

#p$folderpath.data.sumstatImp.genomeReference<-"/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/reference.panel.1KG_Phase3.CLEANED.EUR.cM" #source /users/k1204688/brc_scratch/Public/1KG_Phase3

##LD scores datasets folders (these strings need to have a trailing slash for the GSEM LDSC to work)
p$folderpath.data.mvLDSC.ld.hc1kg <- file.path(p$folderpath.data,"ld_scores","hc1kgp3.b38.eur.l2.jz2024.chr")
p$folderpath.data.mvLDSC.ld.hm3 <- file.path(p$folderpath.data,"ld_scores","eur_w_ld_chr.ordered")
p$folderpath.data.mvLDSC.ld.1kg <- file.path(p$folderpath.data,"ld_scores","1KG_Phase3.WG.CLEANED.EUR_MAF001.1cm.250blocks.ordered")

#Weights, if different from LD-scores
#Set weights to the same folder as ldscores
p$folderpath.data.mvLDSC.wld.hc1kg <- p$folderpath.data.mvLDSC.ld.hc1kg
p$folderpath.data.mvLDSC.wld.1kg <- p$folderpath.data.mvLDSC.ld.1kg
p$folderpath.data.mvLDSC.wld.hm3 <- p$folderpath.data.mvLDSC.ld.hm3

##HDL LD scores reference - needs the trailing slashes!!!
if(p$host=="local") {
  #use the smallest LD reference as default for local tests
  p$folderpath.data.HDL.ld<-paste0(p$folderpath.data,"/UKB_array_SVD_eigen90_extraction/")
} else if (p$host=="cluster") {
  p$folderpath.data.HDL.ld<-paste0(p$folderpath.data,"/UKB_imputed_hm3_SVD_eigen99_extraction/")
}

##Gene mapping
p$filepath.geneMapping<-file.path(p$folderpath.data,"gene_mapping","NCBI37.3.gene.locUPDATETHISTOBUILD38")

##Transcription target maps
p$filepath.transcriptionTargetMap.mirdb <- file.path(p$folderpath.data,"gene_set","miRDB_v6.0_prediction_result.txt.gz")

##Gene sets folder
p$folderpath.geneSet<-file.path(p$folderpath.data,"gene_set")
  
##full script file paths
p$filepath.rmd<-normalizePath(file.path(p$folderpath.scripts,p$filename.rmd))
p$filepath.r<-normalizePath(file.path(p$folderpath.scripts,p$filename.r))

##CFA settings
p$CFA<-c()
p$CFA$correlation<-c("COR","ORT") #ORT, OBL, and COR
p$CFA$estimator=c("ML")
p$CFA$nFactors=c(6)


##latent factor GWAS filter settings
p$lfGWAS$info.filter=.6
p$lfGWAS$maf.filter=0.01

#working directory in case of running as an R-script
setwd(dir = normalizePath(p$folderpath.workingDirectory))

#inactivated python environment until it is used
#use_virtualenv(p$folderpath.pythonVenv)

print("Setup done!")

```

# Functions
```{r plot functions}

#standardised function to extract plotting info from ldsc covstructs
p$getCVDataframe<-function(cCovstruct,labelString,groupString,cv.S_StandLimit = 0.05){
  S_ForCV<-abs(cCovstruct$S)
  g_se <- sqrt(diag(S_ForCV))
  S_se2 <- g_se %*% t(g_se)
  S_ForCV[S_ForCV < S_se2 * cv.S_StandLimit] <- (S_se2 * cv.S_StandLimit)[S_ForCV < S_se2 * cv.S_StandLimit] #cap at min rg 0.05 to avoid extreme values
  S.CV<-(cCovstruct$S.SE/S_ForCV)
  dfOffDiagonal <- data.frame(
    s=cCovstruct$S[lower.tri(cCovstruct$S,diag = F)],
    cv=S.CV[lower.tri(S.CV,diag = F)],
    se=cCovstruct$S.SE[lower.tri(cCovstruct$S.SE,diag = F)],
    m=cCovstruct$m.nonimputed[lower.tri(cCovstruct$m.nonimputed,diag = F)],
    m.imputed=cCovstruct$m.imputed[lower.tri(cCovstruct$m.imputed,diag = F)],
    m.imputed.full=cCovstruct$m.imputed.full[lower.tri(cCovstruct$m.imputed.full,diag = F)],
    m.imputed.partial=cCovstruct$m.imputed.partial[lower.tri(cCovstruct$m.imputed.partial,diag = F)],
    blocks=cCovstruct$cov.blocks[lower.tri(cCovstruct$cov.blocks,diag = F)],
    isDiag=F
    )
  dfDiagonal <- data.frame(
    s=diag(cCovstruct$S),
    cv=diag(S.CV),
    se=diag(cCovstruct$S.SE),
    m=diag(cCovstruct$m.nonimputed),
    m.imputed=diag(cCovstruct$m.imputed),
    m.imputed.full=diag(cCovstruct$m.imputed.full),
    m.imputed.partial=diag(cCovstruct$m.imputed.partial),
    blocks=diag(cCovstruct$cov.blocks),
    isDiag=T
    )
  dfToReturn<-rbind(dfOffDiagonal,dfDiagonal)
dfToReturn$l<-labelString
dfToReturn$g<-groupString
return(dfToReturn)
}


#c(theme.color$contrastDark3,theme.color$contrastLight3)
p$printCorr <- function(corr, pmat=NULL, SE=NULL, filename, addrect = NULL, is.corr = T, number.cex = 1.0, number.digits=2, newnames=NULL, title=NULL, sig.level=c(0.05)){
  # corr = p$mvLD$S.SE.deltaimpnoimp.500
  # pmat = p$mvLD$covstruct.mvLDSC.1kg.500$cov.p
  # filename = file.path(p$folderpath.plots,"covse.deltaimpnoimp.500.png")
  # is.corr = F
  # number.digits = 4
  # newnames = p$sumstats.sel$code.nice
  
  
  #palette<-colorRampPalette(c(theme.color$contrastDark3,"#FFFFFF",theme.color$contrastLight3))
    
  if(!is.null(newnames)){
    rownames(corr)<-newnames
    colnames(corr)<-newnames
    if(!is.null(pmat)){
      rownames(pmat)<-newnames
      colnames(pmat)<-newnames
    }
  }

  corr<-corr[order(colnames(corr)),order(colnames(corr))]
  if(!is.null(pmat)) pmat<-pmat[order(colnames(pmat)),order(colnames(pmat))]
  
  # if(is.null(SE)){
  #   corr.uppCI <- NULL
  #   corr.lowCI <- NULL
  # } else {
  #   corr.uppCI<-clipValues(corr + 1.96 * SE, -1,1)
  #   corr.lowCI<-clipValues(corr - 1.96 * SE, -1,1)
  # }
  
  par.addCoef.col <- NULL
  if(!is.corr) par.addCoef.col <- 'black'
  if(is.corr) par.col.lim <- c(-1,1)
  if(!is.corr) par.col.lim <- c(min(corr), max(corr))
  if(!is.corr & all(corr>=0)) par.col.lim <- c(-max(corr),max(corr))
  if(!is.corr & all(corr<=0)) par.col.lim <- c(min(corr),-min(corr))
  # if(!is.corr & all(corr>=0)) par.col.lim <- c(-round(max(corr),digits = 0),round(max(corr),digits = 0))
  # if(!is.corr & all(corr<=0)) par.col.lim <- c(round(min(corr)-1,digits = 0),-1*(round(min(corr)-1)))

  png(filename = filename, width = 9, height = 9, units = 'in', res = 300, family = "Helvetica")
  par(xpd=TRUE) #keep labels inside margins
  pal<-colorRampPalette(c("#FF6666","#EEEEEE","#6666FF"))
  corrplot(corr = corr, method = "circle", insig='blank', order = "original", p.mat = pmat, sig.level = sig.level, pch.cex = 1.5, full_col=!is.corr, na.label = "square", na.label.col = "grey30", diag=T, addrect = addrect, is.corr = is.corr, number.cex = number.cex, number.digits = number.digits, addCoef.col = par.addCoef.col, col = pal(200), col.lim = par.col.lim, mar=c(0,0,4,0), title = title)$corrPos -> p1  #COL2('RdBu')
  
  #p1$corr <- ifelse(0==-p1$y + ncol(corr)-p1$x + 1, NA_real_, p1$corr)
  if(is.corr) text(p1$x, p1$y, round(p1$corr, number.digits)) #only full coeficients for correlation plots

  dev.off()
  
}

p$printCorrSimplified <- function(corr, SE=NULL, filename, is.corr = T, number.cex = 1.5, number.digits=2, absScale=F){
  
  if(absScale){
    palette<-colorRampPalette(c("#000000",theme.color$contrastLight3))
  } else {
    palette<-colorRampPalette(c(theme.color$contrastDark3,"#000000",theme.color$contrastLight3))
  }
  
  if(is.null(SE)){
    corr.uppCI <- NULL
    corr.lowCI <- NULL
  } else {
    corr.uppCI<-corr + 1.96 * SE
    corr.lowCI<-corr - 1.96 * SE
  }
  
  png(filename = filename, width = 1200, height = 1000)
corrplot(
  corr = corr,
  #uppCI.mat = corr.uppCI,
  #lowCI.mat = corr.lowCI,
  #plotCI = ifelse(is.null(SE),c("n"),"circle"),
  order = "original",#"hclust",
  #hclust.method = "ward.D",
  method = "color",
  type="full",
  #addCoef.col = theme.color$contrastDark1,
  addgrid.col = theme.color$contrastDark1,
  col = palette(200),
  is.corr = is.corr,
  outline = T,
  #addrect = addrect,
  #rect.col = theme.color$contrastLight1,
  #rect.lwd = 6,
  tl.cex = 1.7,
  tl.col = theme.color$contrastDark2,
  tl.srt = 35,
  cl.cex = 2,
  cl.ratio = 0.2,
  #number.cex = number.cex,
  #number.digits = number.digits
  )
dev.off()
  
}

```

# Knitr setup

```{r knitr setup, include=FALSE, purl=FALSE}
knitr::opts_chunk$set(echo = TRUE,comment=NA,prompt=FALSE,cache=FALSE)
knitr::opts_knit$set(root.dir=normalizePath(p$folderpath.workingDirectory))

```

# R script export

```{r purl export, include=FALSE, purl=FALSE, eval=FALSE}
#running this will produce an R-script with the same name as the Rmd-file. Used for running the program on computational clusters (Rosalind), when the code just has to be run rather than creating a knitted output.
knitr::purl(p$filepath.rmd)
  
```


```{r additional source setup, echo=FALSE, warning=F}

#source(normalizePath(file.path(p$folderpath.scripts,"include","ldsc.orig.R")))
#source(file.path(p$folderpath.scripts,"include","fastman.R"))
#source(file.path(p$folderpath.scripts,"include","fastqq.R"))

```


```{r old anxi rg analysis of gwas sumstats, purl=FALSE, eval=FALSE}
p$rg_anxi<-read.table(file.path(p$folderpath.data,"ANXI03_gc.txt"), header=T, quote="", sep = "") %>%
extract(col = p2, into = "code", regex = "_noMHC/(.*)_noMHC.sumstats.", remove = F, )

p$rg_anxi<- p$rg_anxi[which(p$rg_anxi$p<.00001),]
p$rg_anxi$rg.abs<-abs(p$rg_anxi$rg)

#View(p$rg_anxi)

```

# Metadata setup

## Database connection to fetch trait and GWAS metadata - run from Rstudio .rmd
```{r sumstat metadata database load}
p$filepath.sumstats<-file.path(p$folderpath.workingDirectory,paste0("sumstats.",p$setup.code,".Rds"))
if (file.exists(p$filepath.sumstats)) {
  print("Loading summary statistics metadata from previously stored file.")
  p$sumstats<-readRDS(file=p$filepath.sumstats)
} else {

#install.packages('RPostgres')
library(RPostgres)
library(DBI)

p$phenodbcon <- dbConnect(RPostgres::Postgres(),
                 dbname = 'phenodb', 
                 host = '10.200.105.5', 
                 port = 5432,
                 user = 'johan',
                 password = rstudioapi::askForPassword(prompt = "Enter database password for specified user."))

p$phenodbres <- dbSendQuery(p$phenodbcon, "SELECT \"GWAS\".*, category_id, category.name AS category_name, phenotype.name AS phenotype, phenotype.type AS phenotype_type, pmid, year FROM sumstat_old.\"GWAS\", sumstat_old.reference, sumstat_old.phenotype, sumstat_old.category 
WHERE \"GWAS\".reference_id=reference.id AND \"GWAS\".phenotype_id=phenotype.id AND phenotype.category_id = category.id
ORDER BY code,\"GWAS\".id")
p$sumstats<-dbFetch(p$phenodbres)
dbClearResult(p$phenodbres)

#fallback
#p$sumstats<-read.table(file.path(p$folderpath.data,"ukbb_sumstats_download202005.csv"), header=T, quote="\"", sep = ",", fill=T, blank.lines.skip=T,as.is = c(2), strip.white = T)

saveRDS(p$sumstats,file = p$filepath.sumstats)
write.table(p$sumstats, file = file.path(p$folderpath.workingDirectory,paste0(p$setup.code,".sumstats.tsv")), quote = F, sep = "\t", row.names = FALSE, col.names = TRUE)
}



```



## Trait setup
```{r trait setup}
#,echo=FALSE
p$trait<-data.frame(phenotype_id=c())

#ADHD
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-139
p$trait[nrow(p$trait),c("populationPrevalence")]<-.0529 #Worldwide-pooled - THIS WAS WRONG IN THE ANALYSIS THOUGH: .00529
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1176/ajp.2007.164.6.942"

#ALCD
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-141
p$trait[nrow(p$trait),c("populationPrevalence")]<-.125 #US total measurement
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1001/archpsyc.64.7.830"

#ANOR
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-142
p$trait[nrow(p$trait),c("populationPrevalence")]<-.0245 #European average (3 secondary refs)
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1038/S41588-019-0439-2"

#ANXI
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-144
p$trait[nrow(p$trait),c("populationPrevalence")]<-.16 #Any type of anxiety disorder, Via https://doi.org/10.1038/s41380-019-0559-1 (2019), originally from https://doi.org/10.1017/S1121189X00001421 (2009)
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1017/S1121189X00001421"

#AUTI
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-145
p$trait[nrow(p$trait),c("populationPrevalence")]<-.0122
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1186/s12874-016-0280-6"

#BIPO - also wrong in the analysis
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-146
p$trait[nrow(p$trait),c("populationPrevalence")]<- .025 #LT
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1002/mpr.1359"
# p$trait[nrow(p$trait),c("populationPrevalence")]<- (40 * 0.007)/2 #mean of male and female global prevalence rate (2013) from https://doi.org/10.1111/bdi.12423 (2016), assuming sample mean age of 40y. THIS IS PROBABLY WRONG - TOO HIGH?
# p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1111/bdi.12423"

#DEPR
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-149
p$trait[nrow(p$trait),c("populationPrevalence")]<-.146 #MDD, .15 from the LD-calculations in https://doi.org/10.1038/s41588-018-0090-3 (2018), but with a possible reference to https://doi.org/10.1146/annurev-publhealth-031912-114409 (2013) which states 14.6% lifetime prevalence of MDE in high-income countries.
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1146/annurev-publhealth-031912-114409"

#DEPR - broad
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-1499 #this is just an adhoc id
p$trait[nrow(p$trait),c("populationPrevalence")]<-.302 #MDD, .15 from the partitioned heritability analysis in https://doi.org/10.1038/s41593-018-0326-7
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1038/s41593-018-0326-7"

#INSO
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-171
p$trait[nrow(p$trait),c("populationPrevalence")]<-.69 #using the 'prevalence', the higher estimate of persistent symptoms after 1y follow up.
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1093/sleep/30.3.274"

#PTSD
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-140
# p$trait[nrow(p$trait),c("populationPrevalence")]<-.3 #using the moderate estimate of lifetime prevalence GIVEN TRAUMA HAD TAKEN PLACE as used in ref
# p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1038/s41467-019-12576-w"
# #alt without given trauma
p$trait[nrow(p$trait),c("populationPrevalence")]<-.057 #LT
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1002/mpr.1359"

#SCHI
p$trait[nrow(p$trait)+1,c("phenotype_id")]<-151
p$trait[nrow(p$trait),c("populationPrevalence")]<-.0072 #using the lifetime morbid risk estimate, slightly more suitable for the CLOZUK sample as described in the GWAS but not using the lower point estimate of 0.4%
p$trait[nrow(p$trait),c("referenceDOI")]<-"https://doi.org/10.1093/epirev/mxn001"

p$trait
setDT(p$trait)
setkeyv(p$trait, cols = c("phenotype_id"))

```


## Setup GWAS summary statistics datasets

```{r setup GWAS sumstat dataset}

#, echo=FALSE


#View(p$sumstats)

#rename and add columns
names(p$sumstats)[names(p$sumstats)=="n_cases"]<-"n_case"
names(p$sumstats)[names(p$sumstats)=="n_controls"]<-"n_control"
p$sumstats$gwas_name.nice<-NA_character_
p$sumstats$code.trait<-NA_character_
p$sumstats$reference_doi<-NA_character_
p$sumstats$effect.logit<-as.logical(NA)
p$sumstats$dependent_variable.linprob<-as.logical(NA)
p$sumstats$se.logit<-as.logical(NA)
p$sumstats$dependent_variable.OLS<-as.logical(NA)
p$sumstats$age.min<-NA_integer_
p$sumstats$age.max<-NA_integer_
p$sumstats$age.mean<-NA_real_
p$sumstats$age.sd<-NA_real_

#add missing datasets and data
p$sumstats[nrow(p$sumstats)+1,c("code","n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="DEPR05",
  n_case=16823,
  n_control=25632,
  n_total=42455,
  phenotype_id=149,
  reference_id=127,
  reference_doi=c("https://doi.org/10.1038/s41588-018-0090-3")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="DEPR12",
  n_case=14260,
  n_control=15480,
  n_total=29740,
  phenotype_id=149,
  reference_id=127,
  reference_doi=c("https://doi.org/10.1038/s41588-018-0090-3")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="NEUR04",
  name="Neuroticism",
  n_case=449484,
  n_control=NA_integer_,
  n_total=449484,
  phenotype_id=176,
  reference_id=NA_integer_,
  reference_doi=c("https://doi.org/10.1038/s41588-018-0151-7")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="HEIG09",
  name="Height",
  n_case=3612229,
  n_control=NA_integer_,
  n_total=3612229,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_,
  reference_doi=c("https://doi.org/10.1038/s41586-022-05275-y")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="BODYTM",
  name="2013 BMI, male",
  n_case=60586,
  n_control=NA_integer_,
  n_total=60586,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_,
  reference_doi=c("10.1371/journal.pgen.1003500")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="BODY14",
  name="2019 BMI",
  n_case=806834,
  n_control=NA_integer_,
  n_total=806834,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_,
  reference_doi=c("https://doi.org/10.1093/hmg/ddy327")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="BODYTF",
  name="2013 BMI, female",
  n_case=73137,
  n_control=NA_integer_,
  n_total=73137,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_,
  reference_doi=c("10.1371/journal.pgen.1003500")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="BODYT2M",
  name="GWIS of BMI, male",
  n_case=60586, #this is the N of the BMI GWAS of the same sample
  n_control=NA_integer_,
  n_total=60586,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_,
  reference_doi=c("10.1016/j.ajhg.2016.07.020")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="BODYT2MI",
  name="GWIS of BMI, male, imputed H&W",
  n_case=60586, #this is the N of the BMI GWAS of the same sample
  n_control=NA_integer_,
  n_total=60586,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_,
  reference_doi=c("10.1016/j.ajhg.2016.07.020")
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="SMRV01",
  name="Simulated trait 1",
  n_case=36746,
  n_control=NA_integer_,
  n_total=36746,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_
  )

p$sumstats[nrow(p$sumstats)+1,c("code","name", "n_case","n_control","n_total","phenotype_id","reference_id","reference_doi")]=list(
  code="SMRV02",
  name="Simulated trait 2",
  n_case=36746,
  n_control=NA_integer_,
  n_total=36746,
  phenotype_id=NA_integer_,
  reference_id=NA_integer_
  )

#reformat columns
p$sumstats$name<-as.character(p$sumstats$name)

#set datatable
setDT(p$sumstats)
setkeyv(p$sumstats, cols = c("code"))

##set specific trait data corrections - before join!
p$sumstats[which(p$sumstats$code=="DEPR09"),]$phenotype_id<-1499

##Add trait/disorder information
p$sumstats<-as.data.frame(p$sumstats[p$trait,on=c('phenotype_id'),c('populationPrevalence','referenceDOI') :=list(i.populationPrevalence,i.referenceDOI)]) #reset to dataframe

#set code as rowname
rownames(p$sumstats)<-p$sumstats$code


##Add comprehensive names as in the Google sheet
p$sumstats$name[which(p$sumstats$code=="DEPR05")]="Major depressive disorder (PGC2 29) - only clinical ascertainment"


##Add sumstat full file paths
p$sumstats$cleanedpath<-file.path(p$folderpath.data.sumstats.cleaned,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath<-file.path(p$folderpath.data.sumstats.munged,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.supermunge.1kg.orig.unfiltered<-file.path(p$folderpath.data.sumstats.munged.supermunge.1kg.orig.unfiltered,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$imputedpath<-file.path(p$folderpath.data.sumstats.imputed,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$imputedpath.05cm<-file.path(p$folderpath.data.sumstats.imputed.05cm,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$imputedpath.ssimp<-file.path(p$folderpath.data.sumstats.imputed.ssimp,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.hm3.plainN<-file.path(p$folderpath.data.sumstats.munged.ldsc.hm3.plainN,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.hm3<-file.path(p$folderpath.data.sumstats.munged.ldsc.hm3,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.supermunge.hm3.unfiltered<-file.path(p$folderpath.data.sumstats.munged.supermunge.hm3.unfiltered,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.1kg.orig.maf0_05<-file.path(p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_05,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.1kg.orig.maf0_05.info0<-file.path(p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_05.info0,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.1kg.orig.maf0_01<-file.path(p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_01,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.1kg.orig.maf0_1<-file.path(p$folderpath.data.sumstats.munged.ldsc.1kg.orig.maf0_1,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.1kg.maf0_05<-file.path(p$folderpath.data.sumstats.munged.ldsc.1kg.maf0_05,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath.ldsc.1kg.maf0_1<-file.path(p$folderpath.data.sumstats.munged.ldsc.1kg.maf0_1,paste0(p$sumstats$code,".sumstats",p$filename.suffix.data.sumstats.munged))

p$sumstats$mungedpath2<-file.path(p$folderpath.data.sumstats.export.cleaned.nohighld,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$imputedpath2.500<-file.path(p$folderpath.data.sumstats.export.500.cleaned.nohighld,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))

p$sumstats$imputedpath2.ssimp<-file.path(p$folderpath.data.sumstats.export.ssimp.cleaned.nohighld,paste0(p$sumstats$code,p$filename.suffix.data.sumstats.munged))


if(p$host=="local"){
  p$sumstats["BODY14",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/bmi.giant-ukbb.meta-analysis.combined.23May2018.txt.gz"
} else {
  p$sumstats["BODY14",]$cleanedpath<-"/users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/raw/bmi.giant-ukbb.meta-analysis.combined.23May2018.txt.gz"
}

if(p$host=="local"){
  p$sumstats["BODYTM",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_BMI_MEN_N.txt.gz"
} else {
  p$sumstats["BODYTM",]$cleanedpath<-"/scratch/prj/gwas_sumstats/new/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_BMI_MEN_N.txt.gz"
}

if(p$host=="local"){
  p$sumstats["BODYTF",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_BMI_WOMEN_N.txt.gz"
} else {
  p$sumstats["BODYTF",]$cleanedpath<-"/scratch/prj/gwas_sumstats/new/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_BMI_WOMEN_N.txt.gz"
}

if(p$host=="local"){
  p$sumstats["NEUR04",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/sumstats_neuroticism_ctg_format.txt.gz"
} else {
  p$sumstats["NEUR04",]$cleanedpath<-"/users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/raw/sumstats_neuroticism_ctg_format.txt.gz"
}

if(p$host=="local"){
  p$sumstats["INSO03",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/insomnia_ukb2b_EUR_sumstats_20190311_with_chrX_mac_100.txt.gz"
} else {
  p$sumstats["INSO03",]$cleanedpath<-"/users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/raw/insomnia_ukb2b_EUR_sumstats_20190311_with_chrX_mac_100.txt.gz"
}

if(p$host=="local"){
  p$sumstats["DEPR12",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/PGC2_MDD_Wray/daner_ukb_170227_aligned.assoc.gz"
} else {
  p$sumstats["DEPR12",]$cleanedpath<-"/scratch/prj/gwas_sumstats/original/PGC2_MDD_Wray/daner_ukb_170227_aligned.assoc.gz"
}

if(p$host=="local"){
  p$sumstats["HEIG09",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/GIANT_HEIGHT_YENGO_2022_GWAS_SUMMARY_STATS_EUR.gz"
} else {
  p$sumstats["HEIG09",]$cleanedpath<-"/scratch/prj/gwas_sumstats/new/GIANT_HEIGHT_YENGO_2022_GWAS_SUMMARY_STATS_EUR.gz"
}

if(p$host=="local"){
  p$sumstats["HEIG04F",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_HEIGHT_WOMEN_N.txt"
} else {
  p$sumstats["HEIG04F",]$cleanedpath<-"/scratch/prj/gwas_sumstats/original/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_HEIGHT_WOMEN_N.txt"
}

if(p$host=="local"){
  p$sumstats["HEIG04M",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_HEIGHT_MEN_N.txt"
} else {
  p$sumstats["HEIG04M",]$cleanedpath<-"/scratch/prj/gwas_sumstats/original/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_HEIGHT_MEN_N.txt"
}

if(p$host=="local"){
  p$sumstats["WEIG04F",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_WEIGHT_WOMEN_N.txt"
} else {
  p$sumstats["WEIG04F",]$cleanedpath<-"/scratch/prj/gwas_sumstats/original/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_WEIGHT_WOMEN_N.txt"
}

if(p$host=="local"){
  p$sumstats["WEIG04M",]$cleanedpath<-"/Users/jakz/Documents/local_db/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/raw/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_WEIGHT_MEN_N.txt"
} else {
  p$sumstats["WEIG04M",]$cleanedpath<-"/scratch/prj/gwas_sumstats/original/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_WEIGHT_MEN_N.txt"
}




##add reference year - UPDATE THESE!!
p$sumstats["ANXI03",]$year=2019
p$sumstats["BODY14",]$year=2019
p$sumstats["BODYTM",]$year=2013
p$sumstats["DEPR05",]$year=2018
p$sumstats["DEPR12",]$year=2018
p$sumstats["NEUR04",]$year=2018


##Add doi links for easy access to dataset publication
p$sumstats["ALCD03",]$reference_doi="https://doi.org/10.1038/s41593-018-0275-1"
p$sumstats["ANXI03",]$reference_doi="https://doi.org/10.1038/s41380-019-0559-1"
p$sumstats["BODY11",]$reference_doi="https://doi.org/10.1093/hmg/ddy271"
p$sumstats["DEPR09",]$reference_doi="https://doi.org/10.1038/s41593-018-0326-7"
p$sumstats["EDUC03",]$reference_doi="https://doi.org/10.1038/s41588-018-0147-3"
p$sumstats["HEAL01",]$reference_doi="https://doi.org/10.1093/ije/dyw219"


##Add PMID
p$sumstats["ANXI03",]$pmid="31748690"
p$sumstats["DEPR05",]$pmid="29700475"

##add dependent variable type
p$sumstats$dependent_variable[which(p$sumstats$code=="DEPR05")]="binary"

##add participant numbers
p$sumstats$n_total[which(p$sumstats$code=="NEUR01")]=170911
p$sumstats$n_case[which(p$sumstats$code=="HEIG04F" | p$sumstats$code=="HEIG04M" | p$sumstats$code=="WEIG04F" | p$sumstats$code=="WEIG04M")]=60586
p$sumstats$n_case[which(p$sumstats$code=="EDUC03")]=766345 #not sure why it is missing here
p$sumstats$n_total[which(p$sumstats$code=="EDUC03")]=766345

##add sex details
p$sumstats$sex[which(p$sumstats$code=="DEPR05")]="both"

##add age range
p$sumstats$age.min[which(p$sumstats$code=="SUBJ01")]=40
p$sumstats$age.max[which(p$sumstats$code=="SUBJ01")]=73
p$sumstats$age.mean[which(p$sumstats$code=="SUBJ01")]=56.91
p$sumstats$age.sd[which(p$sumstats$code=="SUBJ01")]=7.93
p$sumstats$age.min[which(p$sumstats$code=="TIRE01")]=40
p$sumstats$age.max[which(p$sumstats$code=="TIRE01")]=73
p$sumstats$age.mean[which(p$sumstats$code=="TIRE01")]=56.91
p$sumstats$age.sd[which(p$sumstats$code=="TIRE01")]=7.93


##add sample prevalence for all datasets
p$sumstats$samplePrevalence<-ifelse(is.na(p$sumstats$n_control),NA_real_, p$sumstats$n_case/p$sumstats$n_total)
p$sumstats$samplePrevalence.balanced<-ifelse(is.na(p$sumstats$n_control),NA_real_, 0.5)

##add number of cases or total column
p$sumstats$n_case_total<-ifelse(is.na(p$sumstats$n_control),p$sumstats$n_case,p$sumstats$n_total)

##Add nice trait names to be used in the report
p$sumstats$name.nice<-p$sumstats$name
#p$sumstats$name.nice[which(p$sumstats$code=="ADHD05")]="ADHD"
p$sumstats$name.nice[which(p$sumstats$code=="ALCD03")]="Alcohol dependence"
p$sumstats$name.nice[which(p$sumstats$code=="ANOR02")]="Anorexia nervosa"
p$sumstats$name.nice[which(p$sumstats$code=="ANXI03")]="Anxiety disorder"
p$sumstats$name.nice[which(p$sumstats$code=="ANXI04")]="Generalised anxiety symptoms"
p$sumstats$name.nice[which(p$sumstats$code=="AUTI07")]="Autism spectrum disorder"
p$sumstats$name.nice[which(p$sumstats$code=="BODY11")]="2018 BMI"
p$sumstats$name.nice[which(p$sumstats$code=="BODY14")]="BMI"
p$sumstats$name.nice[which(p$sumstats$code=="DEPR05")]="MDD"
p$sumstats$name.nice[which(p$sumstats$code=="DEPR12")]="MDD symptoms"
p$sumstats$name.nice[which(p$sumstats$code=="DEPR08")]="MDD"
p$sumstats$name.nice[which(p$sumstats$code=="DEPR09")]="Broad depression"
p$sumstats$name.nice[which(p$sumstats$code=="EDUC03")]="Educational attainment"
p$sumstats$name.nice[which(p$sumstats$code=="EXTR01")]="Extraversion"
p$sumstats$name.nice[which(p$sumstats$code=="HEIG04F")]="2013 Height, female"
p$sumstats$name.nice[which(p$sumstats$code=="HEIG04M")]="2013 Height, male"
p$sumstats$name.nice[which(p$sumstats$code=="INCO03")]="Social deprivation"
p$sumstats$name.nice[which(p$sumstats$code=="INSO02")]="Insomnia"
p$sumstats$name.nice[which(p$sumstats$code=="INTE03")]="Cognitive ability"
p$sumstats$name.nice[which(p$sumstats$code=="LONG07")]="Longevity"
p$sumstats[c("NEUR01","NEUR04"),]$name.nice<-"Neuroticism"
p$sumstats$name.nice[which(p$sumstats$code=="RISK01")]="General risk tolerance"
p$sumstats$name.nice[which(p$sumstats$code=="RISK02")]="Risktaking, automobile"
p$sumstats$name.nice[which(p$sumstats$code=="RISK03")]="Risktaking, sex"
p$sumstats$name.nice[which(p$sumstats$code=="SCHI04")]="Schizophrenia"
p$sumstats$name.nice[which(p$sumstats$code=="SUBJ01")]="Subjective well-being"
p$sumstats$name.nice[which(p$sumstats$code=="TIRE01")]="Self-reported tiredness"
p$sumstats$name.nice[which(p$sumstats$code=="WEIG04F")]="2013 Weight, female"
p$sumstats$name.nice[which(p$sumstats$code=="WEIG04M")]="2013 Weight, male"

##Add trait codes for publication rather than internal codes
p$sumstats$code.nice<-p$sumstats$code
p$sumstats["ADHD05",]$code.nice<-"ADHD"
p$sumstats["ANOR02",]$code.nice<-"ANOR"
p$sumstats["ANXI03",]$code.nice<-"ANXD"
p$sumstats["ANXI04",]$code.nice<-"ANXS"
p$sumstats["AUTI07",]$code.nice<-"AUTI"
p$sumstats["BODY11",]$code.nice<-"BMI2018"
p$sumstats["BODY14",]$code.nice<-"BMI"
p$sumstats["BODYTM",]$code.nice<-"BMI2013M"
p$sumstats["BIPO02",]$code.nice<-"BIPD"
p$sumstats["DEPR05",]$code.nice<-"MDD"
p$sumstats["DEPR12",]$code.nice<-"MDDS"
p$sumstats["DEPR09",]$code.nice<-"MDDB"
p$sumstats["EDUC03",]$code.nice<-"EDUA"
p$sumstats["EXTR01",]$code.nice<-"EXTR"
p$sumstats["HEIG04F",]$code.nice<-"H2013F"
p$sumstats["HEIG04M",]$code.nice<-"H2013M"
p$sumstats["BODYT2M",]$code.nice<-"BMIGWIS"
p$sumstats["BODYT2MI",]$code.nice<-"BMIGWISI"
p$sumstats["INCO03",]$code.nice<-"SDEP"
p$sumstats["INSO02",]$code.nice<-"INSO"
p$sumstats["INTE03",]$code.nice<-"COGA"
p$sumstats["LONG07",]$code.nice<-"LONG"
p$sumstats["NEUR01",]$code.nice<-"NEUR1"
p$sumstats["NEUR04",]$code.nice<-"NEUR"
p$sumstats["RISK01",]$code.nice<-"RISK"
p$sumstats["RISK02",]$code.nice<-"RCAR"
p$sumstats["RISK03",]$code.nice<-"RSEX"
p$sumstats["SCHI04",]$code.nice<-"SCHI"
p$sumstats["SMRV01",]$code.nice<-"SIM1"
p$sumstats["SMRV02",]$code.nice<-"SIM2"
p$sumstats["SUBJ01",]$code.nice<-"SUBJ"
p$sumstats["TIRE01",]$code.nice<-"TIRE"
p$sumstats["WEIG04F",]$code.nice<-"W2013F"
p$sumstats["WEIG04M",]$code.nice<-"W2013M"

#INFO (genotype imputation) metadata
p$sumstats$hasinfo<-NA
p$sumstats["ANXI03",]$hasinfo<-T
p$sumstats["BIPO02",]$hasinfo<-T
p$sumstats["BODY14",]$hasinfo<-T
p$sumstats["BODYTM",]$hasinfo<-F
p$sumstats["DEPR05",]$hasinfo<-T
p$sumstats["DEPR12",]$hasinfo<-T
p$sumstats["DEPR09",]$hasinfo<-F
p$sumstats["EDUC03",]$hasinfo<-F
p$sumstats["BODYT2MI",]$hasinfo<-F
p$sumstats["INSO03",]$hasinfo<-T
p$sumstats["NEUR04",]$hasinfo<-T
p$sumstats["SCHI04",]$hasinfo<-F


##Add a combined nice name plus code label
p$sumstats$name.nice.and_code<-paste0(p$sumstats$name.nice," (",p$sumstats$code.nice,")")

##Add nice dataset citation
p$sumstats$reference.nice<-NA_character_
p$sumstats["ADHD05",]$reference.nice <- "Demontis et al."
p$sumstats["ALCD03",]$reference.nice <- "Walters et al."
p$sumstats["ANOR02",]$reference.nice <- "Watson et al."
p$sumstats["ANXI03",]$reference.nice <- "Purves et al."
p$sumstats["ANXI04",]$reference.nice <- "Purves et al."
p$sumstats["AUTI07",]$reference.nice <- "Grove et al."
p$sumstats["BODY11",]$reference.nice <- "Yengo et al."
p$sumstats["BODY14",]$reference.nice <- "Pulit et al."
p$sumstats["BODYTM",]$reference.nice <- "Randall et al."
p$sumstats["BIPO02",]$reference.nice <- "Stahl et al."
p$sumstats["BVOL01",]$reference.nice <- "Jansen et al."
p$sumstats["DEPR05",]$reference.nice <- "Wray et al."
p$sumstats["DEPR12",]$reference.nice <- "Wray et al."
p$sumstats["DEPR08",]$reference.nice <- "Wray et al."
p$sumstats["DEPR09",]$reference.nice <- "Howard et al."
p$sumstats["EDUC03",]$reference.nice <- "Lee et al."
p$sumstats["HEAL01",]$reference.nice <- "Harris et al."
p$sumstats["HEIG04F",]$reference.nice <- "Randall et al."
p$sumstats["HEIG04M",]$reference.nice <- "Randall et al."
p$sumstats["HEIG09",]$reference.nice <- "Yengo et al."
p$sumstats["INCO03",]$reference.nice <- "Hill et al."
p$sumstats["INSO02",]$reference.nice <- "Jansen et al."
p$sumstats["NEUR04",]$reference.nice <- "Nagel et al."
p$sumstats["PTSD04",]$reference.nice <- "Nievergelt et al."
p$sumstats["RISK02",]$reference.nice <- "Karlsson Linnér et al."
p$sumstats["RISK03",]$reference.nice <- "Karlsson Linnér et al."
p$sumstats["SCHI04",]$reference.nice <- "Pardiñas et al."
p$sumstats["SUBJ01",]$reference.nice <- "Okbay et al."
p$sumstats["TIRE01",]$reference.nice <- "Deary et al."
p$sumstats["WEIG04F",]$reference.nice <- "Randall et al."
p$sumstats["WEIG04M",]$reference.nice <- "Randall et al."

#p$sumstats[which(p$sumstats$code=="ANXI04"),]$populationPrevalence<-0.2 #set this to larger than for ANXI03 as this phenotype is broader
#p$sumstats[which(p$sumstats$code=="DEPR08"),]$populationPrevalence<-0.2 #set this to larger than for DEPR05 as this phenotype is broader

#set order of datasets to sorted by code
p$sumstats<-p$sumstats[order(p$sumstats$code),]


#read manual metadata describing imputation stats etc
p$filepath.manualmeta <- file.path(p$folderpath.data,"manualAnnotations.tsv.txt")
if(file.exists(p$filepath.manualmeta)){
  manualMeta <- as.data.frame(shru::readFile(filePath = p$filepath.manualmeta))
  rownames(manualMeta)<-manualMeta$trait
  manualMeta$trait<-NULL
  p$sumstats[rownames(manualMeta),colnames(manualMeta)]<-manualMeta[rownames(manualMeta),colnames(manualMeta)]
  rm(manualMeta)
}


# p$filepath.manualmeta.impstats <- file.path(p$folderpath.data,"impstats.tsv.txt")
# 
# if(file.exists(p$filepath.manualmeta.impstats)){
#   manualMeta <- as.data.frame(shru::readFile(filePath = p$filepath.manualmeta.impstats))
#   rownames(manualMeta)<-manualMeta$trait
#   p$sumstats[rownames(manualMeta),c("h_cpu_ldimp","h_cpu_ssimp", "variants_raw","variants_ldimp","variants_ssimp")]<-manualMeta[rownames(manualMeta),c("h_ldimp","h_ssimp","n","n_ldimp","n_ssimp")]
#   rm(manualMeta)
# }



#save the project data
#saveRDS(project,file = file.path(p$folderpath.workingDirectory,paste0("project.",p$setup.code,".Rds")))

#View(p$sumstats)

```

## Setup gene sets
```{r Setup gene sets}
p$genesets<-data.frame(matrix(data=NA,nrow = 0,ncol = 0))

p$genesets["gtex8_rnaseq_median",c("name","filepath.raw")]<-c("GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct",file.path(p$folderpath.geneSet,"GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz"))

p$genesets["brainspan_rnaseq_mean",c("name","filepath.raw")]<-c("BRAINSPAN_RNA-Seq_Gencode_v10_summarized_to_genes_genes_matrix_csv",file.path(p$folderpath.geneSet,"BRAINSPAN_RNA-Seq_Gencode_v10_summarized_to_genes_genes_matrix_csv","expression_matrix.csv"))


#p$genesets["msigdb.v7.5.1",c("name","filepath.raw")]<-c("All MSigDB",file.path(p$folderpath.geneSet,"msigdb.v7.5.1.symbols.gmt"))
p$genesets["msigdb.h.all.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Hallmark",file.path(p$folderpath.geneSet,"msigdb.h.all.v7.5.1.symbols.gmt"))
p$genesets["msigdb.c2.cp.biocarta.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Canonical pathways Biocarta",file.path(p$folderpath.geneSet,"msigdb.c2.cp.biocarta.v7.5.1.symbols.gmt"))
p$genesets["msigdb.c2.cp.kegg.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Canonical pathways KEGG",file.path(p$folderpath.geneSet,"msigdb.c2.cp.kegg.v7.5.1.symbols.gmt"))
p$genesets["msigdb.c3.mir.mirdb.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Regulatory target miRDB",file.path(p$folderpath.geneSet,"msigdb.c3.mir.mirdb.v7.5.1.symbols.gmt"))
p$genesets["msigdb.c3.tft.gtrd.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Regulatory target GTRD",file.path(p$folderpath.geneSet,"msigdb.c3.tft.gtrd.v7.5.1.symbols.gmt"))
p$genesets["msigdb.c5.all.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Ontology",file.path(p$folderpath.geneSet,"msigdb.c5.all.v7.5.1.symbols.gmt"))
p$genesets["msigdb.c7.all.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Immunologic signature",file.path(p$folderpath.geneSet,"msigdb.c7.all.v7.5.1.symbols.gmt"))
p$genesets["msigdb.c8.all.v7.5.1",c("name","filepath.raw")]<-c("MSigDB Cell type",file.path(p$folderpath.geneSet,"msigdb.c8.all.v7.5.1.symbols.gmt"))

```

# Variable selection and project configuration
```{r GWAS sumstat dataset variable selection}

#selection based on specific traits
#p$sumstats.sel.code<-c("ANXI03","DEPR09","NEUR02")
#p$sumstats.sel.code<-c("ADHD05","ANXI04")
#p$sumstats.sel.code<-c("RISK02","RISK03","SCHI04","SUBJ01","TIRE01")
#p$sumstats.sel.code<-c("TIRE01")

#tests
#p$sumstats.sel.ssimp.code <- c("SCHI04","DEPR5B") #TEST
#p$sumstats.sel.ssimp.code <- c("ANXI03","BODY11","DEPR5B","SCHI04") #TEST
#p$sumstats.sel.ssimp.code <- c("ANXI03","BODY11","BIPO02","BVOL01","DEPR05","DEPR09","EDUC03","INSO02","NEUR02","SCHI04")
#full, but excluding BVOL01 as of now
p$sumstats.sel.ssimp.code<-c("ANXI03","BODY14","BIPO02","DEPR05","DEPR12","DEPR09","EDUC03","INSO02","NEUR04","SCHI04")
#p$sumstats.sel.ssimp.code<-c("ANXI03","BODY14") #TEST

p$sumstats.sel.infobin.code<-c("BODY14","BIPO02","DEPR05")
#p$sumstats.sel.infobin.code<-p$sumstats.sel.ssimp.code

p$sumstats.sel.bmigwas.code<-c("HEIG09","WEIG04F","WEIG04M")

#p$sumstats.sel.gwis.code<-c("GWISI","BODYTM","BODY11") #TESTING!
#p$sumstats.sel.gwis.code<-c("GWISM","BODY11")
p$sumstats.sel.gwis.code<-c("ANXI03","BODYTM","BODY14","BIPO02","DEPR05","DEPR12","DEPR09","EDUC03","BODYT2MI","INSO02","NEUR04","SCHI04")

p$sumstats.sel.raw.code<-c("ANXI03","BODYTM","BODY14","HEIG04M","WEIG04M","BIPO02","DEPR05","DEPR12","DEPR09","EDUC03","INSO02","NEUR04","SCHI04")

p$sumstats.sel.sim.code<-c("SMRV01","SMRV02")

p$sumstats.sel.code <- p$sumstats.sel.gwis.code


#all traits
#p$sumstats.sel.code<-c("ADHD05","ALCD03","ANOR02","ANXI03","ANXI04","AUTI07","BODY14","BODYTM","BODYTF","BIPO02", "BVOL01","DEPR05","DEPR5B","DEPR09","EDUC03","HEAL01","HEIG04F","HEIG04M","HEIG09","INCO03","INSO03","INSO02","NEUR02","PTSD04","RISK02","RISK03","SCHI04","SUBJ01","TIRE01","WEIG04F","WEIG04M")


p$sumstats.sel<-p$sumstats[which(p$sumstats$code %in% p$sumstats.sel.code),]
p$sumstats.sel.raw<-p$sumstats[which(p$sumstats$code %in% p$sumstats.sel.raw.code),]
p$sumstats.sel.bmigwas<-p$sumstats[which(p$sumstats$code %in% p$sumstats.sel.bmigwas.code),]
p$sumstats.sel.ssimp<-p$sumstats[which(p$sumstats$code %in% p$sumstats.sel.ssimp.code),]
p$sumstats.sel.infobin<-p$sumstats[which(p$sumstats$code %in% p$sumstats.sel.infobin.code),]
p$sumstats.sel.sim<-p$sumstats[which(p$sumstats$code %in% p$sumstats.sel.sim.code),]
#p$sumstats.sel.infobin<-rbind(p$sumstats["DEPR05",],p$sumstats["DEPR05",])

#p$sumstats.sel$code_orig<-p$sumstats.sel$code
#p$sumstats.sel$code<-p$sumstats.sel$code.trait
#p$sumstats.sel[,c("code","name","name.nice","name.nice.and_code","reference.nice", "year","n_case","n_control","n_total","pmid","reference_doi","samplePrevalence","populationPrevalence","dependent_variable.OLS","dependent_variable.linprob","se.logit","mungedpath")]
p$k.sel<-nrow(p$sumstats.sel)
#View(p$sumstats.sel[,c("code","n_total","pmid","reference_doi","samplePrevalence","populationPrevalence","mungedpath")])

write.table(p$sumstats.sel[,c("code", "name.nice","year","reference.nice", "n_case","n_control","n_total","samplePrevalence","populationPrevalence", "reference_doi")], file = file.path(p$folderpath.workingDirectory,paste0(p$setup.code,".sumstatinfo.tsv")), quote = TRUE, sep = "\t", row.names = FALSE, col.names = TRUE)

#View(p$sumstats.sel)

```



# Raw data preparatory steps

## Create df with high-LD regions for GRCh37
```{r High-LD regions}
#From 10.1016/j.ajhg.2008.06.005
#via 10.1101/211821
p$highld_b37<-data.frame(matrix(data=NA,ncol=0,nrow=0))

p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(1,48287980,52287979)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(2,86088342,101041482)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(2,134666268,138166268)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(2,183174494,190174494)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(3,47524996,50024996)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(3,83417310,96017310)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(5,44464243,50464243)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(5,97972100,100472101)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(5,128972101,131972101)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(5,135472101,138472101)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(6,25392021,33392022)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(6,56892041,63942041)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(6,139958307,142458307)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(7,55225791,66555850)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(8,7962590,11962591)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(8,42880843,49837447)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(8,111930824,114930824)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(10,36959994,43679994)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(11,46043424,57243424)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(11,87860352,90860352)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(12,33108733,41713733)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(12,111037280,113537280)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(17,31799963,33389579)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(17,40928985,42139672)
p$highld_b37[(nrow(p$highld_b37)+1),c("CHR","BP","BP2")]<-c(20,32536339,35066586)

setDT(p$highld_b37)
setkeyv(p$highld_b37, cols = c("CHR","BP","BP2"))

p$filepath.highld_b38<-file.path(p$folderpath.data,"gene_mapping","highld_b38.gz")
if(file.exists(p$filepath.highld_b38)){
  p$highld_b38 <- fread(file = p$filepath.highld_b38, header = T)
} else {

  #create lift-over build 38- move to data/gene_mapping
  res <- supermunge(
    list_df = list(highld_b38=p$highld_b37),
    chainFilePath = "../data/alignment_chains/hg19ToHg38.over.chain.gz",
    pathDirOutput = p$folderpath.workingDirectory,
    lossless = T,
    process = F
    )
  p$highld_b38<-res
}
setkeyv(p$highld_b38, cols = c("CHR","BP","BP2"))
#View(res$last[order(res$last$CHR,res$last$BP),])

#p$highld_b37<-NULL

```

## Simulate real genomic association effects
This is used as input to GCTA to generate simulated GWAS with known h2
```{r Simulate real genomic association effects}

if(p$clOptions$task=="sim") {

  #Known variants
  variantsToSimulate <- shru::readFile(filePath = p$filepath.SNPReference.1kg, nThreads = p$nThread)
  #variantsToSimulate<-variantsToSimulate[MAF < 0.45 | MAF > 0.55,] #MAF is not >0.5 here though
  
  nrow(variantsToSimulate)
  
  #remove a random set of variants to achieve targeted size
  variantsToSimulate$SELV1<-rbinom(nrow(variantsToSimulate), 1, 0.6)
  variantsToSimulate$SELV2<-rbinom(nrow(variantsToSimulate), 1, 0.6)
  nrow(variantsToSimulate[SELV1==1 & MAF >0.01,])
  nrow(variantsToSimulate[SELV2==1 & MAF >0.01,])
  
  #L2 info
  L2.mean <- mean(variantsToSimulate$L2,na.rm=T)
  L2.median <- median(variantsToSimulate$L2,na.rm=T)
  
  #Simulate completely randomised phenotype, h2=0.2
  if(file.exists(file.path(p$folderpath.workingDirectory,"simulatedVars.Rds"))) {
    print("Using existing effect simulation.")
    simulatedVars<-readRDS(file=file.path(p$folderpath.workingDirectory,"simulatedVars.Rds"))
  } else {

    SigmaCovarianceMatrix <- matrix(c(0.2,0.2*0.4,0.2*0.4,0.2),2,2)
    simulatedVars <- mvrnorm(n = nrow(variantsToSimulate), mu = c(0,0),Sigma = SigmaCovarianceMatrix)
    saveRDS(simulatedVars,file = file.path(p$folderpath.workingDirectory,"simulatedVars.Rds"))
  }
  
  
  #Original real values
  variantsToSimulate$RV1<-simulatedVars[,1]
  variantsToSimulate$RV2<-simulatedVars[,2]
  
  meanAbsRV1<-mean(abs(variantsToSimulate$RV1),na.rm=T)
  meanAbsRV2<-mean(abs(variantsToSimulate$RV2),na.rm=T)
  
  #These are the true genome-wide correlations
  simulatedVars.true.rg <- cor(variantsToSimulate[SELV1==1 & SELV2==1,c("RV1","RV2")])
  print(simulatedVars.true.rg)
#           RV1       RV2
# RV1 1.0000000 0.4003097
# RV2 0.4003097 1.0000000
  simulatedVars.true.covg <- cov(variantsToSimulate[SELV1==1 & SELV2==1,c("RV1","RV2")])
  print(simulatedVars.true.covg)
#            RV1        RV2
# RV1 0.20002236 0.08006982
# RV2 0.08006982 0.20001705
  
  #without MAF effect, in case inactivated
  variantsToSimulate[,EFFECT1:=RV1]
  variantsToSimulate[,EFFECT2:=RV2]
  
  # #MAF
  # variantsToSimulate[,EFFECT1:=RV1/((0.5+MAF)^2)]
  # variantsToSimulate[,EFFECT2:=RV2/((0.5+MAF)^2)]
  # meanAbsEFFECT1<-mean(abs(variantsToSimulate$EFFECT1),na.rm=T)
  # meanAbsEFFECT2<-mean(abs(variantsToSimulate$EFFECT2),na.rm=T)
  # EFFECT1.ratio<-meanAbsRV1/meanAbsEFFECT1
  # EFFECT2.ratio<-meanAbsRV2/meanAbsEFFECT2
  # variantsToSimulate[,EFFECT1:=EFFECT1*eval(EFFECT1.ratio)]
  # variantsToSimulate[,EFFECT2:=EFFECT2*eval(EFFECT2.ratio)]
  # 
  # #These should also be the true genome-wide correlations
  # simulatedVars.true.rg <- cor(variantsToSimulate[SELV1==1 & SELV2==1,c("EFFECT1","EFFECT2")])
  # print(simulatedVars.true.rg)
  # simulatedVars.true.covg <- cov(variantsToSimulate[SELV1==1 & SELV2==1,c("EFFECT1","EFFECT2")])
  # print(simulatedVars.true.covg)
  
  #L2
  ##QC
  LDSC.M<-7184778 #the LDSC 'M' value (#variants in LD score panel with MAF > 0.05)
  #[,L:=sqrt(L2)]
  variantsToSimulate[L2<0,L2:=0.01][!is.finite(L2),L2:=eval(L2.median)][,L:=sqrt(L2)][,rL:=sqrt(L2)/sqrt(eval(LDSC.M))]
  #variantsToSimulate[,EFFECT1.LD:=sqrt((EFFECT1^2)*L2/eval(LDSC.M))][,EFFECT2.LD:=sqrt((EFFECT2^2)*L2/eval(LDSC.M))] #works fairly well
  variantsToSimulate[,EFFECT1.LD:=EFFECT1*rL][,EFFECT2.LD:=EFFECT2*rL] #linear version - CURRENT version (ld)
  meanAbsEFFECT1.LD<-mean(abs(variantsToSimulate$EFFECT1.LD),na.rm=T)
  meanAbsEFFECT2.LD<-mean(abs(variantsToSimulate$EFFECT2.LD),na.rm=T)
  EFFECT1.LD.ratio<-meanAbsRV1/meanAbsEFFECT1.LD
  EFFECT2.LD.ratio<-meanAbsRV2/meanAbsEFFECT2.LD
  variantsToSimulate[,EFFECT1.LD:=EFFECT1.LD*eval(EFFECT1.LD.ratio)]
  variantsToSimulate[,EFFECT2.LD:=EFFECT2.LD*eval(EFFECT2.LD.ratio)]
  
  #These are the almost true genome-wide correlations after simulating L2
  simulatedVars.true.l2.rg <- cor(variantsToSimulate[SELV1==1 & SELV2==1,c("EFFECT1.LD","EFFECT2.LD")],use = "pairwise.complete.obs")
  print(simulatedVars.true.l2.rg)
#  EFFECT1.LD EFFECT2.LD
# EFFECT1.LD   1.000000   0.399341
# EFFECT2.LD   0.399341   1.000000
  simulatedVars.true.l2.covg <- cov(variantsToSimulate[SELV1==1 & SELV2==1,c("EFFECT1.LD","EFFECT2.LD")], use = "pairwise.complete.obs")
  print(simulatedVars.true.l2.covg)
#   EFFECT1.LD EFFECT2.LD
# EFFECT1.LD  0.3851489  0.1537597
# EFFECT2.LD  0.1537597  0.3849187
  
  
  fwrite(x = variantsToSimulate[SELV1==1 & MAF>0.01,.(SNP,EFFECT1.LD)], file = file.path(p$folderpath.workingDirectory,"simEffects1.ld"), col.names = FALSE,nThread = p$nThread, sep = "\t")
  fwrite(x = variantsToSimulate[SELV2==1 & MAF>0.01,.(SNP,EFFECT2.LD)], file = file.path(p$folderpath.workingDirectory,"simEffects2.ld"), col.names = FALSE,nThread = p$nThread, sep = "\t")
  
#   chrs<-1:23
#   for(iChr in chrs){
#     fwrite(x = variantsToSimulate[CHR==eval(iChr),.(SNP,EFFECT1)], file = file.path(p$folderpath.workingDirectory,paste0("simEffects1.",iChr)), col.names = FALSE,nThread = p$nThread, sep = "\t")
#     fwrite(x = variantsToSimulate[CHR==eval(iChr),.(SNP,EFFECT2)], file = file.path(p$folderpath.workingDirectory,paste0("simEffects2.",iChr)), col.names = FALSE,nThread = p$nThread, sep = "\t")
#   }
# 
#   cf <- data.frame(command=paste0(
#   "\n#Chr=",chrs,"
#   sbatch --time 12:00:00 --partition cpu --job-name=\"sim\" --ntasks 1 --cpus-per-task 6 --mem 256G --wrap=\"gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects1.",chrs," --simu-qt --simu-hsq 0.20 --simu-rep 1 --out simTrait1.",chrs,".020\" --output \"setup8.sim1.",chrs,".020.$(date +%Y%m%d).out.txt\"
# 
#   sbatch --time 12:00:00 --partition cpu --job-name=\"sim\" --ntasks 1 --cpus-per-task 6 --mem 256G --wrap=\"gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects2.",chrs," --simu-qt --simu-hsq 0.20 --simu-rep 1 --out simTrait2.",chrs,".020\" --output \"setup8.sim2.",chrs,".020.$(date +%Y%m%d).out.txt\"
# 
#   "
# )
# )

#fwrite(cf,file = file.path(p$folderpath.workingDirectory,"SIMcommands.sh"), append = F, quote = F, sep = "\t", col.names = F, nThread=p$nThread, na = "NA")


}

if(p$clOptions$task=="sim") quit(save = "no")

```


##Create keep file for the simulate GWASs
```{r keep file}
if(p$clOptions$task=="sim") {
  keep.ids <- fread(file = "data/geno/GLADv3_EDGIv1_NBRv2/genotyped/GLAD_EDGI_NBR_v3_EUR_20230512_maf0.01_sample95.SNP95.fam", na.strings =c(".",NA,"NA",""), encoding = "UTF-8",check.names = T, fill = T, blank.lines.skip = T, data.table = T, nThread = p$nThread, showProgress = F)
  
  keep.ids<-keep.ids[!is.na(V1) & !is.na(V2),]
  #keep.ids<-as.data.frame(keep.ids[!is.na(V1) & !is.na(V2), .(V1,V2)])
  nrow(keep.ids)

  fwrite(keep.ids[,.(V1,V2)],file = file.path(p$folderpath.workingDirectory,"keep.all.txt"), append = F, quote = F, sep = "\t", col.names = F, nThread=p$nThread, na = "NA")
  rm(keep.ids)
}

if(p$clOptions$task=="sim") quit(save = "no")

```

##Create covar files for the simulated GWASs
```{r covar files}
if(p$clOptions$task=="sim") {
covar.ids <- fread(file = file.path("data","geno","GLADv3_EDGIv1_NBRv2","genotyped", "GLADv3_EDGIv1_NBRv2_EUR_covariates_1kg_projecting_pcs_batch_array.txt"), na.strings =c(".",NA,"NA",""), encoding = "UTF-8",check.names = T, fill = T, blank.lines.skip = T, data.table = T, showProgress = F)

fwrite(covar.ids,file = file.path(p$folderpath.workingDirectory,"covar.txt"), append = F, quote = F, sep = "\t", col.names = T, na = "NA") #regenie needs NA as 'NA'

#.(FID,IID,pc1,pc2,pc3,pc4,pc5,pc6,pc7,pc8,pc9,pc10)]
fwrite(covar.ids[, .(FID,IID,pc1,pc2,pc3,pc4,pc5,pc6,pc7)],
       file = file.path(p$folderpath.workingDirectory,"covar.q.txt"), append = F, quote = F, sep = "\t", col.names = T, na = "NA") #regenie needs NA as 'NA'

fwrite(covar.ids[, .(FID,IID,Array,Source,batch)],
       file = file.path(p$folderpath.workingDirectory,"covar.c.txt"), append = F, quote = F, sep = "\t", col.names = T, na = "NA") #regenie needs NA as 'NA'

rm(covar.ids)
}

if(p$clOptions$task=="sim") quit(save = "no")

```

##Create joint pheno-file for GCTA bivariate GREML
```{r Merge the two separate simulated trait phenofiles, eval=FALSE, purl=FALSE}

pheno1 <- shru::readFile(filePath = file.path(p$folderpath.workingDirectory,"simTrait1.020.ld.phen"),nThreads = p$nThread)

pheno2 <- shru::readFile(filePath = file.path(p$folderpath.workingDirectory,"simTrait2.020.ld.phen"),nThreads = p$nThread)

pheno1[pheno2,on=c("V1"),V4:=i.V3]

fwrite(x = pheno1,file = file.path(p$folderpath.workingDirectory,"simTrait1and2.020.ld.phen"),col.names = F, sep = "\t")

rm(pheno1)
rm(pheno2)

```




## Simulate GWAS associations (noisy environmental associations etc)
NOT USED
This is used to generate simulated GWAS output
```{r Simulate GWAS associations, eval=FALSE, purl=FALSE}

if(p$clOptions$task=="sim2") {
  chrs<-1:23
  
  variantsToSimulate <- shru::readFile(filePath = p$filepath.SNPReference.1kg,nThreads = p$nThread)
  #variantsToSimulate<-variantsToSimulate[MAF < 0.45 | MAF > 0.55,] #MAF is not >0.5 here though
  
  #L2 info
  L2.mean <- mean(variantsToSimulate$L2,na.rm=T)
  L2.median <- median(variantsToSimulate$L2,na.rm=T)
  
  nrow(variantsToSimulate)
  
  #real world data GWAS reference
  realWorldGWAS <- shru::readFile(filePath = file.path(p$folderpath.data.sumstats.munged.supermunge.1kg.orig.unfiltered,"ANXI03.gz"),nThreads = p$nThread)
  
  #Simulate N
  randomN.mean <- mean(realWorldGWAS$N,na.rm=T)
  randomN.max <- max(realWorldGWAS$N,na.rm=T)
  randomN.sd <- randomN.mean*0.2 #sd(realWorldGWAS$N,na.rm=T) #this is 0 for ANXI03 - setting to arbitrary number
  randomNSigmaCovarianceMatrix <- matrix(c(randomN.sd^2,0.8*randomN.sd^2,0.8*randomN.sd^2,randomN.sd^2),2,2)
  simulatedNDelta <- mvrnorm(n = nrow(variantsToSimulate), mu = c(0,0),Sigma = randomNSigmaCovarianceMatrix)
  
  
  #Simulate unknown INFO values
  randomINFO.sd<-sd(realWorldGWAS$INFO,na.rm=T)
  randomINFO.mean<-mean(realWorldGWAS$INFO,na.rm=T)
  randomINFOSigmaCovarianceMatrix <- matrix(c(randomINFO.sd^2,0.8*randomINFO.sd^2,0.8*randomINFO.sd^2,randomINFO.sd^2),2,2)
  simulatedINFODelta <- mvrnorm(n = nrow(variantsToSimulate), mu = c(0,0),Sigma = randomINFOSigmaCovarianceMatrix)
  
  #Read real effects 1 & 2
  #realEffects1 <- shru::readFile(filePath = file.path(p$folderpath.data,"gwas_sumstats","raw","simTrait1.020.par"),nThreads = p$nThread)
  #basic info
  realEffects1<-as.data.table(matrix(NA,0,0))
  for(iChr in chrs){
    realEffects1.c <- shru::readFile(filePath = file.path(p$folderpath.data,"gwas_sumstats","raw",paste0("simTrait1.",iChr,".020.par")),nThreads = p$nThread)
    realEffects1<-rbind(realEffects1,realEffects1.c)
  }
  realEffects1[variantsToSimulate, on=c(QTL="SNP"), c("CHR","BP","A1","A2"):=.(i.CHR,i.BP,i.A1,i.A2)]
  #check #realEffects1[RefAllele!=A1,]
  realEffects2<-as.data.table(matrix(NA,0,0))
  for(iChr in chrs){
    realEffects2.c <- shru::readFile(filePath = file.path(p$folderpath.data,"gwas_sumstats","raw",paste0("simTrait2.",iChr,".020.par")),nThreads = p$nThread)
    realEffects2<-rbind(realEffects2,realEffects2.c)
  }
  realEffects2[variantsToSimulate, on=c(QTL="SNP"), c("CHR","BP","A1","A2"):=.(i.CHR,i.BP,i.A1,i.A2)]
  
  
  #N, VSNP, SE
  realEffects1$N <- randomN.max - abs(simulatedNDelta[1:nrow(realEffects1),1])
  realEffects1[N<0 | !is.finite(N),N:=0.001*eval(randomN.mean)]
  realEffects1[,VSNP:=2*Frequency*(1-Frequency)][,SE:=sqrt(1/(VSNP*N))][,VAR:=SE^2]
  realEffects2$N <- randomN.max - abs(simulatedNDelta[1:nrow(realEffects2),2])
  realEffects2[N<0 | !is.finite(N),N:=0.001*eval(randomN.mean)]
  realEffects2[,VSNP:=2*Frequency*(1-Frequency)][,SE:=sqrt(1/(VSNP*N))][,VAR:=SE^2]
  
  
  realEffects.VAR.mean<-(mean(realEffects1[is.finite(VAR),]$VAR,na.rm=T)+mean(realEffects2[is.finite(VAR),]$VAR,na.rm=T))/2
  realEffects.ABSBETA.mean<-(mean(abs(realEffects1[is.finite(Effect),]$Effect),na.rm=T)+mean(abs(realEffects1[is.finite(Effect),]$Effect),na.rm=T))/2
  
  
  
  
  #Simulate environmental correlated noise/bias - what is an optimal correlation setting for this noise? Now set to equal the genetic-phenotypic associations previously.
  realEnvironmentalProportionOfVariance=0.8 #compare with the heritability setting of 0.2
  realWorldGWAS[,VAR:=SE^2]
  realWorldGWAS.VAR.mean<-mean(realWorldGWAS[is.finite(VAR),]$VAR,na.rm=T)
  realWorldGWAS.ABSBETA.mean<-mean(abs(realWorldGWAS[is.finite(BETA),]$BETA),na.rm=T)
  randomEnvironmentSigmaCovarianceMatrix <- matrix(c(realEnvironmentalProportionOfVariance*1,realEnvironmentalProportionOfVariance*0.4,realEnvironmentalProportionOfVariance*0.4,realEnvironmentalProportionOfVariance*1),2,2)
  simulatedEnvironmentDelta <- mvrnorm(n = nrow(variantsToSimulate), mu = c(0,0),Sigma = randomEnvironmentSigmaCovarianceMatrix)
  
  #L2
  realEffects1[variantsToSimulate, on=c(QTL="SNP"), L2:=i.L2] 
  realEffects1[L2<0, L2:=0.01][!is.finite(L2),L2:=eval(L2.median)]
  
  #INFO
  realEffects1[realWorldGWAS, on=c(QTL="SNP"), REALINFO:=i.INFO] #Real world INFO, i.e. which variants are genotyped INFO==1 rather than imputed - note that this notation may be different across GWAS summary statistics
  #realEffects1[REALINFO==1,INFO:=REALINFO] #add only genotyped INFO=1 from real world data
  #add more INFO values
  realEffects1[,INFO:=REALINFO] #if the dataset is representative we can use all of the available INFO values
  #realEffects1$RINFO<-pmin(randomINFO.mean+simulatedINFOdelta[,1],1,na.rm = T) #alt version
  realEffects1$RINFO <- (1 - abs(simulatedINFODelta[1:nrow(realEffects1),1]))
  #realEffects1[,RINFOA:=RINFO*abs(L2/(eval(L2.median)+L2))]
  realEffects1[!is.finite(INFO),INFO:=RINFO*abs(L2/(eval(L2.median)+L2))] #adjustment for L2
  
 
  
  #environmental and other noise and bias, using correlated errors
  realEffects1$RNOISEBETA<-simulatedEnvironmentDelta[1:nrow(realEffects1),1] #noise from environment
  realEffects1$INFONOISEBETA<-rnorm(n = nrow(realEffects1),mean = 0, sd = (1-realEffects1$INFO))
  realEffects1[,BETA.RNE:=Effect+RNOISEBETA][,BETA.RNEI:=Effect+RNOISEBETA+INFONOISEBETA]
  
  realEffects1$RNOISESE<-sqrt(realEnvironmentalProportionOfVariance/realEffects1$N)
  realEffects1$INFONOISESE<-sqrt(((1-realEffects1$INFO)^2)/realEffects1$N)
  realEffects1[,SE.RNE:=sqrt(SE^2+RNOISESE^2)][,SE.RNEI:=sqrt(SE^2+RNOISESE^2+INFONOISESE^2)]
  
  shru::writeFile(realEffects1[is.finite(BP) & is.finite(Effect) & Frequency > 0.009 ,.(SNP=QTL,CHR,BP,A1,A2,FRQ=Frequency,BETA=Effect,SE,N)],filePath = file.path(p$folderpath.data.sumstats.cleaned,"SMRV01.gz"))
  shru::writeFile(realEffects1[is.finite(BP) & is.finite(BETA.RNE) & Frequency > 0.009,.(SNP=QTL,CHR,BP,A1,A2,FRQ=Frequency,BETA=BETA.RNE,SE=SE.RNE,N)],filePath = file.path(p$folderpath.data.sumstats.cleaned,"SMNE01.gz"))
  shru::writeFile(realEffects1[is.finite(BP) & is.finite(BETA.RNEI) & Frequency > 0.009,.(SNP=QTL,CHR,BP,A1,A2,FRQ=Frequency,BETA=BETA.RNEI,SE=SE.RNEI,INFO,N)],filePath = file.path(p$folderpath.data.sumstats.cleaned,"SMNI01.gz"))
  
  rm(realEffects1)
  
  supermunge(filePaths = c(
    file.path(p$folderpath.data.sumstats.cleaned,"SMRV01.gz"),
    file.path(p$folderpath.data.sumstats.cleaned,"SMNE01.gz"),
    file.path(p$folderpath.data.sumstats.cleaned,"SMNI01.gz")
  ),
  refFilePath = p$filepath.SNPReference.1kg,
  traitNames = c("SMRV01",
                 "SMNE01",
                 "SMNI01"),nThreads = p$nThread,pathDirOutput = p$folderpath.data.sumstats.munged.supermunge.1kg.orig.unfiltered)
  
  
  
  
  #Real effects 2
  realEffects2 <- shru::readFile(filePath = file.path(p$folderpath.data,"gwas_sumstats","raw","simTrait1.020.par"),nThreads = p$nThread)
  #basic info
  realEffects2[variantsToSimulate, on=c(QTL="SNP"), c("CHR","BP","A1","A2"):=.(i.CHR,i.BP,i.A1,i.A2)]
  #check #realEffects1[RefAllele!=A1,]
  
  #L2
  realEffects2[variantsToSimulate, on=c(QTL="SNP"), L2:=i.L2] 
  realEffects2[L2<0, L2:=0.01][!is.finite(L2),L2:=eval(L2.median)]
  
  #INFO
  realEffects2[realWorldGWAS, on=c(QTL="SNP"), REALINFO:=i.INFO] #Real world INFO, i.e. which variants are genotyped INFO==1 rather than imputed - note that this notation may be different across GWAS summary statistics
  #realEffects1[REALINFO==1,INFO:=REALINFO] #add only genotyped INFO=1 from real world data
  #add more INFO values
  realEffects2[,INFO:=REALINFO] #if the dataset is representative we can use all of the available INFO values
  #realEffects1$RINFO<-pmin(randomINFO.mean+simulatedINFOdelta[,1],1,na.rm = T) #alt version
  realEffects2$RINFO <- 1 - abs(simulatedINFODelta[,2])
  #realEffects1[,RINFOA:=RINFO*abs(L2/(eval(L2.median)+L2))]
  realEffects2[!is.finite(INFO),INFO:=RINFO*abs(L2/(eval(L2.median)+L2))] #adjustment for L2
  
  #N, VSNP, SE
  realEffects2$N <- randomN.max - abs(simulatedNDelta[,2])
  realEffects2[N<0 | !is.finite(N),N:=0.001*eval(randomN.mean)]
  realEffects2[,VSNP:=2*Frequency*(1-Frequency)][,SE:=sqrt(1/(VSNP*N))]
  
  #environmental and other noise and bias, using correlated errors
 
  
  realEffects2$RNOISEBETA<-simulatedEnvironmentDelta[,2] #noise from environment
  realEffects2$INFONOISEBETA<-rnorm(n = nrow(realEffects2),mean = 0, sd = (1-realEffects2$INFO))
  realEffects2[,BETA.RNE:=Effect+RNOISEBETA][,BETA.RNEI:=Effect+RNOISEBETA+INFONOISEBETA]
  
  realEffects2$RNOISESE<-sqrt(realEnvironmentalProportionOfVariance/realEffects2$N)
  realEffects2$INFONOISESE<-sqrt(((1-realEffects2$INFO)^2)/realEffects2$N)
  realEffects2[,SE.RNE:=sqrt(SE^2+RNOISESE^2)][,SE.RNEI:=sqrt(SE^2+RNOISESE^2+INFONOISESE^2)]
  
  shru::writeFile(realEffects2[is.finite(BP) & is.finite(Effect) & Frequency > 0.009 ,.(SNP=QTL,CHR,BP,A1,A2,FRQ=Frequency,BETA=Effect,SE,N)],filePath = file.path(p$folderpath.data.sumstats.cleaned,"SIMR02.gz"))
  shru::writeFile(realEffects2[is.finite(BP) & is.finite(BETA.RNE) & Frequency > 0.009,.(SNP=QTL,CHR,BP,A1,A2,FRQ=Frequency,BETA=BETA.RNE,SE,INFO,N)],filePath = file.path(p$folderpath.data.sumstats.cleaned,"SIMRN02.gz"))
  shru::writeFile(realEffects2[is.finite(BP) & is.finite(BETA.RNEI) & Frequency > 0.009,.(SNP=QTL,CHR,BP,A1,A2,FRQ=Frequency,BETA=BETA.RNEI,SE,INFO,N)],filePath = file.path(p$folderpath.data.sumstats.cleaned,"SIMRNI02.gz"))
  
  rm(realEffects2)
  
}

if(p$clOptions$task=="sim2") quit(save = "no")

```

## OLD Simulate GWAS (incativated)
```{r Simulate GWAS OLD, eval=FALSE, purl=FALSE}
# if(p$clOptions$task=="hap") { 
#   
# 
#   
# }


if(p$clOptions$task=="sim") {
#devtools::install_github("chr1swallace/simGWAS")
#install.packages('hapsim')

#simulate haplotypes test  -from https://chr1swallace.github.io/simGWAS/articles/intro.html
set.seed(42)
nsnps <- 100
nhaps <- 1000
lag <- 5 # genotypes are correlated between neighbouring variants
maf <- runif(nsnps+lag,0.05,0.5) # common SNPs
laghaps <- do.call("cbind", lapply(maf, function(f) rbinom(nhaps,1,f)))
haps <- laghaps[,1:nsnps]
for(j in 1:lag) 
    haps <- haps + laghaps[,(1:nsnps)+j]
haps <- round(haps/matrix(apply(haps,2,max),nhaps,nsnps,byrow=TRUE))

snps <- colnames(haps) <- paste0("s",1:nsnps)
freq <- as.data.frame(haps+1)
freq$Probability <- 1/nrow(freq)
sum(freq$Probability)

#specify causal vriants
##Next, we need to specify the causal variants, and their effects on disease, as odds ratios. We create a vector CV with snp names that are a subset of column names in freq and a vector of odds ratios. In our simulated data, we pick two causal variants at random, with odds ratios of 1.4 and 1.2.
nExposed <- colSums(haps) #d1+h1
nNonexposed <- nhaps-nExposed #d0+h0

CV=sample(snps[which(colMeans(haps)>0.1)],2)
g1 <- c(1.004,1.002)

#variance tests
# #variance1: Two random variables (Bernoulli) E and D - PROBABLY WRONG!
# pE<-nExposed[CV]
# pD<-0.5 #this is set to 0.5 for the comparison
# varOR1<- (pE*(1-pE)+pE^2) * (pD*(1-pD)+pD^2) - (pE^2)*(pD^2)


#variance2: variance of log(OR): X ∼ B(n, p) and Y ∼ B(m, q), independent; 0 < p, q < 1, while n and m are large: 1/np + 1/(n(1-p)) + 1/mq + 1/(m(1-q))
#https://www.stat.berkeley.edu/~freedman/oddsrat.pdf
#p1<-0.7 #unknown
p0<-0.1

# OR<-(p1*(1-p1))/(p0*(1-p0))
# OR*(p0*(1-p0))=p1-p1^2
# p1^2-p1+OR*(p0*(1-p0))=0

#roots 1/2 +-sqrt( (-1/2)^2 - OR*(p0*(1-p0)))
p1 <- 1/2 + sqrt( (-1/2)^2 - g1*(p0*(1-p0)))
#p1 <- 1/2 - sqrt( (-1/2)^2 - g1*(p0*(1-p0)))
varLogOR <- 1/(nExposed[CV]*p1) + 1/(nExposed[CV]*(1-p1)) + 1/(nNonexposed[CV]*p0) + 1/(nNonexposed[CV]*(1-p0))



#old
#varLogOR <- ((1/g1)+1)/nExposed[CV] + (g1+1)/nExposed[CV] + 4/nNonexposed[CV]

#heritability is
sum(varLogOR) #variance formulas for sum of random variables

#using d0/h0=1:
#OR=d1/h1, n1=d1+h1 => d1 = n1/(1/OR+1)
#OR=d1/h1, n1=d1+h1 => h1 = n1/(OR+1)
#OR=d1/h1, n0=d0+h0, d0/h0=1 => d0=h0=n0/2
#var(log(OR))=1/d1+1/h1+1/d0+1/h0 = ((1/OR)+1)/n1 + (OR+1)/n1 + 4/n0


##Now we simulate the results of a GWAS. There are two key functions, makeGenoProbList calculates genotype probabilities at each SNP conditional on genotypes at the causal variants, then est_statistic generates the vector of Z scores across all SNPs, conditional on the causal variants and their odds ratios.
FP <- simGWAS::make_GenoProbList(snps=snps,W=CV,freq=freq)
zexp <- simGWAS::expected_z_score(N0=10000, # number of controls
              N1=10000, # number of cases
              snps=snps, # column names in freq of SNPs for which Z scores should be generated
              W=CV, # causal variants, subset of snps
              gamma.W=log(g1), # odds ratios
              freq=freq, # reference haplotypes
              GenoProbList=FP) # FP above


#new routine - simulate random variant effects from a specified trait heritability measure

#p$nThread<-6
#test - generate haplotypes with hapsim
filepath.varlist <- normalizePath(file.path(p$folderpath.data.variantLists,"hc1kgp3.b38.eur.l2.jz2023.gz")) #use the new version here rather than the old
varlist<-readFile(filepath.varlist)
gwasColumns<-stdGwasColumnNames(colnames(varlist), ancestrySetting = "EUR")
gwasColumns$orig
gwasColumns$std
colnames(varlist)<-gwasColumns$std
varlist<-varlist[FRQ>0,]
randomTraitWeightsTot<-sqrt(rchisq(n=nrow(varlist), ncp = 1, df = 1))
names(randomTraitWeightsTot)<-varlist$SNP

heritabilityPopulationM <- nrow(varlist) #nrow(varlist[FRQ>0.05,]) #improve this value a bit, what is the reference value to use here? nr of non-rare variants in a reference population?

th<-0.05 #target heritability
ratioCausal <- 0.2
initialMeanOR<-1.000001
targetMeanHeritabilityPerCausalVariant <- th/(heritabilityPopulationM*ratioCausal)
targetMeanHeritabilityPerVariant <- th/(heritabilityPopulationM)
nSimCas<-500000
nSimCon<-500000
nSimTot<-nSimCas+nSimCon


#genome-wide OR-table
ORtable<-data.frame(SNP=varlist$SNP,initialMeanOR=initialMeanOR,w=randomTraitWeightsTot)
setDT(ORtable)
setkeyv(ORtable, cols = c("SNP"))
ORtable[sample(.N, ratioCausal*nrow(ORtable)),causal:=T]

filepath.individuals <- file.path(p$folderpath.data,"data","reference_panel","hc1kgp3.b38.plink","hc1kgp3.b38.eur.jz2023.vcf.gz")
filepath.test.individuals <- file.path(p$folderpath.workingDirectory,"test.individuals.Rds")  #test
individuals.all <- fread(file = filepath.individuals, na.strings =c(".",NA,"NA",""), encoding = "UTF-8", header = T, fill = T, blank.lines.skip = T, data.table = T, nThread = p$nThread, showProgress = T, skip =  31, nrows = 10000) #test
individuals.all <- readRDS(file=filepath.test.individuals) #test
rownames(individuals.all)<-individuals.all$ID #does this do anything when DT?
setkeyv(individuals.all, cols = c("ID"))

n.individuals<-ncol(individuals.all)-9
m.blockVariants<-nrow(individuals.all)

#freq <- as.data.frame(matrix(data=NA,nrow = n.individuals, ncol = m.blockVariants))
freq<-t(individuals.all[,10:ncol(individuals.all)][,ifelse(.SD=="0/0",0,ifelse(.SD=="1/1",2,1))])
colnames(freq)<-individuals.all$ID
#rownames(freq)<-colnames(individuals.all)[10:ncol(individuals.all)]
#setDT(freq)

#saveRDS(object = individuals.all, file = filepath.test.individuals)


randomTraitWeights<-randomTraitWeightsTot[colnames(freq)]
nExposed<-colSums(freq) #this is an approximation, counting the homozygous alt allele twice
nNonexposed<-n.individuals-nExposed
exposedRatio<-nExposed/n.individuals
nonexposedRatio<-nNonexposed/n.individuals
#ORtable.chunk<- ORtable[SNP %in% individuals.all$ID,]
freq.sum<-data.frame(SNP=individuals.all$ID,nExposed=colSums(freq),nNonexposed=n.individuals-nExposed,exposedRatio=nExposed/n.individuals,
nonexposedRatio=nNonexposed/n.individuals)
setDT(freq.sum)
setkeyv(freq.sum, cols = c("SNP"))


ORtable.chunk <- merge(x = ORtable, y = freq.sum, by = "SNP", all = F)
ORtable.chunk[,OR:=NA_real_]

p1<-NA #unknown
p0<-0.1 #fix setting
nSimExposedMean<-nSimTot*mean(ORtable.chunk$exposedRatio)
nSimNonexposedMean<-nSimTot-nSimExposedMean

# tVariance <- 1/(nSimExposedMean*p1) + 1/(nSimExposedMean*(1-p1)) + 1/(nSimNonexposedMean*p0) + 1/(nSimNonexposedMean*(1-p0))
# tVariance - 1/(nSimNonexposedMean*p0) - 1/(nSimNonexposedMean*(1-p0))<- 1/(nSimExposedMean*p1) + 1/(nSimExposedMean*(1-p1)) 
# (nSimExposedMean*p1)*(nSimExposedMean*(1-p1)) 
# = 2/(tVariance - 1/(nSimNonexposedMean*p0) - 1/(nSimNonexposedMean*(1-p0)))
# nSimExposedMean^2+p1*nSimExposedMean+nSimExposedMean-nSimExposedMean*p1+p1-p1^2
# = 2/(tVariance - 1/(nSimNonexposedMean*p0) - 1/(nSimNonexposedMean*(1-p0)))
# p1^2-p1-nSimExposedMean-nSimExposedMean^2+2/(tVariance - 1/(nSimNonexposedMean*p0) - 1/(nSimNonexposedMean*(1-p0)))


#p1<- 1/2 + sqrt( (-1/2)^2 - (nSimExposedMean-nSimExposedMean^2+2/(targetMeanHeritabilityPerCausalVariant - 1/(nSimNonexposedMean*p0) - 1/(nSimNonexposedMean*(1-p0)) ) ) )
#this does not work


#roots 1/2 +-sqrt( (-1/2)^2 - OR*(p0*(1-p0)))
p1 <- 1/2 + sqrt( (-1/2)^2 - initialMeanOR*(p0*(1-p0)))
#p1 <- 1/2 - sqrt( (-1/2)^2 - g1*(p0*(1-p0)))

varLogOR <- 1/(nSimExposedMean*p1) + 1/(nSimExposedMean*(1-p1)) + 1/(nSimNonexposedMean*p0) + 1/(nSimNonexposedMean*(1-p0))
# nonexposedRatio<-(1-exposedRatio)
# tVariance<-((1/initialMeanOR)+1)/(exposedRatio*nSimTot) + (initialMeanOR+1)/(exposedRatio*nSimTot) + 4/(nonexposedRatio*nSimTot)
#tVariance <- 1/(nSimTot*pE) + 1/(nSimTot*(1-pE)) + 1/(nSimTot*pD) + 1/(nSimTot*(1-pD))

ORtable.chunk[,OR:=initialMeanOR*w]


#hypothetical variance in the target simulation using the same proportions of exposed and non-exposed individuals as in the haplotype sample
#cVariance<-((1/initialMeanOR)+1)/(exposedRatio*nSimTot) + (initialMeanOR+1)/(exposedRatio*nSimTot) + 4/(nonexposedRatio*nSimTot)

ORtable.chunk[,variance:=((1/OR)+1)/(exposedRatio*eval(nSimTot)) + (OR+1)/(exposedRatio*eval(nSimTot)) + 4/(nonexposedRatio*eval(nSimTot))]

sum(cVariance[is.finite(cVariance)])
mean(cVariance[is.finite(cVariance)])


names(varLogOR)<-snps
#check obtained heritability mean
mean(varLogOR)*heritabilityPopulationM

#roots  #does not work
# a <- varLogOR[CV]
# b <-  (4*(nExposed[CV]-1)/nNonexposed[CV])-varLogOR[CV]*nExposed[CV]
# c <- nExposed[CV]
# 
# d1r1 <- abs(((-b)+sqrt(b^2-4*a*c))/(2*a))
# d1r2 <- abs(((-b)-sqrt(b^2-4*a*c))/(2*a))
# 
# g1 <- abs(d1r1/(nExposed[CV]-d1r1))
#test
#((1/g1)+1)/nExposed[CV] + (g1+1)/nExposed[CV] + 4/nNonexposed[CV] 


#run manually

# Create SNP-list for simulated GWAS - not used as the GCTA simulation does not seem to work using the list

# p$simTemplate <- shru::readFile(filePath = file.path(p$folderpath.data.sumstats.munged,"NEUR02.gz"),nThreads = p$nThread) #based on NEUR02
# 
# fwrite(x = p$simTemplate[P<1e-2,c("SNP")], file = file.path(p$folderpath.workingDirectory,"neuro02.sim.varlist.gz"), sep = "\t",append = F, na = ".",nThread = p$nThread, col.names = F)
  
}

if(p$clOptions$task=="hap" | p$clOptions$task=="sim") quit(save = "no")

```


## GWAS summary statistics munge - if needed
Munging of GWAS sumstat datasets. The summary statistics is also prepared for latent factor GWAS (previously, the sumstats()-function of the GSEM package). Here reference alleles are standardised across all datasets, and SNP effects and their s.e. are scaled to match unit-variance phenotypes.
```{r GWAS sumstat munge}
#test
#p$clOptions$task<-"munge"

p$filepath.lfgwas.sumstats<-file.path(p$folderpath.workingDirectory,paste0("lfGWAS.sumstats.",p$setup.code,".Rds"))
p$filepath.lfgwas.sumstats_full<-file.path(p$folderpath.workingDirectory,paste0("lfGWAS.sumstats_full.",p$setup.code,".Rds"))
p$filepath.munge.meta<-file.path(p$folderpath.workingDirectory,paste0("munge.meta.",p$setup.code,".Rds"))
p$munge<-c()

if(p$clOptions$task=="munge" & (!all(file.exists(p$sumstats.sel$mungedpath)) | !is.na(p$clOptions$task_argument))){
  
  #test
  #p$clOptions$task_argument<-"BIPO02"
  if(!is.na(p$clOptions$task_argument)){
    #task argument sets the specific dataset to munge
    
    cat("\nSet to munge with arg",p$clOptions$task_argument,"\n")
    p$munge$filesToUse<-p$sumstats.sel[which(p$sumstats.sel$code==p$clOptions$task_argument),]$cleanedpath
    p$munge$traitNamesToUse<-p$sumstats.sel[which(p$sumstats.sel$code==p$clOptions$task_argument),]$code
    p$munge$NToUse<-p$sumstats.sel[which(p$sumstats.sel$code==p$clOptions$task_argument),]$n_total
    
  } else {
    #defaults
    p$munge$filesToUse<-p$sumstats.sel$cleanedpath
    p$munge$traitNamesToUse<-p$sumstats.sel$code
    p$munge$NToUse<-p$sumstats.sel$n_total
  }
  
  # cat("\nMunging with:\n")
  # print(p$munge$filesToUse)
  # print(p$filepath.SNPReference.hc1kg)
  # print(p$folderpath.data.mvLDSC.ld.1kg)
  # print(p$munge$traitNamesToUse)
  # print(p$munge$NToUse)
  # print(p$munge$OLSToUse)
  # print(p$munge$se.logitToUse)
  # print(p$munge$propToUse)
  # print(p$folderpath.data.sumstats.munged)
  
  #mask<-c(F,F,F,F,F,F,F,F,F,F,F,F,F,F,F,F,F,F,F,T,T,T,T,T)
  #munging with no filters applied - only munge without imputation!
  
  p$munge$results <- supermunge(
            filePaths = p$munge$filesToUse,
            rsSynonymsFilePath = p$filepath.rsSynonyms.dbSNP151,
            refFilePath = p$filepath.SNPReference.hc1kg,
            traitNames = p$munge$traitNamesToUse,
            chainFilePath = file.path(p$folderpath.data,"alignment_chains","hg19ToHg38.over.chain.gz"),
            N = p$munge$NToUse,
            pathDirOutput = p$folderpath.data.sumstats.munged,
            ancestrySetting = "EUR"
              )
    
  
  p$munge$results$meta %>% gt()
  gtsave(data = (p$munge$results$meta %>% gt()), filename = file.path(p$folderpath.workingDirectory,"munge.meta.rtf"))
  saveRDS(object = p$munge$results$meta, file = p$filepath.munge.meta)
  
  
  
} else if(file.exists(p$filepath.munge.meta)){
  p$munge$results$meta<-readRDS(file=p$filepath.munge.meta)
  print("Read munging results metadata from previous munge result file.")
}

print(p$munge$results$meta)


if(p$clOptions$task=="impute"
   #& !all(file.exists(p$sumstats.sel$imputedpath))
   ){
  #test
  #p$clOptions$task_argument<-"ANXI03"
  if(!is.na(p$clOptions$task_argument)){
    #task argument sets the specific dataset to munge
    
    cat("\nSet to IMPUTE with arg",p$clOptions$task_argument,"\n")
    p$munge$filesToUse<-p$sumstats.sel$mungedpath[which(p$sumstats.sel$code==p$clOptions$task_argument)]
    p$munge$traitNamesToUse<-p$sumstats.sel$code[which(p$sumstats.sel$code==p$clOptions$task_argument)]
    #p$munge$NToUse<-p$sumstats.sel$n_total[which(p$sumstats.sel$code==p$clOptions$task_argument)]
    
  } else {
    #defaults
    p$munge$filesToUse<-p$sumstats.sel$mungedpath
    p$munge$traitNamesToUse<-p$sumstats.sel$code
    p$munge$NToUse<-p$sumstats.sel$n_total
  }
  
  
  #Munge with imputation
  p$munge$results <- supermunge(
            filePaths = p$munge$filesToUse,
            refFilePath = p$filepath.SNPReference.hc1kg,
            traitNames = p$munge$traitNamesToUse,
            imputeFromLD=T,
            # imputeFrameLenBp = 500000, #use default setting of 0.5 CM
            # imputeFrameLenCM=NULL,
            # filter.region.imputation.df=p$highld_b38, #provide the high-ld regions to not use for imputation
            #N = p$munge$NToUse,
            #test = T, #REMOVE THIS!
            imputeFromLD.validate.q=1.0, #validate with all of the original dataset
            pathDirOutput = p$folderpath.data.sumstats.imputed.05cm,
            ancestrySetting = "EUR"
              )
}

if(file.exists(p$filepath.lfgwas.sumstats)) {
  # p$lfGWAS$sumstats<-readRDS(file=p$filepath.lfgwas.sumstats)
  # print("Read summary statistics for latent factor GWAS from file.")
  
  #compare with genomic SEM produced sumstats 
  # p$filepath.lfgwas.sumstats2<-file.path(p$folderpath.workingDirectory,paste0("lfGWAS.sumstats.setup4.Rds"))
  # p$lfGWAS$sumstats2<-readRDS(file=p$filepath.lfgwas.sumstats2)
  # 
  # View(p$lfGWAS$sumstats[which(p$lfGWAS$sumstats$SNP==c("rs1000000","rs10000013"))])
  
  # colBeta<-colnames(p$lfGWAS$sumstats)[grep("^BETA\\.", ignore.case = TRUE,colnames(p$lfGWAS$sumstats))]
  # colSE<-colnames(p$lfGWAS$sumstats)[grep("^SE\\.", ignore.case = TRUE,colnames(p$lfGWAS$sumstats))]
  # colBeta2<-colnames(p$lfGWAS$sumstats2)[grep("^BETA\\.", ignore.case = TRUE,colnames(p$lfGWAS$sumstats2))]
  # colSE2<-colnames(p$lfGWAS$sumstats2)[grep("^SE\\.", ignore.case = TRUE,colnames(p$lfGWAS$sumstats2))]
  # p$lfGWAS$sumstats<-as.data.frame(p$lfGWAS$sumstats)
  # for(iCol in 1:length(colBeta)){
  #   #iCol<-1
  #   print(colBeta[iCol])
  #   cBeta<-p$lfGWAS$sumstats[,colBeta[iCol]]
  #   cSE<-p$lfGWAS$sumstats[,colSE[iCol]]
  #   cBeta2<-p$lfGWAS$sumstats2[,colBeta2[iCol]]
  #   cSE2<-p$lfGWAS$sumstats2[,colSE2[iCol]]
  #   
  #   cat(mean(cBeta2,na.rm=T),"(",median(cSE2,na.rm=T),")\n")
  #   cat(mean(cBeta,na.rm=T),"(",median(cSE,na.rm=T),")\n")
  #   
  #   cat(max(abs(cBeta2),na.rm=T),"(",max(cSE2,na.rm=T),")\n")
  #   cat(max(abs(cBeta),na.rm=T),"(",max(cSE,na.rm=T),")\n")
  #   
  #   cat(min(abs(cBeta2),na.rm=T),"(",min(cSE2,na.rm=T),")\n")
  #   cat(min(abs(cBeta),na.rm=T),"(",min(cSE,na.rm=T),")\n")
  #   
  # }
  
  
  #Compare with old munged ANXI04
  #ssANXI04_old <- read.table(file.path(p$folderpath.data.sumstats.munged,"ANXI04_old.gz"),header=T, quote="\"",fill=T,na.string=c(".",NA,"NA",""))
  #setDT(ssANXI04_old)
  #setkeyv(ssANXI04_old, cols = c("SNP","BP","CHR"))
  #
  # ssANXI04 <- read.table(file.path(p$folderpath.data.sumstats.munged,"ANXI04.gz"),header=T, quote="\"",fill=T,na.string=c(".",NA,"NA",""))
  # setDT(ssANXI04)
  # setkeyv(ssANXI04, cols = c("SNP","BP","CHR"))
  #
  # ssANXI04.merged.snp<-ssANXI04[ssANXI04_old, on=c(SNP='SNP'), nomatch=0]
  # View(ssANXI04.merged.snp[1:100000,c("SNP","BP","CHR","A1","A2","i.A1","i.A2")])
  # sum(!(ssANXI04.merged.snp$A1==ssANXI04.merged.snp$i.A1 & ssANXI04.merged.snp$A2==ssANXI04.merged.snp$i.A2))
  # View(ssANXI04.merged.snp[1:100000,c("SNP","BP","CHR","EFFECT","i.EFFECT","Z","i.Z","P","i.P")])
  # View(ssANXI04.merged.snp[1:100000,c("SNP","BP","CHR","N","i.N")])

  # head(p$lfGWAS$sumstats)
  
}

if(p$clOptions$task=="munge2"){
  
  p$munge$results<-supermunge(
    filePaths = p$sumstats.sel$mungedpath,
    refFilePath = p$filepath.SNPReference.hc1kg,
    traitNames = p$sumstats.sel$code,
    filter.info = 0.6,
    filter.maf = 0.01,
    N = p$sumstats.sel$n_total,
    filter.region.df = p$highld_b38,
    #filter.chr = list(1,2,6,7,9,11),
    pathDirOutput = p$folderpath.data.sumstats.export.cleaned.nohighld
    )

  p$munge$results<-supermunge(
    filePaths = p$sumstats.sel$imputedpath.500,
    refFilePath = p$filepath.SNPReference.hc1kg,
    traitNames = p$sumstats.sel$code,
    filter.info = 0.6,
    filter.maf = 0.01,
    N = p$sumstats.sel$n_total,
    filter.region.df = p$highld_b38,
    #filter.chr = list(1,2,6,7,9,11),
    pathDirOutput = p$folderpath.data.sumstats.export.500.cleaned.nohighld
    )

  p$munge$results<-supermunge(
    filePaths = p$sumstats.sel.ssimp$imputedpath.ssimp,
    refFilePath = p$filepath.SNPReference.hc1kg,
    traitNames = p$sumstats.sel.ssimp$code,
    filter.info = 0.6,
    filter.maf = 0.01,
    N = p$sumstats.sel.ssimp$n_total,
    filter.region.df = p$highld_b38,
    filter.chr = list(1,2,6,7,9,11),
    pathDirOutput = p$folderpath.data.sumstats.export.ssimp.cleaned.nohighld
    )

  print("Done exporting cleaned + munged + no high LD regions")
} else if(p$clOptions$task=="composite"){
  p$munge$results<-supermunge(
    filePaths = p$sumstats.sel$imputedpath,
    refFilePath = p$filepath.SNPReference.hc1kg,
    traitNames = p$sumstats.sel$code,
    produceCompositeTable = T,
    process = F,
    standardiseEffectsToExposure = T,
    writeOutput = F,
    N = p$sumstats.sel$n_total,
    OLS=p$sumstats.sel$dependent_variable.OLS,
    linprob=p$sumstats.sel$dependent_variable.linprob,
    se.logit = p$sumstats.sel$se.logit,
    prop=(p$sumstats.sel$n_case/p$sumstats.sel$n_total),
    info.filter = 0.55
#    maf.filter = 0.001
    
    ) #remove rare variants only
    
  p$lfGWAS$sumstats<-p$munge$results$composite

  saveRDS(object = p$lfGWAS$sumstats, file = p$filepath.lfgwas.sumstats)
  print("Done preparing summary statistics for latent factor GWAS. The result should have been saved to a file.")
}

if(p$clOptions$task=="munge" ||  p$clOptions$task=="impute" || p$clOptions$task=="munge2" || p$clOptions$task=="composite") quit(save = "no")

```

## Format SSIMP munged datasets
```{r ssimp munged format}

if(p$clOptions$task=="ssimpformat"){
  
  cat("\nFormatting SSIMP munged datasets!\n")
  
  for(cTraitCode in p$sumstats.sel.ssimp.code) {
      cSumstats <- shru::readFile(filePath = p$sumstats.sel[cTraitCode,]$mungedpath, nThreads = 5)
      cSumstats[,BP:=NULL][,CHR:=NULL]
      fwrite(x = cSumstats,file = file.path(p$folderpath.data.sumstats.munged.4ssimp,paste0(cTraitCode,".gz")),append = F,quote = F,sep = "\t",col.names = T)
  }
  if(p$clOptions$task=="ssimpformat") quit(save = "no")
}

```


## Merge SSIMP imputed datasets

Delete the previous results first or they will be merged together with the intermediate results files!!!!

```{r GWAS imputed merge}

if(p$clOptions$task=="merge"){
  
  cat("\nMerging SSIMP fractured datasets!\n")
  
  #cRef <- shru::readFile(filePath = p$filepath.SNPReference.hc1kg,nThreads = 5)
  
  for(cTraitCode in p$sumstats.sel.ssimp.code) {
    #cTraitCode <- "NEUR02"
    #cTraitCode <- "BODY11"
    files<-list.files(path = p$folderpath.data.sumstats.imputed.ssimp, pattern = paste0("^",cTraitCode,"\\..+"))
    print(files)
    fullTraitList<-lapply(files, function(x){
      shru::readFile(filePath = file.path(p$folderpath.data.sumstats.imputed.ssimp,x),nThreads = 5)
    }
    )
    
    if(length(fullTraitList)>0){
      fullTrait <- rbindlist(l = fullTraitList, use.names = T, fill = T)
      rm(fullTraitList)
      #print(colnames(fullTrait))
      #test
      # fullTrait<-shru::readFile(filePath = p$sumstats.sel[cTraitCode,]$imputedpath.ssimp,nThreads = 5)
      
      #fullTrait
      
      colnames(fullTrait)[colnames(fullTrait)=="LD_IMP"]<-"LD_IMP.old"
      colnames(fullTrait)[colnames(fullTrait)=="Z"]<-"Z.old"
      colnames(fullTrait)[colnames(fullTrait)=="N"]<-"N.old"
      colnames(fullTrait)[colnames(fullTrait)=="P"]<-"P.old"
      colnames(fullTrait)[colnames(fullTrait)=="CHR"]<-"CHR.old"
      colnames(fullTrait)[colnames(fullTrait)=="BP"]<-"BP.old"
      colnames(fullTrait)[colnames(fullTrait)=="FRQ"]<-"FRQ.old"
      
      colnames(fullTrait)[colnames(fullTrait)=="r2.pred"]<-"LD_IMP"
      colnames(fullTrait)[colnames(fullTrait)=="z_imp"]<-"Z"
      colnames(fullTrait)[colnames(fullTrait)=="N.imp"]<-"N"
      colnames(fullTrait)[colnames(fullTrait)=="P.imp"]<-"P"
      
      print(colnames(fullTrait))
      
      
      #fix faulty SSimp output 
      mungedTrait <- shru::readFile(filePath = p$sumstats.sel[cTraitCode,]$mungedpath,nThreads = 5)
      #rename mungedTrait columns
      colnames(mungedTrait) <- paste0(colnames(mungedTrait),".m")
      print(colnames(mungedTrait))
      
      
      mergedFullTrait <- merge(fullTrait,mungedTrait, by.x="SNP", by.y = "SNP.m", all = T)
      print(colnames(mergedFullTrait))
      
      mergedFullTrait[(is.na(Z)&!is.na(Z.m)) || P>P.m,c("A1","A2","Z","P","N", "LD_IMP"):=list(A1.m,A2.m,Z.m,P.m,N.m,1.0)]
      mergedFullTrait[is.na(chr),chr:=CHR.m][is.na(pos),pos:=BP.m][is.na(maf),maf:=FRQ.m]
      
      mergedFullTrait<-mergedFullTrait[,.(SNP,A1,A2,CHR=chr,BP=pos,FRQ=maf,Z=Z,N=N,P=P,LD_IMP)]
      
      print(colnames(mergedFullTrait))
      
      shru::supermunge(list_df = list(mergedFullTrait) , traitNames = cTraitCode, parse=F, process = F, nThreads = 5, pathDirOutput = p$folderpath.data.sumstats.imputed.ssimp)
      
      rm(mergedFullTrait)
      rm(fullTrait)
      rm(mungedTrait)
    }
    
    rm(fullTraitList)
  }
  
  #rm(cRef)
  
  if(p$clOptions$task=="merge") quit(save = "no")
  
}



```

## GWIS of height and weight into BMI

```{r GWIS}

if(p$clOptions$task=="gwis"){
  
  cat("\nPerforming GWIS of height and weight\n")
  
  nCode = "BODYT2MI" #what to call the new sumstat
  
  
  
  # original GWIS implementation
  
  Weight_mean <- 83.858621
  Height_mean <- 1.81942192
  
  
  Weight_SD <- 12.981540
  Height_SD <- 0.07325379
  
  HW_cor <- 0.370959

  # Approximate the SD for BMI:
  BMI_SD <- deltamethod(~x1/(x2^2) ,mean=c(Weight_mean,Height_mean),cov=diag(c(Weight_SD,Height_SD)) %*% matrix(c(1,HW_cor,HW_cor,1),2,2) %*% diag(c(Weight_SD,Height_SD )))

  GIANT_Height_Men <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.imputed.05cm, "HEIG04M.gz")))
  GIANT_Weight_Men <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.imputed.05cm, "WEIG04M.gz")))
  
  if(any(colnames(GIANT_Height_Men)=="SINFO") & any(colnames(GIANT_Weight_Men)=="SINFO")){
    #TODO
    
    #reset validation SINFO
    #head(GIANT_Height_Men[GIANT_Height_Men$BETA!=GIANT_Height_Men$BETA.I | GIANT_Height_Men$SE!=GIANT_Height_Men$SE.I,])
    
    #y1[,IMPUTED:=(is.finite(BETA.I) & BETA.I==BETA)] #condition used in ldscpp
    GIANT_Height_Men$IMPUTED<-is.finite(GIANT_Height_Men$BETA.I) & GIANT_Height_Men$BETA.I==GIANT_Height_Men$BETA
    GIANT_Weight_Men$IMPUTED<-is.finite(GIANT_Weight_Men$BETA.I) & GIANT_Weight_Men$BETA.I==GIANT_Weight_Men$BETA
    
    GIANT_Height_Men$SINFO<-ifelse(GIANT_Height_Men$IMPUTED==T,GIANT_Height_Men$SINFO,NA_real_)
    GIANT_Weight_Men$SINFO<-ifelse(GIANT_Weight_Men$IMPUTED==T,GIANT_Weight_Men$SINFO,NA_real_)
    
    # GIANT_Height_Men<-GIANT_Height_Men[GIANT_Height_Men$FRQ>0.01 & (GIANT_Height_Men$SINFO>0.1 | is.na(GIANT_Height_Men$SINFO)),]
    # GIANT_Weight_Men<-GIANT_Weight_Men[GIANT_Weight_Men$FRQ>0.01 & (GIANT_Weight_Men$SINFO>0.1 | is.na(GIANT_Weight_Men$SINFO)),]
  }

  # GIANT_Height_Men <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.munged, "HEIG04M.gz")))
  # GIANT_Weight_Men <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.munged, "WEIG04M.gz")))

  GIANT_Combined <- merge(GIANT_Weight_Men,GIANT_Height_Men,by=1)
  rm(GIANT_Height_Men)
  rm(GIANT_Weight_Men)
  # compute new beta:

    # Rewscale beta and SE:
    Bw <- Weight_SD *  GIANT_Combined$BETA.x
    Bh <- Height_SD *  GIANT_Combined$BETA.y
  
  
    SEw <- Weight_SD *  GIANT_Combined$SE.x
    SEh <- Height_SD *  GIANT_Combined$SE.y
  
    # Frequency:
    FRQ <- as.vector(GIANT_Combined$FRQ.x)
    
  
    # Compute af weighted mean of 2 aproximated beta's
    AF1 <- 2*FRQ*(1-FRQ) / (2*FRQ*(1-FRQ) + FRQ^2)
  
    B1 <- ( ((Weight_mean  + Bw )/ (Height_mean + Bh)^2) - (Weight_mean /(Height_mean^2)))
  
    AF2 <- ((FRQ^2) ) / (2*FRQ*(1-FRQ) + FRQ^2)
  
    B2 <-(((Weight_mean  + 2*Bw) / ((Height_mean + 2*Bh)^2)) - ( Weight_mean /(Height_mean^2))) / 2
  
  
    new.beta.us <-    (AF1 * B1 + AF2 * B2 )
    new.beta <- (new.beta.us / BMI_SD) 
  
  # compute new SE:
  
    N <- length(new.beta)
    cat("\nGWIS has ", N,  " rows after filters.\n\n")
    #deltamethod
    cat("\nComputing SEs with the deltamethod...")
  
    new.se <- matrix(NA,nrow=length(Bh),ncol=1)
    for(i in 1:length(new.beta)){
      if(i %% 100000 == 0) cat("*") else if(i %% 10000 == 0) cat(".")
      
      AF1_l <- AF1[i]
      AF2_l <- AF2[i] 
  
      new.se[i]<- deltamethod(~( AF1_l*(((Weight_mean+x1)/((Height_mean+x2)^2)) -(Weight_mean/Height_mean^2)) + (AF2_l*(((Weight_mean+2*x1)/((Height_mean+2*x2)^2)) -(Weight_mean/Height_mean^2)))/2)/BMI_SD, mean=c(Bw[i],Bh[i]),
                               
                               cov=cbind(rbind((SEw[i])^2,           (SEw[i])*HW_cor*(SEh[i])),
                                         rbind((SEw[i])*HW_cor*(SEh[i]),      (SEh[i])^2)))
      }
  
  
  
  # build an output file:
  
  GWIS_BMI <- as.data.frame(cbind(GIANT_Combined$SNP,GIANT_Combined$CHR.x,GIANT_Combined$BP.x,GIANT_Combined$A1.x,GIANT_Combined$A2.x,new.beta,new.se,(pchisq((new.beta/new.se)^2,1,lower.tail=F)),(GIANT_Combined$N.x+GIANT_Combined$N.y)/2))
  
  colnames(GWIS_BMI) <- c("SNP","CHR","BP","A1","A2","BETA","SE","P","N")
  
  if(any(colnames(GIANT_Combined)=="SINFO.x") | any(colnames(GIANT_Combined)=="SINFO.y")){
    GWIS_BMI$SINFO<-(GIANT_Combined$SINFO.x+GIANT_Combined$SINFO.y)/2
  }
  
  GWIS_BMI$CHR<-as.integer(GWIS_BMI$CHR)
  GWIS_BMI$BP<-as.integer(GWIS_BMI$BP)
  GWIS_BMI$P<-as.numeric(GWIS_BMI$P)
  
  supermunge(
    list_df = list(GWIS_BMI),
    refFilePath = p$filepath.SNPReference.hc1kg,
    traitNames = c(nCode),
    pathDirOutput = p$folderpath.data.sumstats.munged
      )
  
  if(p$clOptions$task=="gwis") quit(save = "no") #end gwis task - only manual stuff below
  
  
  #read previous results
  GWISM_BMI <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.munged, "BODYT2M.gz")))
  GWISI_BMI <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.imputed, "BODYT2MI.gz")))
  
 
  
  # PLOT GWIS Mwenahttan:
  
  
  
  #manhattan(na.omit(GWIS_BMI),chr="GIANT_Combined$chrom",bp="GIANT_Combined$pos",p="(pchisq((new.beta/new.se)^2, 1, lower.tail = F))",ylim=c(0,50))
  
  cPlot<-plot.manhattan.custom(GWISM_BMI,var = "P")
  ggsave(filename=file.path(p$folderpath.plots, paste0("manp.GWIS.GWISM.png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  
  cPlot<-plot.manhattan.custom(GWISI_BMI,var = "P")
  ggsave(filename=file.path(p$folderpath.plots, paste0("manp.GWIS.GWISI.png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  
  #  Read in the real BMI results:
  
  
  REFERENCE_BMI <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.munged, "BODYTM.gz")))
  REFERENCE_BMI2 <- as.data.frame(shru::readFile(file.path(p$folderpath.data.sumstats.munged, "BODY11.gz")))
  
  
  # Merge to test:
  
  GWIS_GWAS <- merge(GWISI_BMI,REFERENCE_BMI,by=1)
  GWIS_GWAS.f <- GWIS_GWAS[GWIS_GWAS$N.x > 58000 & GWIS_GWAS$N.y > 58000,] # & (GWIS_GWAS$Z.x)^2>0.01579077 & (GWIS_GWAS$Z.y)^2>0.01579077
  
  
  cor(GWIS_GWAS.f$BETA.x,GWIS_GWAS.f$BETA.y,use="pairwise.co")^2
  cor(GWIS_GWAS.f$SE.x,GWIS_GWAS.f$SE.y,use="pairwise.co")^2
  cor(GWIS_GWAS.f$Z.x,GWIS_GWAS.f$Z.y,use="pairwise.co")^2
  
  
  tiff(file.path(p$folderpath.plots,"GWISI_REF1_corr.tiff"), width = 4, height = 8, units = 'in', res = 300)
  # Make plot
  
  par(mfrow=c(3,1))
  
  plot(GWIS_GWAS.f$BETA.x,GWIS_GWAS.f$BETA.y, col=rgb(0,0,1,.2),pch=19,ylab="Beta's GWAS BMI", xlab="Beta's GWIS BMI Male Height" )
  legend("topleft",expression(R^2 ~ .935))
  abline(0,1,col="red")
  
  plot(GWIS_GWAS.f$SE.x,GWIS_GWAS.f$SE.y, col=rgb(0,0,1,.2),pch=19,ylab="SE GWAS BMI", xlab="SE GWIS BMI Male Height" )
  legend("topleft",expression(R^2 ~ .996))
  abline(0,1,col="red")
  
  plot(GWIS_GWAS.f$Z.x,GWIS_GWAS.f$Z.y, col=rgb(0,0,1,.2),pch=19,ylab="Z GWAS BMI", xlab="Z GWIS BMI Male Height" )
  legend("topleft",expression(R^2 ~ .936))
  abline(0,1,col="red")
  
  dev.off()
  
  
  #`1000G_rs_ch_bp` <- read.table("C:/8. post hoc GWAS/FINAL Analysis paper/1000G_rs_ch_bp", quote="\"")
  
  
  #manhattan(na.omit(GWIS_GWAS),chr="GIANT_Combined$chrom",bp="GIANT_Combined$pos",p="P.2gc",ylim=c(0,50))
  cPlot<-plot.manhattan.custom(REFERENCE_BMI,var = "P")
  ggsave(filename=file.path(p$folderpath.plots, paste0("manp.GWIS.BODYTM.png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  
  
  
  # Merge to test 2:
  
  GWIS_GWAS2 <- merge(GWISI_BMI,REFERENCE_BMI2,by=1)
  GWIS_GWAS2.f <- GWIS_GWAS2[GWIS_GWAS2$N.x > 58000 & GWIS_GWAS2$N.y > 58000,] #& (GWIS_GWAS2$Z.x)^2>0.1 & (GWIS_GWAS2$Z.y)^2>0.1 #0.01579077
  
  
  cor(GWIS_GWAS2.f$BETA.x,GWIS_GWAS2.f$BETA.y,use="pairwise.co")^2
  cor(GWIS_GWAS2.f$SE.x,GWIS_GWAS2.f$SE.y,use="pairwise.co")^2
  cor(GWIS_GWAS2.f$Z.x,GWIS_GWAS2.f$Z.y,use="pairwise.co")^2
  
  
  tiff(file.path(p$folderpath.plots,"GWISI_REF2_corr.tiff"), width = 4, height = 8, units = 'in', res = 300)
  # Make plot
  
  par(mfrow=c(3,1))
  
  plot(GWIS_GWAS2.f$BETA.x,GWIS_GWAS2.f$BETA.y, col=rgb(0,0,1,.2),pch=19,ylab="Beta's GWAS BMI", xlab="Beta's GWIS BMI Male Height" )
  legend("topleft",expression(R^2 ~ .156))
  abline(0,1,col="red")
  
  plot(GWIS_GWAS2.f$SE.x,GWIS_GWAS2.f$SE.y, col=rgb(0,0,1,.2),pch=19,ylab="SE GWAS BMI", xlab="SE GWIS BMI Male Height" )
  legend("topleft",expression(R^2 ~ .889))
  abline(0,1,col="red")
  
  plot(GWIS_GWAS2.f$Z.x,GWIS_GWAS2.f$Z.y, col=rgb(0,0,1,.2),pch=19,ylab="Z GWAS BMI", xlab="Z GWIS BMI Male Height" )
  legend("topleft",expression(R^2 ~ .186))
  abline(0,1,col="red")
  
  dev.off()
  
  
  #`1000G_rs_ch_bp` <- read.table("C:/8. post hoc GWAS/FINAL Analysis paper/1000G_rs_ch_bp", quote="\"")
  
  
  #manhattan(na.omit(GWIS_GWAS),chr="GIANT_Combined$chrom",bp="GIANT_Combined$pos",p="P.2gc",ylim=c(0,50))
  cPlot<-plot.manhattan.custom(REFERENCE_BMI2,var = "P")
  ggsave(filename=file.path(p$folderpath.plots, paste0("manp.GWIS.BODY11.png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  
  
  
  
  # B B Plot
  #plot(GWIS_GWAS[,5],GWIS_GWAS[,12],col=rgb(0,0,1,.1),ylim=c(-.35,.35),pch=19,ylab="BMI GWAS Beta",xlab="BMI GWIS Beta")
  # SE SE Plot
  #plot(GWIS_GWAS[,6],GWIS_GWAS[,13],col=rgb(0,0,1,.1),ylim=c(0,.16),pch=19,ylab="BMI GWAS SE",xlab="BMI GWIS SE")
  # P P Plot
  plot(-log10(GWIS_GWAS$P.x),-log10(GWIS_GWAS$P.y),col=rgb(0,0,1,.1),ylim=c(0,16),xlim=c(0,16),pch=19,ylab="BMI GWAS -log10(P)",xlab="BMI GWIS -log10(P)")
  
  # Number of Hits GWAS:
  nrow(GWIS_GWAS[GWIS_GWAS$P.x < 5e-8,])
  
  # Numer of hits replicated:
  sum(GWIS_GWAS[GWIS_GWAS$P.x < 5e-8,]$P.y < 5e-8)
  
  # numer of false positives:
  sum(GWIS_GWAS[GWIS_GWAS$P.x < 5e-8,]$P.y > 5e-8)
  
  
  # Table the false positives for inclusion in a supplment:
  
  fp <-GWIS_GWAS[GWIS_GWAS$P.x < 5e-8 & GWIS_GWAS$P.y > 5e-8 ,]
  
  write.table(fp,"false_positive_male.txt")

  
  # Number of Hits GWAS2:
  nrow(GWIS_GWAS2[GWIS_GWAS2$P.x < 5e-8,])
  
  # Numer of hits replicated:
  sum(GWIS_GWAS2[GWIS_GWAS2$P.x < 5e-8,]$P.y < 5e-8)
  
  # numer of false positives:
  sum(GWIS_GWAS2[GWIS_GWAS2$P.x < 5e-8,]$P.y > 5e-8)
  
  
  # # nwe implementation - on pause
  # #performGWIS <- function(nCode,pathGwasH,patghGwasWM,pathGwasWF){
  #   pathGWISI = p$sumstats["GWISI",]$mungedpath
  # 
  #   
  #   pathGwasH = p$sumstats["HEIG09",]$imputedpath
  #   patghGwasWM = p$sumstats["WEIG04M",]$imputedpath
  #   pathGwasWF = p$sumstats["WEIG04F",]$imputedpath
  #   
  #   
  #   
  #   # pathGwasH = p$sumstats["HEIG09",]$mungedpath
  #   # patghGwasWM = p$sumstats["WEIG04M",]$mungedpath
  #   # pathGwasWF = p$sumstats["WEIG04F",]$mungedpath
  #   
  #   HW_cor <- 0.370959 #from GWIS example
  #   
  #   #10.1194/jlr.P041673
  #   mHeightM <- (168.85/15095+181.31/9809)/(1/15095+1/9809)/100
  #   sdHeightM <- sqrt(((6.54^2)/15095+(7.37^2)/9809)/(1/15095+1/9809))/100
  #   mBMIKgpM <- (24.290/16312+24.772/10709)/(1/16312+1/10709)
  #   sdBMIKgpM <- sqrt(((4.389^2)/16312+(4.389^2)/10709)/(1/16312+1/10709))
  #   #mWeightKg <- mBMIKgpM * mHeightM^2
  #   
  #   #10.1038/ijo.2008.11
  #   mWeightKg <- (82.0/6580+69.3/7021)/(1/6580+1/7021)
  #   sdWeightKg <- sqrt(((0.19^2)/6580+(0.20^2)/7021)/(1/6580+1/7021))
  #   
  #   #https://stats.stackexchange.com/questions/201358/distribution-of-the-square-of-a-non-standard-normal-random-variable
  #   #E[Non-central Chi-squared variable, k] = k + l
  #   #Var[Non-central Chi-squared variable, k] = 2(k +2l)
  #   # varHeight2 <-  2 + 2*((mHeightM/sdHeightM)^2)
  #   # mHeight2 <- 1 + (mHeightM/sdHeightM)^2
  #   #Find variance for BMI*Height^2=Weight
  #   #Goodman's formula. How to compute cov(X^2,Y^2) in the dependent variables formula???
  #   #Var(XY)=Cov(X^2,Y^2) + (Var(X) + [E(X)]^2) * (Var(Y) + [E(Y)]^2) - [Cov(X,Y) + E(X)*E(Y)]^2
  #   #for the reciprocal normal distribution the Var(1/X) ~= Var(X)
  #   # covH2BMI<-0.75*sqrt(varHeight2)*sdBMIKgpM #Just an arbitrary placeholder
  #   # covH22BMI2<-sqrt(covH2BMI) #Just an arbitrary placeholder
  #   # sdWeightKg <- sqrt(
  #   #   covH22BMI2 + (sdBMIKgpM^2+mBMIKgpM^2) * (varHeight2 + mHeight2^2) - (covH2BMI + mBMIKgpM*mHeight2)^2
  #   #   )
  #   
  #   gwasGWISI<-readFile(filePath = pathGWISI) #for QC
  # 
  #   gwasH<-readFile(filePath = pathGwasH)
  #   gwasWM<-readFile(filePath = patghGwasWM)
  #   gwasWF<-readFile(filePath = pathGwasWF)
  #   
  #   #The deltamethod is not fast so let's filter the data to what we will be using later anyway
  #   gwasH<-gwasH[FRQ>0.01 & (Z*Z)>0.01579077,]
  #   gwasWM<-gwasWM[FRQ>0.01 & (Z*Z)>0.01579077,]
  #   gwasWF<-gwasWF[FRQ>0.01 & (Z*Z)>0.01579077,]
  #   
  #   #this is probably not best practice
  #   # gwasH[FRQ==0,FRQ:=1/(10*N)]
  #   # gwasWM[FRQ==0,FRQ:=1/(10*N)]
  #   # gwasWF[FRQ==0,FRQ:=1/(10*N)]
  #   
  #   gwasH[,W:=1/(SE^2)] #meta-analytical weights
  #   gwasWM[,W:=1/(SE^2)] #meta-analytical weights
  #   gwasWF[,W:=1/(SE^2)] #meta-analytical weights
  #   
  #   if(!any(colnames(gwasH)=="SINFO")) { 
  #     gwasH$SINFO<-NA_real_
  #   } else {
  #     #Height has wrong SINFO for some reason
  #     gwasH[,SINFO:=as.double(SINFO)]
  #     gwasH[LDIMP.L2.SUM>0,LDIMP.Q:=LDIMP.K*L2_REF/(LDIMP.L2.SUM)][LDIMP.L2.SUM>0,c('SINFO'):=list(pnorm(q = LDIMP.Q, mean = 1, sd = 1))]
  #   }
  #   if(!any(colnames(gwasWM)=="SINFO")) gwasWM$SINFO<-NA_real_
  #   if(!any(colnames(gwasWF)=="SINFO")) gwasWF$SINFO<-NA_real_
  #   
  #   #merge male and female weight GWAS
  #   gwasW<-copy(gwasWM)
  #   gwasW[gwasWF, on="SNP", c("BETA","SE","N","FRQ","SINFO"):=list((W*BETA+i.W*i.BETA)/(W+i.W), sqrt(1/(W+i.W)), N+i.N, shru::clipValues((FRQ/N+i.FRQ/i.N)/(1/N+1/i.N),min = 1/(10*N), max = 1),ifelse(is.na(SINFO) & is.na(i.SINFO), NA_real_, (ifelse(is.na(SINFO),1,SINFO)+ifelse(is.na(i.SINFO),1,i.SINFO))/2))][,W:=1/(SE^2)]
  #   #gwasW[gwasWF, on="SNP", c("BETA","SE","N","FRQ"):=list( (BETA/(N*FRQ)+i.BETA/(i.N*i.FRQ))/(1/(N*FRQ)+1/(i.N*i.FRQ)), sqrt((SE^2/(N*FRQ)+i.SE^2/(i.N*i.FRQ))/(1/(N*FRQ)+1/(i.N*i.FRQ))), N+i.N, shru::clipValues( (FRQ/N+i.FRQ/i.N)/(1/N+1/i.N),min = 1/(10*N), max = 1))] #old, non-meta-analysis
  #   
  #   #standardisation and GWIS formulas
  #   gwasBMI <- copy(gwasH)
  #   gwasBMI[gwasW,on=c("SNP"),c("BETA.W","SE.W","BETA.H","SE.H","N","FRQ","Z.H","Z.W","N.H","N.W","SINFO"):=list(i.BETA, i.SE, BETA, SE, (N+i.N)/2, shru::clipValues( (FRQ/N+i.FRQ/i.N)/(1/N+1/i.N),min = 1/(10*N), max = 1), Z, i.Z, N, i.N,ifelse(is.na(SINFO) & is.na(i.SINFO), NA_real_, (ifelse(is.na(SINFO),1,SINFO)+ifelse(is.na(i.SINFO),1,i.SINFO))/2))]
  #   
  #   gwasBMI<-gwasBMI[is.finite(BETA.W) & is.finite(BETA.H),]
  #   
  #   #different types of Z
  #   gwasBMI[,Z.BMI:=Z.W/(Z.H^2)]
  #   gwasBMI[,Z.WEIGHTED:=(Z.W/(N.W*(FRQ-FRQ^2+1/2))+Z.H/(N.H*(FRQ-FRQ^2+1/2)))/(1/(N.W*(FRQ-FRQ^2+1/2))+1/(N.H*(FRQ-FRQ^2+1/2)))]
  #   
  #   gwasBMI[,c("I.W","I.H","I.BMI"):=list(eval(..mWeightKg),eval(..mHeightM),eval(..mWeightKg)/(eval(..mHeightM)^2))]
  #   
  #   #scaling etc
  #   gwasBMI[,c("BETA.W.S","BETA.H.S","SE.W.S","SE.H.S"):=list(
  #     BETA.W*eval(..sdWeightKg), #scaled beta
  #     BETA.H*eval(..sdHeightM), #scaled beta
  #     SE.W*eval(..sdWeightKg), #scaled standard error
  #     SE.H*eval(..sdHeightM) #scaled standard error
  #   )][,c('AF1','AF2'):=list((2*FRQ*(1-FRQ)/(2*FRQ*(1-FRQ)+FRQ^2)),(FRQ^2)/(2*FRQ*(1-FRQ)+FRQ^2))]
  #   
  #   #H2 variable estimates - HERE!!! SEE THE NEW FORMULAS ON THE PAPER!
  #   gwasBMI[,c("SE.H2.S","E.H2.S"):=list(
  #     sqrt(SE.H.S^4 * 2), # Var(aX) = a^2 * Var(X), Var(X)= 2k
  #     1 * SE.H.S^2 # k * var(H)
  #     # sqrt(2 + 2*(2*abs(BETA.H.S)/((FRQ-FRQ^2+1/2)*SE.H.S))^2), #scaled standard error, m = 2*BETA/X
  #     # 1 + 2*abs(BETA.H.S)/(FRQ-FRQ^2+1/2) #scaled estimated value - THIS IS THE MEAN OF H2!!!
  #   )]
  #  
  #   #GWIS
  #   gwasBMI[,c("BETA"):=list(
  #     (AF1*((I.W+BETA.W.S)/((I.H+BETA.H.S)^2) - I.BMI) + (1/2)*AF2*((I.W+2*BETA.W.S)/((I.H+2*BETA.H.S)^2) - I.BMI))/eval(..sdBMIKgpM) #this last denominator thing was added in the GWIS code example, but is not documented in the paper (I think). 
  #   )]
  #   
  #   
  #   #SE of BMI - a ratio of random variables - no closed form solution for the variance
  #   
  #   
  #   #The GWIS solution from the original implementation
  #   
  #   # #test
  #   # HW_cor <- 0.370959
  #   # AF1_l <- (2*0.5154780*(1-0.5154780)/(2*0.5154780*(1-0.5154780)+0.5154780^2))
  #   # AF2_l <- (1/2)*((0.5154780^2)/(2*0.5154780*(1-0.5154780)+0.5154780^2))
  #   # 
  #   # new.se <- deltamethod(~( AF1_l*(((mWeightKg+x1)/((mHeightM+x2)^2)) - mBMIKgpM) + (AF2_l*(((mWeightKg+2*x1)/((mHeightM+2*x2)^2)) - mBMIKgpM))/2)/sdBMIKgpM, mean=c(-0.008090578*sdWeightKg,0.000489998*sdHeightM),
  #   #                            
  #   #                            cov=cbind(rbind((0.004523826*sdWeightKg)^2,           (0.004523826*sdWeightKg)*HW_cor*(0.00105*sdHeightM)),
  #   #                                      rbind((0.004523826*sdWeightKg)*HW_cor*(0.00105*sdHeightM),(0.00105*sdHeightM)^2)))
  #   
  #   
  #   
  #   # https://www.stat.cmu.edu/~hseltman/files/ratio.pdf
  #   # https://math.stackexchange.com/questions/40713/calculating-the-variance-of-the-ratio-of-random-variables
  #   gwasBMI[,c("SE.DERIVED","E.BMI","SE.BMI.Z.DERIVED","SE.WEIGHTED"):=list(
  #       sqrt(
  #         1/(E.H2.S^4)*((SE.W.S^2)/1 - 2*eval(..HW_cor)*SE.W.S*SE.H2.S/((BETA.W.S)^2) + (SE.H2.S^2)/(E.H2.S^2))
  #            ), #use the HW correlation as approximation. the approximation of a ratio random variables using taylor series. Issue with E[W]=0, approximately Var(W) when E[W] -> 0 ??
  #       1/E.H2.S - eval(..HW_cor)*SE.W.S*SE.H2.S/((BETA.W.S)^2) + (SE.H2.S^2)*1/((BETA.W.S)^3), #use the HW correlation as approximation. the approximation of a ratio random variables using taylor series.
  #       abs(BETA/Z.BMI),
  #       abs(BETA/Z.WEIGHTED)
  #       
  #   )]
  #   
  #   N <- nrow(gwasBMI)
  #   cat("\nGWIS has ", N,  " rows after filters.\n\n")
  #   #deltamethod
  #   cat("\nComputing SEs with the deltamethod...")
  #   for(i in 1:N){ #nrow(gwasBMI)
  #     #i<-1
  #     if(i %% 100000 == 0) cat("*")
  #     else if(i %% 10000 == 0) cat(".")
  #     
  #     AF1_l <- gwasBMI[i,]$AF1
  #     AF2_l <- gwasBMI[i,]$AF2
  #     
  #     set(x = gwasBMI,i = i,j = "SE.DELTAMETHOD",
  #                 value = deltamethod(~( AF1_l*(((mWeightKg+x1)/((mHeightM+x2)^2)) - mBMIKgpM) + ((1/2)*AF2_l*((mWeightKg+2*x1)/((mHeightM+2*x2)^2) - mBMIKgpM)))/sdBMIKgpM,
  #                     mean=c(gwasBMI[i,]$BETA.W.S,gwasBMI[i,]$BETA.H.S),
  #                     cov=cbind(rbind(gwasBMI[i,]$SE.W.S^2,gwasBMI[i,]$SE.W.S*HW_cor*gwasBMI[i,]$SE.H.S),rbind(gwasBMI[i,]$SE.W.S*HW_cor*gwasBMI[i,]$SE.H.S,gwasBMI[i,]$SE.H.S^2)))
  #             )
  #     # gwasBMI[i,c('SE.DELTAMETHOD')] <- deltamethod(~( AF1_l*(((mWeightKg+x1)/((mHeightM+x2)^2)) - mBMIKgpM) + ((1/2)*AF2_l*((mWeightKg+2*x1)/((mHeightM+2*x2)^2) - mBMIKgpM)))/sdBMIKgpM,
  #     #                 mean=c(gwasBMI[i,]$BETA.W.S,gwasBMI[i,]$BETA.H.S),
  #     #                 cov=cbind(rbind(gwasBMI[i,]$SE.W.S^2,gwasBMI[i,]$SE.W.S*HW_cor*gwasBMI[i,]$SE.H.S),rbind(gwasBMI[i,]$SE.W.S*HW_cor*gwasBMI[i,]$SE.H.S,gwasBMI[i,]$SE.H.S^2)))
  #   }
  #   cat("done!")
  #   #SEs of squared random variables
  #   # gwasBMI[,c("SE.W.S.2","SE.H2.S.2"):=list(
  #   #   sqrt(SE.W.S^4+2*I.W^2), #approximation of these as uncorrelated
  #   #   sqrt(SE.H2.S^4+2*I.H2^2)  #approximation of these as uncorrelated
  #   #   # sqrt(SE.W.S^4+2*(SE.W.S^2)*I.W^2-2*I.W^2), #correlated
  #   #   # sqrt(SE.H2.S^4+2*(SE.H2.S^2)*I.H2^2-2*I.H2^2) #correlated
  #   # )]
  #   #test
  #   #sqrt(0.001248959^4 + 2*75.85589^2) #uncorrelated     
  #   #sqrt(0.001248959^4 + 2*(0.001248959^2)*75.85589^2-2*75.85589^2) #correlated
  #   
  #   gwasBMI[,SE:=SE.DELTAMETHOD][,c("Z"):=list(BETA/SE)][,P:=2*pnorm(q = abs(Z), mean = 0, sd = 1, lower.tail = F)]
  #   
  #   gwasBMI <- gwasBMI[is.finite(BETA.W) & is.finite(P), c("SNP","A1","A2","CHR","BP","FRQ","BETA","SE","Z","P","N","SINFO")]
  #   
  #   cat("\nNumber of rows in GWIS: ", nrow(gwasBMI))
  #   
  #   fwrite(x = gwasBMI,file = file.path(p$folderpath.data.sumstats.munged,paste0(nCode,".gz")),append = F,quote = F,sep = "\t",col.names = T,nThread=5)
  #   
  # #}
  # 
  # # performGWIS(
  # #   nCode = "GWISM",
  # #   pathGwasH = p$sumstats["HEIG09",]$mungedpath,
  # #   patghGwasWM = p$sumstats["WEIG04M",]$mungedpath,
  # #   pathGwasWF = p$sumstats["WEIG04F",]$mungedpath
  # #   )
  # # 
  # # performGWIS(
  # #   nCode = "GWISI",
  # #   pathGwasH = p$sumstats["HEIG09",]$imputedpath,
  # #   patghGwasWM = p$sumstats["WEIG04M",]$imputedpath,
  # #   pathGwasWF = p$sumstats["WEIG04F",]$imputedpath
  # #   )
  # 
  if(p$clOptions$task=="gwis") quit(save = "no")
  
}



```



# Visualise GWAS effect distributions
```{r visualise distributions, fig.width=9, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}

  #munged
  for(iTrait in 1:nrow(p$sumstats.sel)){
    cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$mungedpath, nThreads = 5)
    #cSumstats <- cSumstats[,]
    cPlot<-plot.manhattan.custom(cSumstats,var = "BETA")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manbeta.munged.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "SE")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manse.munged.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "P")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manp.munged.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  }

  #imputed ld-imp
  for(iTrait in 1:nrow(p$sumstats.sel)){
    cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$imputedpath.500, nThreads = 5)
    cPlot<-plot.manhattan.custom(cSumstats,var = "BETA")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manbeta.ld-imp.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "SE")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manse.ld-imp.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "P")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manp.ld-imp.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  }

  #imputed ssimp
  # for(iTrait in 1:nrow(p$sumstats.sel)){
  #   cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$imputedpath.ssimp, nThreads = 5)
  #   # cPlot<-plot.manhattan.custom(cSumstats,var = "BETA")
  #   # ggsave(filename=file.path(p$folderpath.plots, paste0("manbeta.ssimp.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  #   # cPlot<-plot.manhattan.custom(cSumstats,var = "SE")
  #   # ggsave(filename=file.path(p$folderpath.plots, paste0("manse.ssimp.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  #   cPlot<-plot.manhattan.custom(cSumstats,var = "P")
  #   ggsave(filename=file.path(p$folderpath.plots, paste0("manp.ssimp.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  # }
  
  #munged2
  for(iTrait in 1:nrow(p$sumstats.sel)){
    cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$mungedpath2, nThreads = 5)
    cPlot<-plot.manhattan.custom(cSumstats,var = "BETA")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manbeta.munged2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "SE")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manse.munged2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "P")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manp.munged2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  }
  
  #imputed ld-imp, munge2
  for(iTrait in 1:nrow(p$sumstats.sel)){
    cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$imputedpath2.500, nThreads = 5)
    cPlot<-plot.manhattan.custom(cSumstats,var = "BETA")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manbeta.ld-imp2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "SE")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manse.ld-imp2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "P")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manp.ld-imp2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  }
  
  #imputed ssimp, munge2
  for(iTrait in 1:nrow(p$sumstats.sel)){
    cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$imputedpath2.ssimp, nThreads = 5)
    cPlot<-plot.manhattan.custom(cSumstats,var = "BETA")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manbeta.ssimp2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "SE")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manse.ssimp2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
    cPlot<-plot.manhattan.custom(cSumstats,var = "P")
    ggsave(filename=file.path(p$folderpath.plots, paste0("manp.ssimp2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)
  }
  
  #quit(save = "no")
  

```

# Per-dataset statistics

Includes plotting of MAF vs L2. This takes a long time to run as we have to read all datasets.

```{r dataset statistics}

print("***Dataset statistics***")

p$filepath.dsStat<-file.path(p$folderpath.workingDirectory,paste0("dsStat.",p$setup.code,".Rds"))

if(file.exists(p$filepath.dsStat)) {
  print("Using existing dataset statistics.")
  p$dsStat<-readRDS(file=p$filepath.dsStat)
}

if(p$clOptions$task=="stat") {
  
  p$dsStat<-c()
  p$dsStat$sumstats.meta<-data.table(code=p$sumstats.sel.raw$code)
  
  for(iTrait in 1:nrow(p$sumstats.sel.raw)){
    #iTrait<-1
    if(!file.exists(p$sumstats.sel.raw[iTrait,]$imputedpath.05cm)) next
    cat("\n",p$sumstats.sel.raw[iTrait,]$code)
    cSumstats <- shru::readFile(filePath = p$sumstats.sel.raw[iTrait,]$imputedpath.05cm, nThreads = p$nThread)

    if(!any(colnames(cSumstats)=="BETA.I")) next



    #nrow(cSumstats)
    cond.invertedMAF<-cSumstats$FRQ > .5
    cSumstats$MAF<-ifelse(cond.invertedMAF,1-cSumstats$FRQ,cSumstats$FRQ)
    #cSumstats <- cSumstats[MAF>0.01,]
    #nrow(cSumstats)
    cSumstats[,ZZ:=Z*Z][,ZZ.I:=(BETA.I/SE.I)^2][,AC:=FRQ*N][,MAC:=MAF*N][,VSNP:=2*FRQ*(1-FRQ)][,NEFF:=round((((BETA.I/SE.I)/BETA.I)^2)/VSNP,digits = 0)][,IMPUTED:=(is.finite(LDIMP.L2.SUM) & BETA.I==BETA)]

    # cSumstats$SINFO<-as.double(cSumstats$SINFO)
    # cSumstats[IMPUTED==T & is.finite(L2_REF),SINFO:=pnorm(q = LDIMP.K*L2_REF/(LDIMP.L2.SUM), mean = 1, sd = 1)]

    setkeyv(cSumstats,cols = c("CHR","BP"))
    #[is.finite(BETA) & is.finite(BETA.I) & is.finite(SE.I),]
    cSumstats[,c('d.z','d.beta','d.se') :=list((BETA.I/SE.I)-(BETA/SE),BETA.I-BETA,SE.I-SE)]
    cSumstats[,d.z.abs:=abs(d.z)]


    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.genotyped")]<-median(cSumstats[is.finite(ZZ) & IMPUTED==F,]$ZZ,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed")]<-median(cSumstats[is.finite(ZZ.I),]$ZZ.I,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.genotyped.05_50")]<-median(cSumstats[is.finite(ZZ) & IMPUTED==F & MAF>0.05 & MAF<0.5,]$ZZ,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed.05_50")]<-median(cSumstats[is.finite(ZZ.I) & MAF>0.05 & MAF<0.5,]$ZZ.I,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.genotyped.01_05")]<-median(cSumstats[is.finite(ZZ) & IMPUTED==F & MAF>0.01 & MAF<0.05,]$ZZ,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed.01_05")]<-median(cSumstats[is.finite(ZZ.I) & MAF>0.01 & MAF<0.05,]$ZZ.I,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.genotyped.0_01")]<-median(cSumstats[is.finite(ZZ) & IMPUTED==F & MAF<0.01,]$ZZ,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed.0_01")]<-median(cSumstats[is.finite(ZZ.I) & MAF<0.01,]$ZZ.I,na.rm=T)

    # p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.genotyped")]<-median(cSumstats[is.finite(ZZ) & IMPUTED==F,]$ZZ,na.rm=T)
    # p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed")]<-median(cSumstats[is.finite(ZZ.I),]$ZZ.I,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("l_ldimp_reinflation")]<-p$dsStat$sumstats.meta[iTrait,]$medianAssocChi2.imputed/p$dsStat$sumstats.meta[iTrait,]$medianAssocChi2.genotyped
    #p$dsStat$sumstats.meta[iTrait,c("l_ldimp_reinflation")]<-p$dsStat$sumstats.meta[iTrait,]$meanAssocChi2.imputed/p$dsStat$sumstats.meta[iTrait,]$meanAssocChi2.genotyped


    #re-inflation test
    cSumstats[,c(
      'd.z.re',
      'd.z.re2',
      'd.z.re3'
      ) :=list(
        (BETA.I/(eval(p$dsStat$sumstats.meta[iTrait,]$l_ldimp_reinflation)*SE.I))-(BETA/SE),
        (BETA.I/(eval(p$dsStat$sumstats.meta[iTrait,]$l_ldimp_reinflation^(1/2))*SE.I))-(BETA/SE),
        (BETA.I/(eval(p$dsStat$sumstats.meta[iTrait,]$l_ldimp_reinflation^(1/3))*SE.I))-(BETA/SE)
      )
      ]

    #re-inflation
    cSumstats[,ZZ.re:=(BETA.I/(eval(p$dsStat$sumstats.meta[iTrait,]$l_ldimp_reinflation)*SE.I))^2]
    cSumstats[,ZZ.re2:=(BETA.I/(eval(p$dsStat$sumstats.meta[iTrait,]$l_ldimp_reinflation^(1/2))*SE.I))^2]
    cSumstats[,ZZ.re3:=(BETA.I/(eval(p$dsStat$sumstats.meta[iTrait,]$l_ldimp_reinflation^(1/3))*SE.I))^2]

    p$dsStat$sumstats.meta[iTrait,c("meanAssocChi2.imputed.re")]<-mean(cSumstats[is.finite(ZZ.re),]$ZZ.re,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("meanAssocChi2.imputed.re2")]<-mean(cSumstats[is.finite(ZZ.re2),]$ZZ.re2,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("meanAssocChi2.imputed.re3")]<-mean(cSumstats[is.finite(ZZ.re3),]$ZZ.re3,na.rm=T)

    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed.re")]<-median(cSumstats[is.finite(ZZ.re),]$ZZ.re,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed.re2")]<-median(cSumstats[is.finite(ZZ.re2),]$ZZ.re2,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("medianAssocChi2.imputed.re3")]<-median(cSumstats[is.finite(ZZ.re3),]$ZZ.re3,na.rm=T)

    p$dsStat$sumstats.meta[iTrait,c("m")]<-nrow(cSumstats)
    p$dsStat$sumstats.meta[iTrait,c("m.genotyped")]<-nrow(cSumstats[IMPUTED==F,])
    p$dsStat$sumstats.meta[iTrait,c("m.imputed")]<-nrow(cSumstats[IMPUTED==T,])
    p$dsStat$sumstats.meta[iTrait,c("m.05_50")]<-nrow(cSumstats[MAF>0.05 & MAF<0.5,])
     p$dsStat$sumstats.meta[iTrait,c("m.genotyped.05_50")]<-nrow(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,])
    p$dsStat$sumstats.meta[iTrait,c("m.imputed.05_50")]<-nrow(cSumstats[IMPUTED==T & MAF>0.05 & MAF<0.5,])
    p$dsStat$sumstats.meta[iTrait,c("m.genotyped.01_05")]<-nrow(cSumstats[IMPUTED==F & MAF>0.01 & MAF<0.05,])
    p$dsStat$sumstats.meta[iTrait,c("m.imputed.01_05")]<-nrow(cSumstats[IMPUTED==T & MAF>0.01 & MAF<0.05,])
    p$dsStat$sumstats.meta[iTrait,c("m.genotyped.0_01")]<-nrow(cSumstats[IMPUTED==F & MAF<0.01,])
    p$dsStat$sumstats.meta[iTrait,c("m.imputed.0_01")]<-nrow(cSumstats[IMPUTED==T & MAF<0.01,])

    p$dsStat$sumstats.meta[iTrait,c("rmse_z")]<-sqrt(mean(cSumstats[IMPUTED==F,]$d.z^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re")]<-sqrt(mean(cSumstats[IMPUTED==F,]$d.z.re^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re2")]<-sqrt(mean(cSumstats[IMPUTED==F,]$d.z.re2^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re3")]<-sqrt(mean(cSumstats[IMPUTED==F,]$d.z.re3^2,na.rm=T))

    #p$dsStat$sumstats.meta[iTrait,c("rmse_beta")]<-sqrt(mean(cSumstats$d.beta^2,na.rm=T))
    #p$dsStat$sumstats.meta[iTrait,c("rmse_se")]<-sqrt(mean(cSumstats$d.se^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z.05_50")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$d.z^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z.01_05")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF>0.01 & MAF<0.05,]$d.z^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z.0_01")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF<0.01,]$d.z^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re2.05_50")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$d.z.re2^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re2.01_05")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF>0.01 & MAF<0.05,]$d.z.re2^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re2.0_01")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF<0.01,]$d.z.re2^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re3.05_50")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$d.z.re3^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re3.01_05")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF>0.01 & MAF<0.05,]$d.z.re3^2,na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_z_re3.0_01")]<-sqrt(mean(cSumstats[IMPUTED==F & MAF<0.01,]$d.z.re3^2,na.rm=T))

    # cQ<-quantile(sqrt(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$ZZ),0.5)
    # p$dsStat$sumstats.meta[iTrait,c("rmse_z.q_50")]<-sqrt(mean(cSumstats[IMPUTED==F & sqrt(ZZ)<eval(cQ),]$d.z^2,na.rm=T))
    # cQ2<-quantile(sqrt(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$ZZ),0.75)
    # p$dsStat$sumstats.meta[iTrait,c("rmse_z.q_50_75")]<-sqrt(mean(cSumstats[IMPUTED==F & sqrt(ZZ)<eval(cQ2) & sqrt(ZZ)>eval(cQ),]$d.z^2,na.rm=T))
    # p$dsStat$sumstats.meta[iTrait,c("rmse_z.q_75_100")]<-sqrt(mean(cSumstats[IMPUTED==F & sqrt(ZZ)>eval(cQ2),]$d.z^2,na.rm=T))


    p$dsStat$sumstats.meta[iTrait,c("bias_z")]<-mean(cSumstats[IMPUTED==F,]$d.z,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re")]<-mean(cSumstats[IMPUTED==F,]$d.z.re,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re2")]<-mean(cSumstats[IMPUTED==F,]$d.z.re2,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re3")]<-mean(cSumstats[IMPUTED==F,]$d.z.re3,na.rm=T)
    #p$dsStat$sumstats.meta[iTrait,c("bias_beta")]<-mean(cSumstats$d.beta,na.rm=T)
    #p$dsStat$sumstats.meta[iTrait,c("bias_se")]<-mean(cSumstats$d.se,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z.05_50")]<-mean(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$d.z,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z.01_05")]<-mean(cSumstats[IMPUTED==F & MAF>0.01 & MAF<0.05,]$d.z,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z.0_01")]<-mean(cSumstats[IMPUTED==F & MAF<0.01,]$d.z,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re2.05_50")]<-mean(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$d.z.re2,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re2.01_05")]<-mean(cSumstats[IMPUTED==F & MAF>0.01 & MAF<0.05,]$d.z.re2,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re2.0_01")]<-mean(cSumstats[IMPUTED==F & MAF<0.01,]$d.z.re2,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re3.05_50")]<-mean(cSumstats[IMPUTED==F & MAF>0.05 & MAF<0.5,]$d.z.re3,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re3.01_05")]<-mean(cSumstats[IMPUTED==F & MAF>0.01 & MAF<0.05,]$d.z.re3,na.rm=T)
    p$dsStat$sumstats.meta[iTrait,c("bias_z_re3.0_01")]<-mean(cSumstats[IMPUTED==F & MAF<0.01,]$d.z.re3,na.rm=T)


   # test.result <-cor.test(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO),]$d.z.abs,cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO),]$SINFO)
    #p-values are expected to be 0 always due to large sample size.
    p$dsStat$sumstats.meta[iTrait,c("cor_sinfo_adz")]<-cor(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO),]$d.z.abs,cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO),]$SINFO)
    p$dsStat$sumstats.meta[iTrait,c("cor_sinfo_adz.05_50")]<-cor(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & MAF>0.05 & MAF<0.5,]$d.z.abs,cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & MAF>0.05 & MAF<0.5,]$SINFO)
    p$dsStat$sumstats.meta[iTrait,c("cor_sinfo_adz.01_05")]<-cor(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & MAF>0.01 & MAF<0.05,]$d.z.abs,cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & MAF>0.01 & MAF<0.05,]$SINFO)
    p$dsStat$sumstats.meta[iTrait,c("cor_sinfo_adz.0_01")]<-cor(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & MAF<0.01,]$d.z.abs,cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & MAF<0.01,]$SINFO)

    p$dsStat$sumstats.meta[iTrait,c("rmse_sinfo.0_5")]<-sqrt(mean(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & SINFO<0.5,]$d.z^2, na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_sinfo.5_9")]<-sqrt(mean(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & SINFO>0.5 & SINFO<0.9,]$d.z^2, na.rm=T))
    p$dsStat$sumstats.meta[iTrait,c("rmse_sinfo.9_10")]<-sqrt(mean(cSumstats[IMPUTED==F & is.finite(d.z.abs) & is.finite(SINFO) & SINFO>0.9 & SINFO<1,]$d.z^2, na.rm=T))




    saveRDS(object = p$dsStat,file = p$filepath.dsStat)
  }
  
  #paths
  
  p$dsStat$panel.meta<-as.data.table(matrix(data = NA,nrow = 0,ncol = 0))
  p$dsStat$panel.meta<- rbind(p$dsStat$panel.meta,list(code="hm3",path=eval(p$filepath.SNPReference.hm3)))
  p$dsStat$panel.meta<- rbind(p$dsStat$panel.meta,list(code="1kg",path=eval(p$filepath.SNPReference.1kg)))
  p$dsStat$panel.meta<- rbind(p$dsStat$panel.meta,list(code="hc1kg",path=eval(p$filepath.SNPReference.hc1kg)))
  
  print(p$dsStat$panel.meta)
  
  #hard coded
  cat("\n1kg")
  if(file.exists(file.path(p$folderpath.workingDirectory,"1kg.frq.l2.Rds"))) {
    print("Using existing 1kg dataset")
    tgp.old<-readRDS(file=file.path(p$folderpath.workingDirectory,"1kg.frq.l2.Rds"))
  } else {
    tgp.old <- shru::readFile(filePath = file.path("/scratch/prj/gwas_sumstats/1kg_bed","1KG_Phase3.WG.CLEANED.EUR_MAF001.CM23.bim"), nThreads = p$nThread)
    colnames(tgp.old)<-c("CHR","SNP","CM","BP","A1","A2")
    tgp.old.frq <- shru::readFile(filePath = file.path("/scratch/prj/gwas_sumstats/1kg_bed","1KG_Phase3.WG.CLEANED.EUR_MAF001.CM23.frq.frq"), nThreads = p$nThread)
    setkeyv(tgp.old,cols = c("SNP"))
    setkeyv(tgp.old.frq,cols = c("SNP"))
    tgp.old[tgp.old.frq,on=c("SNP"),c("MAF"):=i.MAF]
    tgp.old.l2 <- shru::readFile(filePath = file.path("/scratch/prj/gwas_sumstats/ld_scores/1KG_Phase3.WG.CLEANED.EUR_MAF001.1cm.250blocks","1KG_Phase3.WG.CLEANED.EUR_MAF001.1cm.250blocks.l2.ldscore.gz"), nThreads = p$nThread)
    setkeyv(tgp.old.l2,cols = c("SNP"))
    tgp.old[tgp.old.l2,on=c("SNP"),c("L2"):=i.L2]
    saveRDS(object = tgp.old,file = file.path(p$folderpath.workingDirectory,"1kg.frq.l2.Rds"))
  }
  
  #save as LDSC compatible(?) variant list
  #shru::writeFile(d = tgp.old,filePath = file.path(p$folderpath.data.variantLists,"1kgp3.bX.eur.l2.jz2023.gz"),nThreads = p$nThread)
  
  tgp.old[,g:="1kg"]
  
  p$dsStat$panel.meta[code=="1kg",c("m")]<-nrow(tgp.old)
  p$dsStat$panel.meta[code=="1kg",c("m.0_01")]<-nrow(tgp.old[is.finite(MAF) & MAF<0.01,])
  p$dsStat$panel.meta[code=="1kg",c("m.01_05")]<-nrow(tgp.old[is.finite(MAF) & MAF>0.01 & MAF <0.05,])
  p$dsStat$panel.meta[code=="1kg",c("m.05_50")]<-nrow(tgp.old[is.finite(MAF) & MAF>0.05 & MAF <0.5,])
  #rm(tgp.old)
  
  cat("\nhm3")
  if(file.exists(file.path(p$folderpath.workingDirectory,"hm3.frq.l2.Rds"))) {
    print("Using existing hm3 dataset")
    hm3<-readRDS(file=file.path(p$folderpath.workingDirectory,"hm3.frq.l2.Rds"))
  } else {
    hm3 <- shru::readFile(filePath = p$dsStat$panel.meta[code=="hm3",]$path, nThreads = p$nThread)
    setkeyv(hm3,cols = c("SNP"))
    hm3[tgp.old.frq,on=c("SNP"),c("MAF"):=i.MAF]
    hm3.l2 <- do.call("rbind", lapply(list.files(path = "/scratch/prj/gwas_sumstats/ld_scores/eur_w_ld_chr", pattern = "\\.l2\\.ldscore\\.gz$"), function(i) {
          suppressMessages(fread(file = file.path("/scratch/prj/gwas_sumstats/ld_scores/eur_w_ld_chr", i), na.strings =c(".",NA,"NA",""), encoding = "UTF-8",check.names = T, fill = T, blank.lines.skip = T, showProgress = F, nThread = 6, data.table=T))
        }))
    setkeyv(hm3.l2,cols = c("SNP"))
    hm3[hm3.l2,on=c("SNP"),c("MAF"):=i.MAF][hm3.l2,on=c("SNP"),c("L2"):=i.L2]
    saveRDS(object = hm3,file = file.path(p$folderpath.workingDirectory,"hm3.frq.l2.Rds"))
  }
  
  hm3[,g:="hm3"]
  
  p$dsStat$panel.meta[code=="hm3",c("m")]<-nrow(hm3)
  p$dsStat$panel.meta[code=="hm3",c("m.0_01")]<-nrow(hm3[is.finite(MAF) & MAF<0.01,])
  p$dsStat$panel.meta[code=="hm3",c("m.01_05")]<-nrow(hm3[is.finite(MAF) & MAF>0.01 & MAF <0.05,])
  p$dsStat$panel.meta[code=="hm3",c("m.05_50")]<-nrow(hm3[is.finite(MAF) & MAF>0.05 & MAF <0.5,])
  
  vall<-rbindlist(
    l = list(tgp.old[,c("SNP","MAF","L2","g")],hm3[,c("SNP","MAF","L2","g")]),
    use.names = T,
    fill = T
    )
  rm(hm3)
  rm(tgp.old.frq)
  rm(tgp.old)
  
  
  cat("\nhc1kg")
  tgp <- shru::readFile(filePath = p$dsStat$panel.meta[code=="hc1kg",]$path, nThreads = p$nThread)
  setkeyv(tgp,cols = c("SNP"))
  
  tgp[,g:="hc1kg"]
  
  p$dsStat$panel.meta[code=="hc1kg",c("m")]<-nrow(tgp)
  p$dsStat$panel.meta[code=="hc1kg",c("m.0_01")]<-nrow(tgp[is.finite(MAF) & MAF<0.01,])
  p$dsStat$panel.meta[code=="hc1kg",c("m.01_05")]<-nrow(tgp[is.finite(MAF) & MAF>0.01 & MAF <0.05,])
  p$dsStat$panel.meta[code=="hc1kg",c("m.05_50")]<-nrow(tgp[is.finite(MAF) & MAF>0.05 & MAF <0.5,])

  saveRDS(object = p$dsStat,file = p$filepath.dsStat)
  
  vall<-rbindlist(
    l = list(vall,tgp[,c("SNP","MAF","L2","g")]),
    use.names = T,
    fill = T
    )
  rm(tgp)
  
  
  if(p$host!="cluster"){
    vall<-vall[MAF>0 & !is.na(L2),]
    vall<-as.data.frame(vall)
    #vall$SNP<-as.factor(vall$SNP)
    vall$g<-as.factor(vall$g)
    
    for(cPanel in c("hm3","1kg","hc1kg")){
      #cPanel <- "hm3"
      cPlot <- ggplot(aes(x=MAF, group = g, fill = g), data = vall[vall$g==cPanel,]) + #aes: group = g, fill = g
        geom_bar(stat = 'count', position = 'dodge') +
        scale_x_binned(n.breaks = 50) +
        #ylim(0,1.5e07) +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
        ggtitle(paste("MAF distribution: ",cPanel)) +
        xlab("MAF") + ylab("# variants")
      ggsave(filename=file.path(p$folderpath.plots, paste0("panel.maf.",cPanel,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1.2)
      
      cPlot <- ggplot(aes(x=MAF, y=L2, group = g, fill = g), data = vall[vall$g==cPanel,]) +
        stat_summary_bin(geom='point', size=3, shape=5, color='red', bins=60) +
        #stat_summary_bin(fun.y = "mean", geom="bar", bins=30) +
        #geom_errorbar(aes(ymin= len-se, ymax=len+se),
         #         width=.2,                    # Width of the error bars
          #        position=position_dodge(.9)) +
        stat_summary_bin(geom='errorbar', color='red', bins=60, width=0.1) +
        #stat_summary_bin(fun.y = "mean", geom="bar", bins=30) +
        ylim(0,150) +
        
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
        ggtitle(paste("L2 distribution across MAF: ",cPanel)) +
        xlab("MAF") + ylab("Average L2")
      ggsave(filename=file.path(p$folderpath.plots, paste0("panel.l2_maf.",cPanel,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1.2)
    }
  }
  
}

if(p$clOptions$task=="stat") quit(save = "no")


```

# Preliminary annotation of chosen datasets, excluding mean chi2

```{r annotation of chosen datasets, fig.width=10, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}
#View(p$sumstats.sel)
#View(p$sumstats.sel.raw)

#selected sumstats
p$sumstats.sel.table<-p$sumstats.sel.raw[,c("name.nice.and_code","reference.nice","year","n_case","n_control", "samplePrevalence","populationPrevalence")]
#View(p$sumstats.sel.table)
#p$sumstats.sel.table

p$plots.sumstats.sel.table<-p$sumstats.sel.table %>% 
  gt() %>% 
  fmt_number(columns = vars(samplePrevalence), decimals = 3) %>%
  fmt_number(columns = vars(populationPrevalence), decimals = 4) %>%
  fmt_number(columns = vars(n_case,n_control), decimals = 0) %>%
  tab_header(
    title = "Selected GWAS summary statistics datasets"
  ) %>% cols_label(
    name.nice.and_code  = "Trait",
    reference.nice = "Reference",
    year = "Year",
    #ancestry = "Ancestry",
    #sex = "Sex",
    n_case = "N case or total",
    n_control = "N control",
    samplePrevalence = html("Prev<sub>sample</sub>"),
    populationPrevalence = html("Prev<sub>population</sub>")
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.rtf"))


```

# LD score regression

Original Python 2 LDSC for heritabilities of selected traits.
```{r ld score regression}
print("***Univariate LD score regression***")

p$filepath.LD<-file.path(p$folderpath.workingDirectory,paste0("LD.",p$setup.code,".Rds"))

if (file.exists(p$filepath.LD)) {
  print("Using existing covariance structures from previous uv LD score regression computations.")
  p$LD<-readRDS(file=p$filepath.LD)
} else {
  p$LD<-c()
}

if(p$clOptions$task=="LD")
{
  
  # p$LD$traits.hm3.plainN<-c()
  # for(cTrait in p$sumstats.sel.ssimp.code){
  #   #cTrait<-"ANXI03"
  #   p$LD$traits.hm3.plainN[[cTrait]] <- list(
  #     ldsc.orig(trait_codes = list(cTrait),trait_filepaths = p$sumstats.sel.ssimp[cTrait,]$mungedpath.ldsc.hm3.plainN, ldsc_filepath = "/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py", ld_folderpath = p$folderpath.data.mvLDSC.ld.hm3,samp_prev = p$sumstats.sel.ssimp[cTrait,]$samplePrevalence, pop_prev = p$sumstats.sel.ssimp[cTrait,]$populationPrevalence,output_file_path = paste0(cTrait,"_hm3.plainN.ldsc"), blocks = 200)
  #     )
  # }
  # saveRDS(object = p$LD,file = p$filepath.LD)
  # 
  # p$LD$traits.hm3<-c()
  # for(cTrait in p$sumstats.sel.ssimp.code){
  #   #cTrait<-"ANXI03"
  #   p$LD$traits.hm3[[cTrait]] <- list(
  #     ldsc.orig(trait_codes = list(cTrait),trait_filepaths = p$sumstats.sel.ssimp[cTrait,]$mungedpath.ldsc.hm3, ldsc_filepath = "/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py", ld_folderpath = p$folderpath.data.mvLDSC.ld.hm3,samp_prev = ifelse(!is.na(p$sumstats.sel.ssimp[cTrait,]$samplePrevalence),0.5, NA_real_), pop_prev = p$sumstats.sel.ssimp[cTrait,]$populationPrevalence,output_file_path = paste0(cTrait,"_hm3.ldsc"), blocks = 200)
  #     )
  # }
  # saveRDS(object = p$LD,file = p$filepath.LD)
  
  p$LD$traits.1kg.maf0_01<-c()
  for(cTrait in p$sumstats.sel.ssimp.code){
    #cTrait<-"ANXI03"
    p$LD$traits.1kg.maf0_01[cTrait] <- list(
      ldsc.orig(trait_codes = list(cTrait),trait_filepaths = p$sumstats.sel.ssimp[cTrait,]$mungedpath.ldsc.1kg.orig.maf0_01, ldsc_filepath = "/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py", ld_filepath = file.path(p$folderpath.data.mvLDSC.ld.1kg,"1KG_Phase3.WG.CLEANED.EUR_MAF001.1cm.250blocks"),samp_prev = ifelse(!is.na(p$sumstats.sel.ssimp[cTrait,]$samplePrevalence),0.5, NA_real_), pop_prev = p$sumstats.sel.ssimp[cTrait,]$populationPrevalence,output_file_path = paste0(cTrait,"_1kg.orig.maf0_01.ldsc"), blocks = 200)
      )
  }
  saveRDS(object = p$LD,file = p$filepath.LD)
  
  # p$LD$traits.1kg.maf0_05<-c()
  # for(cTrait in p$sumstats.sel.ssimp.code){
  #   #cTrait<-"ANXI03"
  #   p$LD$traits.1kg.maf0_05[cTrait] <- list(
  #     ldsc.orig(trait_codes = list(cTrait),trait_filepaths = p$sumstats.sel.ssimp[cTrait,]$mungedpath.ldsc.1kg.orig.maf0_05, ldsc_filepath = "/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py", ld_filepath = file.path(p$folderpath.data.mvLDSC.ld.1kg,"1KG_Phase3.WG.CLEANED.EUR_MAF001.1cm.250blocks"),samp_prev = ifelse(!is.na(p$sumstats.sel.ssimp[cTrait,]$samplePrevalence),0.5, NA_real_), pop_prev = p$sumstats.sel.ssimp[cTrait,]$populationPrevalence,output_file_path = paste0(cTrait,"_1kg.orig.maf0_05.ldsc"), blocks = 200)
  #     )
  # }
  # saveRDS(object = p$LD,file = p$filepath.LD)
  # 
  # p$LD$traits.1kg.maf0_1<-c()
  # for(cTrait in p$sumstats.sel.ssimp.code){
  #   #cTrait<-"ANXI03"
  #   p$LD$traits.1kg.maf0_1[cTrait] <- list(
  #     ldsc.orig(trait_codes = list(cTrait),trait_filepaths = p$sumstats.sel.ssimp[cTrait,]$mungedpath.ldsc.1kg.orig.maf0_1, ldsc_filepath = "/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py", ld_filepath = file.path(p$folderpath.data.mvLDSC.ld.1kg,"1KG_Phase3.WG.CLEANED.EUR_MAF001.1cm.250blocks"),samp_prev = ifelse(!is.na(p$sumstats.sel.ssimp[cTrait,]$samplePrevalence),0.5, NA_real_), pop_prev = p$sumstats.sel.ssimp[cTrait,]$populationPrevalence,output_file_path = paste0(cTrait,"_1kg.orig.maf0_1.ldsc"), blocks = 200)
  #     )
  # }
  # saveRDS(object = p$LD,file = p$filepath.LD)
  # 
  # p$LD$traits.hc1kg.maf0_05<-c()
  # for(cTrait in p$sumstats.sel.ssimp.code){
  #   #cTrait<-"ANXI03"
  #   p$LD$traits.hc1kg.maf0_05[cTrait] <- list(
  #     ldsc.orig(trait_codes = list(cTrait),trait_filepaths = p$sumstats.sel.ssimp[cTrait,]$mungedpath.ldsc.1kg.maf0_05, ldsc_filepath = "/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py", ld_folderpath = p$folderpath.data.mvLDSC.ld.hc1kg, samp_prev = ifelse(!is.na(p$sumstats.sel.ssimp[cTrait,]$samplePrevalence),0.5, NA_real_), pop_prev = p$sumstats.sel.ssimp[cTrait,]$populationPrevalence,output_file_path = paste0(cTrait,"_1kg..maf0_05.ldsc"), blocks = 200)
  #     )
  # }
  # saveRDS(object = p$LD,file = p$filepath.LD)
  # 
  # p$LD$traits.hc1kg.maf0_1<-c()
  # for(cTrait in p$sumstats.sel.ssimp.code){
  #   #cTrait<-"ANXI03"
  #   p$LD$traits.hc1kg.maf0_1[cTrait] <- list(
  #     ldsc.orig(trait_codes = list(cTrait),trait_filepaths = p$sumstats.sel.ssimp[cTrait,]$mungedpath.ldsc.1kg.maf0_1, ldsc_filepath = "/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py", ld_folderpath = p$folderpath.data.mvLDSC.ld.hc1kg, samp_prev = ifelse(!is.na(p$sumstats.sel.ssimp[cTrait,]$samplePrevalence),0.5, NA_real_), pop_prev = p$sumstats.sel.ssimp[cTrait,]$populationPrevalence,output_file_path = paste0(cTrait,"_1kg..maf0_05.ldsc"), blocks = 200)
  #     )
  # }
  # saveRDS(object = p$LD,file = p$filepath.LD)

}

if(length(p$LD$traits.hm3.plainN)>0){
  p$sumstats$h2.LDSC.hm3.plainN<-NA_real_
  p$sumstats$h2.se.LDSC.hm3.plainN<-NA_real_
  for(cCode in p$sumstats.sel.ssimp.code){
    #cCode<-"ANXI03"
    cRes<-p$LD$traits.hm3.plainN[[cCode]]
    p$sumstats[cCode,c("h2.LDSC.hm3.plainN")]<-cRes[[1]]$df$h2_liab[[1]]
    p$sumstats[cCode,c("h2.se.LDSC.hm3.plainN")]<-cRes[[1]]$df$h2_liab_se[[1]]
  }
}

if(length(p$LD$traits.hm3)>0){
  p$sumstats$h2.LDSC.hm3<-NA_real_
  p$sumstats$h2.se.LDSC.hm3<-NA_real_
  for(cCode in p$sumstats.sel.ssimp.code){
    #cCode<-"ANXI03"
    cRes<-p$LD$traits.hm3[[cCode]]
    p$sumstats[cCode,c("h2.LDSC.hm3")]<-cRes[[1]]$df$h2_liab[[1]]
    p$sumstats[cCode,c("h2.se.LDSC.hm3")]<-cRes[[1]]$df$h2_liab_se[[1]]
  }
}

if(length(p$LD$traits.1kg.maf0_01)>0){
  p$sumstats$h2.LDSC.1kg.maf0_01<-NA_real_
  p$sumstats$h2.se.LDSC.1kg.maf0_01<-NA_real_
  for(cCode in p$sumstats.sel.ssimp.code){
    #cCode<-"ANXI03"
    cRes<-p$LD$traits.1kg.maf0_01[[cCode]]
    p$sumstats[cCode,c("h2.LDSC.1kg.maf0_01")]<-cRes$df$h2_liab[[1]] #I do not know why these are different
    p$sumstats[cCode,c("h2.se.LDSC.1kg.maf0_01")]<-cRes$df$h2_liab_se[[1]]
  }
}

if(length(p$LD$traits.1kg.maf0_05)>0){
  p$sumstats$h2.LDSC.1kg.maf0_05<-NA_real_
  p$sumstats$h2.se.LDSC.1kg.maf0_05<-NA_real_
  for(cCode in p$sumstats.sel.ssimp.code){
    #cCode<-"ANXI03"
    cRes<-p$LD$traits.1kg.maf0_05[[cCode]]
    p$sumstats[cCode,c("h2.LDSC.1kg.maf0_05")]<-cRes$df$h2_liab[[1]] #I do not know why these are different
    p$sumstats[cCode,c("h2.se.LDSC.1kg.maf0_05")]<-cRes$df$h2_liab_se[[1]]
  }
}

if(length(p$LD$traits.1kg.maf0_1)>0){
  p$sumstats$h2.LDSC.1kg.maf0_1<-NA_real_
  p$sumstats$h2.se.LDSC.1kg.maf0_1<-NA_real_
  for(cCode in p$sumstats.sel.ssimp.code){
    #cCode<-"ANXI03"
    cRes<-p$LD$traits.1kg.maf0_1[[cCode]]
    p$sumstats[cCode,c("h2.LDSC.1kg.maf0_1")]<-cRes$df$h2_liab[[1]] #I do not know why these are different
    p$sumstats[cCode,c("h2.se.LDSC.1kg.maf0_1")]<-cRes$df$h2_liab_se[[1]]
  }
}

if(length(p$LD$traits.hc1kg.maf0_05)>0){
  p$sumstats$h2.LDSC.hc1kg.maf0_05<-NA_real_
  p$sumstats$h2.se.LDSC.hc1kg.maf0_05<-NA_real_
  for(cCode in p$sumstats.sel.ssimp.code){
    #cCode<-"ANXI03"
    cRes<-p$LD$traits.hc1kg.maf0_05[[cCode]]
    p$sumstats[cCode,c("h2.LDSC.hc1kg.maf0_05")]<-cRes$df$h2_liab[[1]] #I do not know why these are different
    p$sumstats[cCode,c("h2.se.LDSC.hc1kg.maf0_05")]<-cRes$df$h2_liab_se[[1]]
  }
}

if(length(p$LD$traits.hc1kg.maf0_1)>0){
  p$sumstats$h2.LDSC.hc1kg.maf0_1<-NA_real_
  p$sumstats$h2.se.LDSC.hc1kg.maf0_1<-NA_real_
  for(cCode in p$sumstats.sel.ssimp.code){
    #cCode<-"ANXI03"
    cRes<-p$LD$traits.hc1kg.maf0_1[[cCode]]
    p$sumstats[cCode,c("h2.LDSC.hc1kg.maf0_1")]<-cRes$df$h2_liab[[1]] #I do not know why these are different
    p$sumstats[cCode,c("h2.se.LDSC.hc1kg.maf0_1")]<-cRes$df$h2_liab_se[[1]]
  }
}

#View(p$sumstats[p$sumstats.sel.ssimp.code,])

```


# Multivariate LD score regression
The multivariate LDSC and HDL procedure calculates genetic correlations between the different traits and estimates of correlation standard errors, producing S (variances and covariances = heritabilities and coheritabilities) and V (diagonal: squared standard errors of elements in S. Off-diagonal: Covariances of elements in S - "dependencies between estimation errors used to model dependencies due to sample overlap") matrices used later in the Genomic SEM model fitting.

HDL is an improved method to account for linkage disequilibrium while estimating genetic effects from GWAS.

```{r multivariate LD score regression}
print("***Multivariate LD score regression***")

p$filepath.mvLD<-file.path(p$folderpath.workingDirectory,paste0("mvLD.",p$setup.code,".Rds"))

if (file.exists(p$filepath.mvLD)) {
  print("Using existing covariance structures from previous mv LD score regression computations.")
  p$mvLD<-readRDS(file=p$filepath.mvLD)
} else {
  p$mvLD<-c()
}

if(p$clOptions$task=="mvLD" || (!file.exists(p$filepath.mvLD) & p$clOptions$task=="0"))
{
  print("Running (multivariate) LD score regression with different methods. This might take a while. If the procedure runs for too long you may want to abort the process.")
  
  cat("The current task is specified as:",p$clOptions$task,"\n")
  
  #run mvLDSC
  
  #GSEM LDSC
  
  

  # cat("\n\n*** LDSC: covstruct.GSEMmvLDSC.hm3.sim ***\n") #this is the main branch implementation
  # p$mvLD$covstruct.GSEMmvLDSC.hm3.sim<-GenomicSEM::ldsc(
  #   traits = p$sumstats.sel.sim$mungedpath.ldsc.hm3,
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  #   n.blocks = 200,
  #   ldsc.log = p$setup.code.date
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.GSEMmvLDSC.hm3.sim ***\n")
  
  
  #  cat("\n\n*** LDSC: covstruct.GSEMmvLDSC.hm3.sim ***\n") #this is the MODIFIED main branch implementation, changed to remove duplicate variants over SNP - the seemingly main difference between the LDSC++ and original Genomic SEM implementations when running with original LDSC-settings (block jackknife, 200 fixed blocks etc.)
  # p$mvLD$covstruct.GSEMmvLDSC.hm3.sim<-ldsc.orig.mod(
  #   traits = p$sumstats.sel.sim$mungedpath.ldsc.hm3,
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  #   n.blocks = 200,
  #   ldsc.log = p$setup.code.date
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.GSEMmvLDSC.hm3.sim ***\n")


  
# 
#   cat("\n\n*** LDSC: covstruct.GSEMmvLDSC.hm3 ***\n")
#   p$mvLD$covstruct.GSEMmvLDSC.hm3<-GenomicSEM::ldsc(
#     traits = p$sumstats.sel.ssimp$mungedpath.ldsc.hm3,
#     sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
#     population.prev = p$sumstats.sel.ssimp$populationPrevalence,
#     trait.names = p$sumstats.sel.ssimp$code,
#     ld = p$folderpath.data.mvLDSC.ld.hm3,
#     n.blocks = 200,
#     ldsc.log = p$setup.code.date
#     )
#   saveRDS(object = p$mvLD,file = p$filepath.mvLD)
#   cat("\n\n***END LDSC: covstruct.GSEMmvLDSC.hm3 ***\n")
# 
# 
#   cat("\n\n*** LDSC: covstruct.mvLDSC.GSEMemulation.hm3 ***\n")
#   p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3<-ldscpp(
#     #traits = p$sumstats.sel.ssimp$mungedpath.ldsc.hm3,
#     traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #do we use the ldsc=munged or the unfiltered supermunged?
#     sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
#     population.prev = p$sumstats.sel.ssimp$populationPrevalence,
#     trait.names = p$sumstats.sel.ssimp$code,
#     ld = p$folderpath.data.mvLDSC.ld.hm3,
#     n.blocks = 200,
#     ldsc.log = p$setup.code.date,
#     preweight.alternativeCorrelationCorrection = F,
#     preweight.ChiSquare = F,
#     correctAttenuationBias = F,
#     doubleRegressionRoutine = F,
#     preweight.INFO=F,
#     resamplingMethod="jn",
#     blocksizeCM = NA, #inactivate CM window def
#     filter.info = 0.6,
#     filter.maf = 0.01,
#     filter.zz.min = 0
#     )
#   saveRDS(object = p$mvLD,file = p$filepath.mvLD)
#   cat("\n\n***END LDSC: covstruct.mvLDSC.GSEMemulation.hm3 ***\n")

  # cat("\n\n*** LDSC: covstruct.mvLDSC.GSEMemulation.hm3.sim ***\n")
  # p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3.sim<-ldscpp(
  #   traits = p$sumstats.sel.sim$mungedpath.ldsc.hm3,
  #   #traits = p$sumstats.sel.sim$mungedpath.supermunge.hm3.unfiltered, #do we use the ldsc=munged or the unfiltered supermunged?
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  #   n.blocks = 200,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.GSEMemulation.hm3.sim ***\n")
  
# 
# 
#   cat("\n\n*** LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_01 ***\n")
#   p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01<-ldscpp(
#     traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #USES THE UNFILTERED FOLDER BECAUSE OF 0.01
#     sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
#     population.prev = p$sumstats.sel.ssimp$populationPrevalence,
#     trait.names = p$sumstats.sel.ssimp$code,
#     ld = p$folderpath.data.mvLDSC.ld.1kg,
#     n.blocks = 200,
#     ldsc.log = p$setup.code.date,
#     preweight.alternativeCorrelationCorrection = F,
#     preweight.ChiSquare = F,
#     correctAttenuationBias = F,
#     doubleRegressionRoutine = F,
#     preweight.INFO=F,
#     resamplingMethod="jn",
#     blocksizeCM = NA, #inactivate CM window def
#     filter.info = 0.6,
#     filter.maf = 0.01,
#     filter.zz.min = 0,
#     force.M = 7184778
#     )
#   saveRDS(object = p$mvLD,file = p$filepath.mvLD)
#   cat("\n\n***END LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_01 ***\n")

  # cat("\n\n*** LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_01.sim ***\n")
  # p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01.sim<-ldscpp(
  #   traits = p$sumstats.sel.sim$mungedpath.supermunge.1kg.orig.unfiltered, #USES THE UNFILTERED FOLDER BECAUSE OF 0.01
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   n.blocks = 200,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_01.sim ***\n")

  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_05 ***\n")
  # p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_05<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.ldsc.1kg.orig.maf0_05,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   n.blocks = 200,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.05,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_05 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_1 ***\n")
  # p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_1<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.ldsc.1kg.orig.maf0_1,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   n.blocks = 200,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.1,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.GSEMemulation.1kg.maf0_1 ***\n")
  # 
  # 
  # #restricted 1kg to hm3 variants
  # df_ld.hm3 <- fread(file = p$filepath.SNPReference.hm3, na.strings =c(".",NA,"NA",""), encoding = "UTF-8",check.names = T, fill = T, blank.lines.skip = T, showProgress = F, nThread = p$nThread, data.table=T)
  # df_ld.1kg <- do.call("rbind", lapply(list.files(path = p$folderpath.data.mvLDSC.ld.1kg, pattern = "\\.l2\\.ldscore\\.gz$"), function(i) {
  #     suppressMessages(fread(file = file.path(p$folderpath.data.mvLDSC.ld.1kg, i), na.strings =c(".",NA,"NA",""), encoding = "UTF-8",check.names = T, fill = T, blank.lines.skip = T, showProgress = F, nThread = p$nThread, data.table=T))
  #   }))
  # 
  # df_ld.hm3<-df_ld.hm3[,c("SNP")]
  # setkeyv(df_ld.hm3,cols = c("SNP"))
  # setkeyv(df_ld.1kg,cols = c("SNP"))
  # df_ld.1kg.hm3res <- df_ld.1kg[df_ld.hm3, on=c("SNP"), nomatch=0]
  # rm(df_ld.hm3)
  # rm(df_ld.1kg)
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.GSEMemulation.1kg.hm3res***\n")
  # p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.hm3res<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   df_ld = df_ld.1kg.hm3res,
  #   n.blocks = 200,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.GSEMemulation.1kg.hm3res***\n")
  # #rm(df_ld.1kg.hm3res)

   #calculate common LDSC block-size for vbcs default settings
  #mean(diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_05$m.nonimputed))/200
  # This is 21782.13 ~ 20K

  #jn

  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.varblock ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.varblock<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.varblock ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock ***\n")
  # 
  # #View(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S.SE)
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_8_10 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_8_10<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.8,
  #   filter.info.lt = 1.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_8_10 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_6_8 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_6_8<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.info.lt = 0.8,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_6_8 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_4_6 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_4_6<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.4,
  #   filter.info.lt = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_4_6 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_0_4 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_0_4<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0,
  #   filter.info.lt = 0.4,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_0_4 ***\n")
  # 
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.rblock ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.rblock<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   #blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.rblock ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   #blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock ***\n")
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock5 ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock5 ***\n")
  # 
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.varblock.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.varblock.altcw ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.altcw ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.hm3.rblock.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.rblock.altcw ***\n")
  # 
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   #blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock.altcw ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock5.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock5.altcw ***\n")
  # 
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.rblock.unfiltered ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.unfiltered<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.rblock.unfiltered ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock.unfiltered ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.unfiltered<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="jn",
  #   filter.maf = 0.05,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock.unfiltered ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.varblock.winfo ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.varblock.winfo ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo ***\n")
  # 
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_8_10 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_8_10<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.8,
  #   filter.info.lt = 1.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_8_10 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_6_8 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_6_8<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.info.lt = 0.8, #filters genotyped variants
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_6_8 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_4_6 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_4_6<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.4,
  #   filter.info.lt = 0.6, #filters genotyped variants
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_4_6 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_0_4 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_0_4<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.info.lt = 0.4, #filters genotyped variants
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo_info_0_4 ***\n")
  # 
  # 
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.rblock.winfo ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.rblock.winfo ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock.winfo ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock.winfo ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock5.winfo ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.winfo<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock5.winfo ***\n")


  #vbcs

  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.varblock ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.varblock ***\n")
  # 
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock ***\n")
  # 
  # #View(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S.SE)
  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock_info_8_10 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_8_10<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.8,
  #   filter.info.lt = 1.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock_info_8_10 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock_info_6_8 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_6_8<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.info.lt = 0.8,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock_info_6_8 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock_info_4_6 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_4_6<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.4,
  #   filter.info.lt = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock_info_4_6 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock_info_0_4 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_0_4<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0,
  #   filter.info.lt = 0.4,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock_info_0_4 ***\n")


  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.rblock ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.rblock ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   #blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock ***\n")
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5 ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5 ***\n")
  # 
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.varblock.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.varblock.altcw ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.altcw ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.altcw ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   #blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.altcw ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.6,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5.altcw ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.unfiltered ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.unfiltered<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered,
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=F,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.unfiltered ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.unfiltered ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.unfiltered<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered,
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=F,
  #   resamplingMethod="vbcs",
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.unfiltered ***\n")
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.varblock.winfo ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.varblock.winfo ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo ***\n")
  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_8_10 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_8_10<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.8,
  #   filter.info.lt = 1.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_8_10 ***\n")
  #
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_6_8 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_6_8<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.6,
  #   filter.info.lt = 0.8, #filters genotyped variants
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_6_8 ***\n")
  #
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_4_6 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_4_6<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.4,
  #   filter.info.lt = 0.6, #filters genotyped variants
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_4_6 ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_0_4 ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_0_4<-ldscpp(
  #   traits = p$sumstats.sel.infobin$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.infobin$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.infobin$populationPrevalence,
  #   trait.names = p$sumstats.sel.infobin$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.info.lt = 0.4, #filters genotyped variants
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_0_4 ***\n")


  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.winfo ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 20, #10 original for 1kG
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.winfo ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.winfo ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = F,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.winfo ***\n")
  # 
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5.winfo ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.winfo<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = F,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5.winfo ***\n")
  # 
  # 
  # 
  # #FINAL PANEL COMPARISONS
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw<-ldscpp(
  # #   traits =  p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw ***\n")
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw<-ldscpp(
  # #   traits =  p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="jn",
  # #   #blocksizeCM = 20, #10 original for 1kG #this may fail
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw ***\n")
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw<-ldscpp(
  # #   traits =  p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = NA, #inactivate CM window def
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw ***\n")
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw<-ldscpp(
  # #   traits =  p$sumstats.sel.ssimp$mungedpath.supermunge.hm3.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.hm3,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="vbcs",
  # #   #blocksizeCM = 20, #10 original for 1kG #this may fail
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw ***\n")
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw ***\n")
  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw.sim ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw.sim<-ldscpp(
  #   traits = p$sumstats.sel.sim$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw.sim ***\n")
  
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw ***\n")
  # 
  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw.sim ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw.sim<-ldscpp(
  #   traits = p$sumstats.sel.sim$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="jn",
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw.sim ***\n")
  
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.jn.rblock5.winfo.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.winfo.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="jn",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.jn.rblock5.winfo.altcw ***\n")
  # 
  # 
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw ***\n")
  # 
  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw.sim ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw.sim<-ldscpp(
  #   traits = p$sumstats.sel.sim$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   blocksizeCM = NA, #inactivate CM window def
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw.sim ***\n")
  
  cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw.sim.x1 ***\n")
  p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw.sim.x1<-ldscpp(
    traits = p$sumstats.sel.sim$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
    sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
    population.prev = p$sumstats.sel.sim$populationPrevalence,
    trait.names = p$sumstats.sel.sim$code,
    ld = p$folderpath.data.mvLDSC.ld.1kg,
    ldsc.log = p$setup.code.date,
    preweight.alternativeCorrelationCorrection = T,
    preweight.ChiSquare = F,
    correctAttenuationBias = F,
    doubleRegressionRoutine = T,
    preweight.INFO=T,
    resamplingMethod="vbcs",
    blocksizeCM = NA, #inactivate CM window def
    filter.info = 0.0,
    filter.maf = 0.01,
    filter.zz.min = 0,
    force.M = 7184778
    )
  saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw.sim.x1 ***\n")
  
  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw<-ldscpp(
  #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  #   trait.names = p$sumstats.sel.ssimp$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw ***\n")
  # 
  
  # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw.sim ***\n")
  # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw.sim<-ldscpp(
  #   traits = p$sumstats.sel.sim$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  #   sample.prev =  p$sumstats.sel.sim$samplePrevalence.balanced,
  #   population.prev = p$sumstats.sel.sim$populationPrevalence,
  #   trait.names = p$sumstats.sel.sim$code,
  #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  #   ldsc.log = p$setup.code.date,
  #   preweight.alternativeCorrelationCorrection = T,
  #   preweight.ChiSquare = F,
  #   correctAttenuationBias = F,
  #   doubleRegressionRoutine = F,
  #   preweight.INFO=T,
  #   resamplingMethod="vbcs",
  #   filter.info = 0.0,
  #   filter.maf = 0.01,
  #   filter.zz.min = 0,
  #   force.M = 7184778
  #   )
  # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw.sim ***\n")
  
  # 
  # # cat("\n\n*** LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5.winfo.altcw ***\n")
  # # p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.winfo.altcw<-ldscpp(
  # #   traits = p$sumstats.sel.ssimp$mungedpath.supermunge.1kg.orig.unfiltered, #we need all columns - INFO
  # #   sample.prev =  p$sumstats.sel.ssimp$samplePrevalence.balanced,
  # #   population.prev = p$sumstats.sel.ssimp$populationPrevalence,
  # #   trait.names = p$sumstats.sel.ssimp$code,
  # #   ld = p$folderpath.data.mvLDSC.ld.1kg,
  # #   filepathVariantsAdditional = p$filepath.SNPReference.1kg,
  # #   ldsc.log = p$setup.code.date,
  # #   preweight.alternativeCorrelationCorrection = T,
  # #   preweight.ChiSquare = F,
  # #   correctAttenuationBias = F,
  # #   doubleRegressionRoutine = F,
  # #   preweight.INFO=T,
  # #   resamplingMethod="vbcs",
  # #   blocksizeCM = 5,
  # #   filter.info = 0.0,
  # #   filter.maf = 0.01,
  # #   filter.zz.min = 0,
  # #   force.M = 7184778
  # #   )
  # # saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  # # cat("\n\n***END LDSC: covstruct.mvLDSC.1kg.vbcs.rblock5.winfo.altcw ***\n")



  #set the default mvLDSC object to use
  p$mvLD$covstruct.mvLDSC<-p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw
  #p$mvLD$covstruct.mvLDSC<-p$mvLD$covstruct.mvLDSC.1kg.noimp
  
  #Additional computations
  #saving the original S in case of smoothing experiments later stored in S
  p$mvLD$covstruct.mvLDSC$S.orig<-p$mvLD$covstruct.mvLDSC$S
  p$mvLD$covstruct.mvLDSC$S.smooth<-as.matrix((nearPD(p$mvLD$covstruct.mvLDSC$S, corr = FALSE))$mat)
  
  rm(readyMergedTraits)
  rm(readyMergedTraitsF)
  
  #save the mvLD output
  saveRDS(object = p$mvLD,file = p$filepath.mvLD)
  print("Multivariate LD score regression is done now and the resulting covariance structure should have been saved to a file.")
  
}


# 
# if(p$clOptions$task=="mvLD.gwis" || (!file.exists(p$filepath.mvLD.gwis) & p$clOptions$task=="0"))
# {
#   print("Running (multivariate) LD score regression. This might take a while. If the procedure runs for too long you may want to abort the process.")
#   
#   cat("The current task is specified as:",p$clOptions$task,"\n")
#   
#   #run mvLDSC
#   
#   #requires the bmigwas selection of traits
#   cat("\n\n*** LDSC: covstruct.mvLDSC.gwis ***\n")
#   p$mvLD.gwis$covstruct.mvLDSC.gwis.nonimputed<-ldscpp(
#     traits = p$sumstats.sel.gwis$mungedpath,
#     sample.prev =  p$sumstats.sel.gwis$samplePrevalence,
#     population.prev = p$sumstats.sel.gwis$populationPrevalence,
#     trait.names = p$sumstats.sel.gwis$code,
#     filepathLD = p$filepath.SNPReference.hc1kg,
#     filter.info = 0.6,
#     filter.maf = 0.01,
#     filter.region.df = p$highld_b38,
#     N = p$sumstats.sel.gwis$n_total,
#     ldsc.log = p$setup.code.date
#     )
#   saveRDS(object = p$mvLD.gwis,file = p$filepath.mvLD.gwis)
#   
#   print("Multivariate LD score regression is done now and the resulting covariance structure should have been saved to a file.")
#   
# }


if(p$clOptions$task=="mvLD" || p$clOptions$task=="mvLD.gwis"){
      quit(save = "no")
    }

#add newly computed heritabilities to the selected summary statistics table

p$sumstats$h2.GSEMmvLDSC.hm3<-NA_real_
p$sumstats$h2.se.GSEMmvLDSC.hm3<-NA_real_
rownames(p$mvLD$covstruct.GSEMmvLDSC.hm3$S)<-colnames(p$mvLD$covstruct.GSEMmvLDSC.hm3$S)
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.GSEMmvLDSC.hm3")]<-diag(p$mvLD$covstruct.GSEMmvLDSC.hm3$S)[p$sumstats.sel.ssimp.code]

#fetch the GSEM LDSC S.E.
p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE<-matrix(0, nrow(p$mvLD$covstruct.GSEMmvLDSC.hm3$S), nrow(p$mvLD$covstruct.GSEMmvLDSC.hm3$S))
p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE[lower.tri(p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE,diag=TRUE)] <-sqrt(diag(p$mvLD$covstruct.GSEMmvLDSC.hm3$V))
p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE[upper.tri(p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE)]<-t(p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE)[upper.tri(p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE)]
colnames(p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE)<-colnames(p$mvLD$covstruct.GSEMmvLDSC.hm3$S)
rownames(p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE)<-rownames(p$mvLD$covstruct.GSEMmvLDSC.hm3$S)
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.GSEMmvLDSC.hm3")]<-diag(p$mvLD$covstruct.GSEMmvLDSC.hm3$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.GSEMemulation.hm3<-NA_real_
p$sumstats$h2.se.mvLDSC.GSEMemulation.hm3<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.GSEMemulation.hm3")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.GSEMemulation.hm3")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.GSEMemulation.1kg.maf0_01<-NA_real_
p$sumstats$h2.se.mvLDSC.GSEMemulation.1kg.maf0_01<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.GSEMemulation.1kg.maf0_01")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.GSEMemulation.1kg.maf0_01")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.GSEMemulation.1kg.maf0_05<-NA_real_
p$sumstats$h2.se.mvLDSC.GSEMemulation.1kg.maf0_05<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.GSEMemulation.1kg.maf0_05")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_05$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.GSEMemulation.1kg.maf0_05")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_05$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.GSEMemulation.1kg.maf0_1<-NA_real_
p$sumstats$h2.se.mvLDSC.GSEMemulation.1kg.maf0_1<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.GSEMemulation.1kg.maf0_1")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_1$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.GSEMemulation.1kg.maf0_1")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_1$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.GSEMemulation.hc1kg.maf0_05<-NA_real_
p$sumstats$h2.se.mvLDSC.GSEMemulation.hc1kg.maf0_05<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.GSEMemulation.hc1kg.maf0_05")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.hc1kg.maf0_05$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.GSEMemulation.hc1kg.maf0_05")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.hc1kg.maf0_05$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.GSEMemulation.hc1kg.maf0_1<-NA_real_
p$sumstats$h2.se.mvLDSC.GSEMemulation.hc1kg.maf0_1<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.GSEMemulation.hc1kg.maf0_1")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.hc1kg.maf0_1$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.GSEMemulation.hc1kg.maf0_1")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.hc1kg.maf0_1$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.GSEMemulation.1kg.hm3res<-NA_real_
p$sumstats$h2.se.mvLDSC.GSEMemulation.1kg.hm3res<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.GSEMemulation.1kg.hm3res")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.hm3res$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.GSEMemulation.1kg.hm3res")]<-diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.hm3res$S.SE)[p$sumstats.sel.ssimp.code]

#1kG
#jn

p$sumstats$h2.mvLDSC.1kg.jn.varblock<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.varblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.varblock<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.varblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock5<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock5<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock5")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock5")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.varblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.varblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock5.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock5.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock5.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock5.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock.unfiltered<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock.unfiltered<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.unfiltered$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.unfiltered$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.varblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.varblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock5.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock5.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock5.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock5.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.varblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.varblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.jn.rblock5.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.jn.rblock5.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.jn.rblock5.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.jn.rblock5.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.jn.rblock5.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

#vbcs
p$sumstats$h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.varblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.rblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.varblock<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.varblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock5<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock5<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock5")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock5")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.varblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.varblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock5.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock5.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock5.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock5.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock.unfiltered<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock.unfiltered<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.unfiltered$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.unfiltered$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.varblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.varblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock5.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock5.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock5.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock5.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.varblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.1kg.vbcs.rblock5.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.1kg.vbcs.rblock5.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.1kg.vbcs.rblock5.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.1kg.vbcs.rblock5.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock5.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

# p$sumstats$h2.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw<-NA_real_
# p$sumstats$h2.se.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw<-NA_real_
# p$sumstats[p$sumstats.sel.code,c("h2.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw$S)[p$sumstats.sel.code]
# p$sumstats[p$sumstats.sel.code,c("h2.se.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw$S.SE)[p$sumstats.sel.code]
# 
# p$sumstats$h2.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw<-NA_real_
# p$sumstats$h2.se.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw<-NA_real_
# p$sumstats[p$sumstats.sel.code,c("h2.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw$S)[p$sumstats.sel.code]
# p$sumstats[p$sumstats.sel.code,c("h2.se.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw$S.SE)[p$sumstats.sel.code]

# p$sumstats$h2.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.ldimp<-NA_real_
# p$sumstats$h2.se.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.ldimp<-NA_real_
# p$sumstats[p$sumstats.sel.code,c("h2.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.ldimp")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.ldimp$S)[p$sumstats.sel.code]
# p$sumstats[p$sumstats.sel.code,c("h2.se.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.ldimp")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.ldimp$S.SE)[p$sumstats.sel.code]
# 
# p$sumstats$h2.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw.ldimp<-NA_real_
# p$sumstats$h2.se.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw.ldimp<-NA_real_
# p$sumstats[p$sumstats.sel.code,c("h2.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw.ldimp")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw.ldimp$S)[p$sumstats.sel.code]
# p$sumstats[p$sumstats.sel.code,c("h2.se.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw.ldimp")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.rblock.winfo.altcw.ldimp$S.SE)[p$sumstats.sel.code]

# #maf 0.01 test
# p$sumstats$h2.mvLDSC.hc1kg.jn.varblock.winfo.altcw.maf01test<-NA_real_
# p$sumstats$h2.se.mvLDSC.hc1kg.jn.varblock.winfo.altcw.maf01test<-NA_real_
# p$sumstats[p$sumstats.sel.code,c("h2.mvLDSC.hc1kg.jn.varblock.winfo.altcw.maf01test")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.jn.varblock.winfo.altcw.maf01test$S)[p$sumstats.sel.code]
# p$sumstats[p$sumstats.sel.code,c("h2.se.mvLDSC.hc1kg.jn.varblock.winfo.altcw.maf01test")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.jn.varblock.winfo.altcw.maf01test$S.SE)[p$sumstats.sel.code]
# 
# p$sumstats$h2.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.maf01test<-NA_real_
# p$sumstats$h2.se.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.maf01test<-NA_real_
# p$sumstats[p$sumstats.sel.code,c("h2.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.maf01test")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.maf01test$S)[p$sumstats.sel.code]
# p$sumstats[p$sumstats.sel.code,c("h2.se.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.maf01test")]<-diag(p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.maf01test$S.SE)[p$sumstats.sel.code]


#-----

#hm3
#jn

p$sumstats$h2.mvLDSC.hm3.jn.varblock<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.varblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.varblock<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.varblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.rblock<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.rblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.varblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.varblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.rblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.rblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.rblock.unfiltered<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.rblock.unfiltered<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.unfiltered$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.unfiltered$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.varblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.varblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.rblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.rblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.jn.varblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.varblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.jn.rblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.jn.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.jn.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

#vbcs
p$sumstats$h2.mvLDSC.hm3.vbcs.varblock<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.varblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.varblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.rblock<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.rblock<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.rblock")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.varblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.varblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.varblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.rblock.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.rblock.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.rblock.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.rblock.unfiltered<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.rblock.unfiltered<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.unfiltered$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.rblock.unfiltered")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.unfiltered$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.varblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.varblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.varblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.rblock.winfo<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.rblock.winfo<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.rblock.winfo")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.varblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.varblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]

p$sumstats$h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw<-NA_real_
p$sumstats$h2.se.mvLDSC.hm3.vbcs.rblock.winfo.altcw<-NA_real_
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw$S)[p$sumstats.sel.ssimp.code]
p$sumstats[p$sumstats.sel.ssimp.code,c("h2.se.mvLDSC.hm3.vbcs.rblock.winfo.altcw")]<-diag(p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw$S.SE)[p$sumstats.sel.ssimp.code]




# add when running original ldsc again
# p$sumstats.sel$h2_orig.liability_mvLDSC.1kg.small<-as.numeric(lapply(X = p$mvLD$h2.munged, FUN = function(x){ifelse(any(colnames(x$df)=="h2_liab"),x$df$h2_liab,x$df$h2_obs)}))
# p$sumstats.sel$h2_orig.se.liability_mvLDSC.1kg.small<-as.numeric(lapply(X = p$mvLD$h2.munged, FUN = function(x){ifelse(any(colnames(x$df)=="h2_liab_se"),x$df$h2_liab_se,x$df$h2_obs_se)}))
# p$sumstats.sel$h2_orig.liability_mvLDSC.1kg.ldimp.small<-as.numeric(lapply(X = p$mvLD$h2.500, FUN = function(x){ifelse(any(colnames(x$df)=="h2_liab"),x$df$h2_liab,x$df$h2_obs)}))
# p$sumstats.sel$h2_orig.se.liability_mvLDSC.1kg.ldimp.small<-as.numeric(lapply(X = p$mvLD$h2.500, FUN = function(x){ifelse(any(colnames(x$df)=="h2_liab_se"),x$df$h2_liab_se,x$df$h2_obs_se)}))
# p$sumstats.sel$h2_orig.liability_mvLDSC.1kg.ssimp.small<-as.numeric(lapply(X = p$mvLD$h2.ssimp, FUN = function(x){ifelse(any(colnames(x$df)=="h2_liab"),x$df$h2_liab,x$df$h2_obs)}))
# p$sumstats.sel$h2_orig.se.liability_mvLDSC.1kg.ssimp.small<-as.numeric(lapply(X = p$mvLD$h2.ssimp, FUN = function(x){ifelse(any(colnames(x$df)=="h2_liab_se"),x$df$h2_liab_se,x$df$h2_obs_se)}))

#add variant counts to sumstats
p$sumstats.sel$variants_raw<-p$mvLD$covstruct.mvLDSC.1kg$impstats$m
p$sumstats.sel$variants_ldimp<-p$mvLD$covstruct.mvLDSC.1kg.05cm$impstats$m.imputed
p$sumstats.sel$variants_ldimp.small<-p$mvLD$covstruct.mvLDSC.1kg.05cm.small$impstats[p$sumstats.sel$code,]$m.imputed
p$sumstats.sel$variants_ssimp.small<-p$mvLD$covstruct.mvLDSC.1kg.ssimp.small$impstats[p$sumstats.sel$code,]$m.imputed

  
```

#Distributions of standardised variances
```{r overview of standardised variance distributions, eval=FALSE, purl=FALSE}

#((p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock$S.SE)^2)

varLowerTri<-((p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock$S.SE))[lower.tri(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock$S.SE,diag = T)]
dfToplot <- data.frame(x=varLowerTri)

skimRes <- skim(varLowerTri)
skimRes$numeric.sd
skimRes$numeric.mean

ggplot(data = dfToplot, aes(x = x)) + 
    #geom_histogram(bins = 30, color = "black", fill = "gray",aes(y = ..count..)) +
    geom_density(adjust=1.5, alpha = 0.5, linewidth=1) + #adjust=1.5 #c(0, Inf)
    labs(y = paste0("Density Y"),
     x = "covG S.E.",
     title = paste0("Distribution of covG VBCS S.E. \nn[non NA]=",sum(!is.na(dfToplot$x)))) +
  xlim(0,0.03) +
  ylim(0,160)
  ggsave(filename=file.path(p$folderpath.plots,paste0("se_distributions.vbcs.png")), plot = last_plot(), width = 150, height = 100, units = "mm")
  
  
varLowerTri<-((p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S.SE))[lower.tri(p$mvLD$covstruct.mvLDSC.1kg.jn.varblock$S.SE,diag = T)]
dfToplot <- data.frame(x=varLowerTri)


skimRes <- skim(varLowerTri)
skimRes$numeric.sd
skimRes$numeric.mean

ggplot(data = dfToplot, aes(x = x)) + 
    #geom_histogram(bins = 20, color = "black", fill = "gray",aes(y = ..count..)) +
    geom_density(adjust=1.5, alpha = 0.5, linewidth=1) + 
    labs(y = paste0("Density Y"),
     x = "covG S.E.",
     title = paste0("Distribution of covG JN S.E. \nn[non NA]=",sum(!is.na(dfToplot$x)))) +
  xlim(0,0.03) +
  ylim(0,160)
  ggsave(filename=file.path(p$folderpath.plots,paste0("se_distributions.jn.png")), plot = last_plot(), width = 150, height = 100, units = "mm")

  
#Standadised versions
  
df.cv <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock, labelString = "label", groupString = "group")
  
varLowerTri<-df.cv$cv
dfToplot <- data.frame(x=varLowerTri)

skimRes <- skim(varLowerTri)
skimRes$numeric.sd
skimRes$numeric.mean

ggplot(data = dfToplot, aes(x = x)) + 
    #geom_histogram(bins = 30, color = "black", fill = "gray",aes(y = ..count..)) +
    geom_density(adjust=1.5, alpha = 0.5, linewidth=1) + #adjust=1.5 #c(0, Inf)
    labs(y = paste0("Density Y"),
     x = "covG CV.",
     title = paste0("Distribution of covG VBCS CV. \nn[non NA]=",sum(!is.na(dfToplot$x)))) +
  xlim(0,0.25) +
  ylim(0,15)
  ggsave(filename=file.path(p$folderpath.plots,paste0("cv_distributions.vbcs.png")), plot = last_plot(), width = 150, height = 100, units = "mm")
  
  
df.cv <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock, labelString = "label", groupString = "group")
  
varLowerTri<-df.cv$cv
dfToplot <- data.frame(x=varLowerTri)

skimRes <- skim(varLowerTri)
skimRes$numeric.sd
skimRes$numeric.mean

ggplot(data = dfToplot, aes(x = x)) + 
    #geom_histogram(bins = 20, color = "black", fill = "gray",aes(y = ..count..)) +
    geom_density(adjust=1.5, alpha = 0.5, linewidth=1) + 
    labs(y = paste0("Density Y"),
     x = "covG CV.",
     title = paste0("Distribution of covG JN CV. \nn[non NA]=",sum(!is.na(dfToplot$x)))) +
  xlim(0,0.25) +
  ylim(0,15)
  ggsave(filename=file.path(p$folderpath.plots,paste0("cv_distributions.jn.png")), plot = last_plot(), width = 150, height = 100, units = "mm")


```


# Overview of LD score regression CV's HM3 NEW
```{r overview ldsc cv NEW HM3, fig.width=9, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}

#individual methods - all
p$cv_coverage.ldsc.hm3.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw, labelString = "hm3.jn.varblock", groupString = "hm3.jn.varblock")

p$cv_coverage.ldsc.hm3.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw, labelString = "hm3.jn.rblock", groupString = "hm3.jn.rblock")

p$cv_coverage.ldsc.hm3.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw, labelString = "hm3.vbcs.varblock", groupString = "hm3.vbcs.varblock")

p$cv_coverage.ldsc.hm3.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw, labelString = "hm3.vbcs.rblock", groupString = "hm3.vbcs.rblock")

p$cv_coverage <- rbind(
  #p$cv_coverage.ldsc.hm3.jn.fixed,
  p$cv_coverage.ldsc.hm3.jn.varblock,
  p$cv_coverage.ldsc.hm3.jn.rblock,
  p$cv_coverage.ldsc.hm3.vbcs.varblock,
  p$cv_coverage.ldsc.hm3.vbcs.rblock
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]

# cTest.m.cv.jn.fixed<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$cv)
# cTest.m.cv.jn.fixed
# cTest.m.cv.jn.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$cv)
# cTest.m.cv.jn.varblock
# cTest.m.cv.jn.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$cv)
# cTest.m.cv.jn.rblock
# cTest.m.cv.vbcs.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$cv)
# cTest.m.cv.vbcs.varblock
# cTest.m.cv.vbcs.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$cv)
# cTest.m.cv.vbcs.rblock


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC number of variants against covG coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage, aes(x=m, y=se, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
  coord_cartesian(ylim = c(0, 0.1)) +
  ggtitle(label = "LDSC number of variants against covG S.E.") +
  stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.covG.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC number of variants against h2 coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=se, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
  coord_cartesian(ylim = c(0, 0.1)) +
  ggtitle(label = "LDSC, number of variants against h2 S.E.") +
  stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.h2.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


#individual methods - additions, jn.varblock
p$cv_coverage.ldsc.hm3.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3, labelString = "hm3.gsem", groupString = "hm3.gsem")

p$cv_coverage.ldsc.hm3.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.varblock, labelString = "hm3.jn.varblock", groupString = "hm3.jn.varblock")

p$cv_coverage.ldsc.hm3.jn.varblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.altcw, labelString = "hm3.jn.varblock.altcw", groupString = "hm3.jn.varblock.altcw")

p$cv_coverage.ldsc.hm3.jn.varblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo, labelString = "hm3.jn.varblock.winfo", groupString = "hm3.jn.varblock.winfo")

p$cv_coverage.ldsc.hm3.jn.varblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw, labelString = "hm3.jn.varblock.winfo.altcw", groupString = "hm3.jn.varblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.hm3.gsem,
  p$cv_coverage.ldsc.hm3.jn.varblock,
  p$cv_coverage.ldsc.hm3.jn.varblock.altcw,
  p$cv_coverage.ldsc.hm3.jn.varblock.winfo,
  p$cv_coverage.ldsc.hm3.jn.varblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC number of variants against covG coefficient of variation (CV), block jackknife, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.hm3.jn.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC number of variants against h2 coefficient of variation (CV), block jackknife, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.hm3.jn.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



#individual methods - additions, vbcs.varblock
p$cv_coverage.ldsc.hm3.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3, labelString = "hm3.gsem", groupString = "hm3.gsem")

p$cv_coverage.ldsc.hm3.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock, labelString = "hm3.vbcs.varblock", groupString = "hm3.vbcs.varblock")

p$cv_coverage.ldsc.hm3.vbcs.varblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.altcw, labelString = "hm3.vbcs.varblock.altcw", groupString = "hm3.vbcs.varblock.altcw")

p$cv_coverage.ldsc.hm3.vbcs.varblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo, labelString = "hm3.vbcs.varblock.winfo", groupString = "hm3.vbcs.varblock.winfo")

p$cv_coverage.ldsc.hm3.vbcs.varblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw, labelString = "hm3.vbcs.varblock.winfo.altcw", groupString = "hm3.vbcs.varblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.hm3.gsem,
  p$cv_coverage.ldsc.hm3.vbcs.varblock,
  p$cv_coverage.ldsc.hm3.vbcs.varblock.altcw,
  p$cv_coverage.ldsc.hm3.vbcs.varblock.winfo,
  p$cv_coverage.ldsc.hm3.vbcs.varblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC number of variants against covG coefficient of variation (CV), vbcs, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.hm3.vbcs.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC number of variants against h2 coefficient of variation (CV), vbcs, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.hm3.vbcs.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


#individual methods - additions, jn.rblock
p$cv_coverage.ldsc.hm3.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3, labelString = "hm3.gsem", groupString = "hm3.gsem")

p$cv_coverage.ldsc.hm3.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.rblock, labelString = "hm3.jn.rblock", groupString = "hm3.jn.rblock")

p$cv_coverage.ldsc.hm3.jn.rblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.altcw, labelString = "hm3.jn.rblock.altcw", groupString = "hm3.jn.rblock.altcw")

p$cv_coverage.ldsc.hm3.jn.rblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo, labelString = "hm3.jn.rblock.winfo", groupString = "hm3.jn.rblock.winfo")

p$cv_coverage.ldsc.hm3.jn.rblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw, labelString = "hm3.jn.rblock.winfo.altcw", groupString = "hm3.jn.rblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.hm3.gsem,
  p$cv_coverage.ldsc.hm3.jn.rblock,
  p$cv_coverage.ldsc.hm3.jn.rblock.altcw,
  p$cv_coverage.ldsc.hm3.jn.rblock.winfo,
  p$cv_coverage.ldsc.hm3.jn.rblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC number of variants against covG coefficient of variation (CV), block jackknife, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.hm3.jn.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC number of variants against h2 coefficient of variation (CV), block jackknife, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.hm3.jn.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



#individual methods - additions, vbcs.rblock
p$cv_coverage.ldsc.hm3.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3, labelString = "hm3.gsem", groupString = "hm3.gsem")

p$cv_coverage.ldsc.hm3.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock, labelString = "hm3.vbcs.rblock", groupString = "hm3.vbcs.rblock")

p$cv_coverage.ldsc.hm3.vbcs.rblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.altcw, labelString = "hm3.vbcs.rblock.altcw", groupString = "hm3.vbcs.rblock.altcw")

p$cv_coverage.ldsc.hm3.vbcs.rblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo, labelString = "hm3.vbcs.rblock.winfo", groupString = "hm3.vbcs.rblock.winfo")

p$cv_coverage.ldsc.hm3.vbcs.rblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw, labelString = "hm3.vbcs.rblock.winfo.altcw", groupString = "hm3.vbcs.rblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.hm3.gsem,
  p$cv_coverage.ldsc.hm3.vbcs.rblock,
  p$cv_coverage.ldsc.hm3.vbcs.rblock.altcw,
  p$cv_coverage.ldsc.hm3.vbcs.rblock.winfo,
  p$cv_coverage.ldsc.hm3.vbcs.rblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC number of variants against covG coefficient of variation (CV), vbcs, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.hm3.vbcs.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC number of variants against h2 coefficient of variation (CV), vbcs, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.hm3.vbcs.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


#block methods
p$cv_coverage.ldsc.hm3.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw, labelString = "hm3.jn.varblock", groupString = "varblock")

p$cv_coverage.ldsc.hm3.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw, labelString = "hm3.jn.rblock", groupString = "rblock")

p$cv_coverage.ldsc.hm3.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw, labelString = "hm3.vbcs.varblock", groupString = "varblock")

p$cv_coverage.ldsc.hm3.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw, labelString = "hm3.vbcs.rblock", groupString = "rblock")

p$cv_coverage <- rbind(
  #p$cv_coverage.ldsc.hm3.jn.fixed,
  p$cv_coverage.ldsc.hm3.jn.varblock,
  p$cv_coverage.ldsc.hm3.jn.rblock,
  p$cv_coverage.ldsc.hm3.vbcs.varblock,
  p$cv_coverage.ldsc.hm3.vbcs.rblock
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]

# cTest.m.cv.jn.fixed<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$cv)
# cTest.m.cv.jn.fixed
# cTest.m.cv.jn.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$cv)
# cTest.m.cv.jn.varblock
# cTest.m.cv.jn.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$cv)
# cTest.m.cv.jn.rblock
# cTest.m.cv.vbcs.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$cv)
# cTest.m.cv.vbcs.varblock
# cTest.m.cv.vbcs.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$cv)
# cTest.m.cv.vbcs.rblock


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC number of variants against covG coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.8, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.block.covG.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage, aes(x=m, y=se, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
  coord_cartesian(ylim = c(0, 0.1)) +
  ggtitle(label = "LDSC number of variants against covG S.E.") +
  stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.block.covG.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC number of variants against h2 coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.8, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.block.h2.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=se, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
  coord_cartesian(ylim = c(0, 0.1)) +
  ggtitle(label = "LDSC, number of variants against h2 S.E.") +
  stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.block.h2.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



# recombination groups
p$cv_coverage.ldsc.hm3.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.varblock.winfo.altcw, labelString = "hm3.jn.varblock", groupString = "jn")

p$cv_coverage.ldsc.hm3.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.jn.rblock.winfo.altcw, labelString = "hm3.jn.rblock", groupString = "jn")

p$cv_coverage.ldsc.hm3.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw, labelString = "hm3.vbcs.varblock", groupString = "vbcs")

p$cv_coverage.ldsc.hm3.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.hm3.vbcs.rblock.winfo.altcw, labelString = "hm3.vbcs.rblock", groupString = "vbcs")

p$cv_coverage <- rbind(
  #p$cv_coverage.ldsc.hm3.jn.fixed,
  p$cv_coverage.ldsc.hm3.jn.varblock,
  p$cv_coverage.ldsc.hm3.jn.rblock,
  p$cv_coverage.ldsc.hm3.vbcs.varblock,
  p$cv_coverage.ldsc.hm3.vbcs.rblock
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC number of variants against covG coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.recomb.covG.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage, aes(x=m, y=se, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
  coord_cartesian(ylim = c(0, 0.1)) +
  ggtitle(label = "LDSC number of variants against covG S.E.") +
  stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.recomb.covG.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC number of variants against h2 coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.4,label.y.npc = 0.8, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.recomb.h2.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=se, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
  coord_cartesian(ylim = c(0, 0.1)) +
  ggtitle(label = "LDSC, number of variants against h2 S.E.") +
  stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.recomb.h2.hm3.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


```

# Overview of LD score regression CV's 1kG NEW
```{r overview ldsc cv NEW 1kG, fig.width=9, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}


#individual methods
p$cv_coverage.ldsc.1kg.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, labelString = "1kg.gsem", groupString = "1kg.gsem")

p$cv_coverage.ldsc.1kg.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw, labelString = "1kg.jn.varblock", groupString = "1kg.jn.varblock")

p$cv_coverage.ldsc.1kg.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw, labelString = "1kg.jn.rblock", groupString = "1kg.jn.rblock")

p$cv_coverage.ldsc.1kg.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, labelString = "1kg.vbcs.varblock", groupString = "1kg.vbcs.varblock")

p$cv_coverage.ldsc.1kg.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw, labelString = "1kg.vbcs.rblock", groupString = "1kg.vbcs.rblock")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.1kg.gsem,
  p$cv_coverage.ldsc.1kg.jn.varblock,
  p$cv_coverage.ldsc.1kg.jn.rblock,
  p$cv_coverage.ldsc.1kg.vbcs.varblock,
  p$cv_coverage.ldsc.1kg.vbcs.rblock
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]

# cTest.m.cv.jn.fixed<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$cv)
# cTest.m.cv.jn.fixed
# cTest.m.cv.jn.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$cv)
# cTest.m.cv.jn.varblock
# cTest.m.cv.jn.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$cv)
# cTest.m.cv.jn.rblock
# cTest.m.cv.vbcs.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$cv)
# cTest.m.cv.vbcs.varblock
# cTest.m.cv.vbcs.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$cv)
# cTest.m.cv.vbcs.rblock


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ number of variants against covG coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

# ggplot(data = p$cv_coverage, aes(x=m, y=se, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
#   coord_cartesian(ylim = c(0, 0.1)) +
#   ggtitle(label = "LDSC number of variants against covG S.E.") +
#   stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.covG.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC++ number of variants against h2 coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

# ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=se, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
#   coord_cartesian(ylim = c(0, 0.1)) +
#   ggtitle(label = "LDSC, number of variants against h2 S.E.") +
#   stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.h2.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



#individual methods - additions, jn.varblock
p$cv_coverage.ldsc.1kg.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, labelString = "1kg.gsem", groupString = "1kg.gsem")

p$cv_coverage.ldsc.1kg.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock, labelString = "1kg.jn.varblock", groupString = "1kg.jn.varblock")

p$cv_coverage.ldsc.1kg.jn.varblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.altcw, labelString = "1kg.jn.varblock.altcw", groupString = "1kg.jn.varblock.altcw")

p$cv_coverage.ldsc.1kg.jn.varblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo, labelString = "1kg.jn.varblock.winfo", groupString = "1kg.jn.varblock.winfo")

p$cv_coverage.ldsc.1kg.jn.varblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw, labelString = "1kg.jn.varblock.winfo.altcw", groupString = "1kg.jn.varblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.1kg.gsem,
  p$cv_coverage.ldsc.1kg.jn.varblock,
  p$cv_coverage.ldsc.1kg.jn.varblock.altcw,
  p$cv_coverage.ldsc.1kg.jn.varblock.winfo,
  p$cv_coverage.ldsc.1kg.jn.varblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ number of variants against covG coefficient of variation (CV),\n block jackknife, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.1kg.jn.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC++ number of variants against h2 coefficient of variation (CV),\n block jackknife, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.1kg.jn.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



#individual methods - additions, vbcs.varblock
p$cv_coverage.ldsc.1kg.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, labelString = "1kg.gsem", groupString = "1kg.gsem")

p$cv_coverage.ldsc.1kg.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock, labelString = "1kg.vbcs.varblock", groupString = "1kg.vbcs.varblock")

p$cv_coverage.ldsc.1kg.vbcs.varblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.altcw, labelString = "1kg.vbcs.varblock.altcw", groupString = "1kg.vbcs.varblock.altcw")

p$cv_coverage.ldsc.1kg.vbcs.varblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo, labelString = "1kg.vbcs.varblock.winfo", groupString = "1kg.vbcs.varblock.winfo")

p$cv_coverage.ldsc.1kg.vbcs.varblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, labelString = "1kg.vbcs.varblock.winfo.altcw", groupString = "1kg.vbcs.varblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.1kg.gsem,
  p$cv_coverage.ldsc.1kg.vbcs.varblock,
  p$cv_coverage.ldsc.1kg.vbcs.varblock.altcw,
  p$cv_coverage.ldsc.1kg.vbcs.varblock.winfo,
  p$cv_coverage.ldsc.1kg.vbcs.varblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ number of variants against covG coefficient of variation (CV),\n vbcs, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.1kg.vbcs.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC++ number of variants against h2 coefficient of variation (CV),\n vbcs, number-of-variants blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.1kg.vbcs.varblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


#individual methods - additions, jn.rblock
p$cv_coverage.ldsc.1kg.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, labelString = "1kg.gsem", groupString = "1kg.gsem")

p$cv_coverage.ldsc.1kg.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock, labelString = "1kg.jn.rblock", groupString = "1kg.jn.rblock")

p$cv_coverage.ldsc.1kg.jn.rblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.altcw, labelString = "1kg.jn.rblock.altcw", groupString = "1kg.jn.rblock.altcw")

p$cv_coverage.ldsc.1kg.jn.rblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo, labelString = "1kg.jn.rblock.winfo", groupString = "1kg.jn.rblock.winfo")

p$cv_coverage.ldsc.1kg.jn.rblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw, labelString = "1kg.jn.rblock.winfo.altcw", groupString = "1kg.jn.rblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.1kg.gsem,
  p$cv_coverage.ldsc.1kg.jn.rblock,
  p$cv_coverage.ldsc.1kg.jn.rblock.altcw,
  p$cv_coverage.ldsc.1kg.jn.rblock.winfo,
  p$cv_coverage.ldsc.1kg.jn.rblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ number of variants against covG coefficient of variation (CV),\n block jackknife, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.1kg.jn.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC++ number of variants against h2 coefficient of variation (CV),\n block jackknife, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.1kg.jn.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



#individual methods - additions, vbcs.rblock
p$cv_coverage.ldsc.1kg.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, labelString = "1kg.gsem", groupString = "1kg.gsem")

p$cv_coverage.ldsc.1kg.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock, labelString = "1kg.vbcs.rblock", groupString = "1kg.vbcs.rblock")

p$cv_coverage.ldsc.1kg.vbcs.rblock.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.altcw, labelString = "1kg.vbcs.rblock.altcw", groupString = "1kg.vbcs.rblock.altcw")

p$cv_coverage.ldsc.1kg.vbcs.rblock.winfo <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo, labelString = "1kg.vbcs.rblock.winfo", groupString = "1kg.vbcs.rblock.winfo")

p$cv_coverage.ldsc.1kg.vbcs.rblock.winfo.altcw <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw, labelString = "1kg.vbcs.rblock.winfo.altcw", groupString = "1kg.vbcs.rblock.winfo.altcw")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.1kg.gsem,
  p$cv_coverage.ldsc.1kg.vbcs.rblock,
  p$cv_coverage.ldsc.1kg.vbcs.rblock.altcw,
  p$cv_coverage.ldsc.1kg.vbcs.rblock.winfo,
  p$cv_coverage.ldsc.1kg.vbcs.rblock.winfo.altcw
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ number of variants against covG coefficient of variation (CV),\n vbcs, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.covG.1kg.vbcs.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC++ number of variants against h2 coefficient of variation (CV),\n vbcs, recombination distance blocks") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 1, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.h2.1kg.vbcs.rblock.png"), plot = last_plot(), width = 200, height = 200, units = "mm")







#block methods
p$cv_coverage.ldsc.1kg.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, labelString = "1kg.gsem", groupString = "1kg.gsem")

p$cv_coverage.ldsc.1kg.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw, labelString = "1kg.jn.varblock", groupString = "varblock")

p$cv_coverage.ldsc.1kg.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw, labelString = "1kg.jn.rblock", groupString = "rblock")

p$cv_coverage.ldsc.1kg.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, labelString = "1kg.vbcs.varblock", groupString = "varblock")

p$cv_coverage.ldsc.1kg.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw, labelString = "1kg.vbcs.rblock", groupString = "rblock")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.1kg.gsem,
  p$cv_coverage.ldsc.1kg.jn.varblock,
  p$cv_coverage.ldsc.1kg.jn.rblock,
  p$cv_coverage.ldsc.1kg.vbcs.varblock,
  p$cv_coverage.ldsc.1kg.vbcs.rblock
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]

# cTest.m.cv.jn.fixed<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.fixed",]$cv)
# cTest.m.cv.jn.fixed
# cTest.m.cv.jn.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.varblock",]$cv)
# cTest.m.cv.jn.varblock
# cTest.m.cv.jn.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="jn.rblock",]$cv)
# cTest.m.cv.jn.rblock
# cTest.m.cv.vbcs.varblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.varblock",]$cv)
# cTest.m.cv.vbcs.varblock
# cTest.m.cv.vbcs.rblock<-corr.test(x = p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$m, y=p$cv_coverage[p$cv_coverage$g=="vbcs.rblock",]$cv)
# cTest.m.cv.vbcs.rblock


library(ggpubr)
ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ number of variants against covG coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.8, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.block.covG.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

# ggplot(data = p$cv_coverage, aes(x=m, y=se, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
#   coord_cartesian(ylim = c(0, 0.1)) +
#   ggtitle(label = "LDSC number of variants against covG S.E.") +
#   stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.block.covG.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC++ number of variants against h2 coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.8, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.block.h2.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

# ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=se, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
#   coord_cartesian(ylim = c(0, 0.1)) +
#   ggtitle(label = "LDSC, number of variants against h2 S.E.") +
#   stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.block.h2.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



# recombination groups
p$cv_coverage.ldsc.1kg.gsem <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, labelString = "1kg.gsem", groupString = "1kg.gsem")

p$cv_coverage.ldsc.1kg.jn.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw, labelString = "1kg.jn.varblock", groupString = "jn")

p$cv_coverage.ldsc.1kg.jn.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw, labelString = "1kg.jn.rblock", groupString = "jn")

p$cv_coverage.ldsc.1kg.vbcs.varblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, labelString = "1kg.vbcs.varblock", groupString = "vbcs")

p$cv_coverage.ldsc.1kg.vbcs.rblock <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw, labelString = "1kg.vbcs.rblock", groupString = "vbcs")

p$cv_coverage <- rbind(
  p$cv_coverage.ldsc.1kg.gsem,
  p$cv_coverage.ldsc.1kg.jn.varblock,
  p$cv_coverage.ldsc.1kg.jn.rblock,
  p$cv_coverage.ldsc.1kg.vbcs.varblock,
  p$cv_coverage.ldsc.1kg.vbcs.rblock
  )

p$cv_coverage<-p$cv_coverage[!is.na(p$cv_coverage$cv),]


ggplot(data = p$cv_coverage, aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ number of variants against covG coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.recomb.covG.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

# ggplot(data = p$cv_coverage, aes(x=m, y=se, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
#   coord_cartesian(ylim = c(0, 0.1)) +
#   ggtitle(label = "LDSC number of variants against covG S.E.") +
#   stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.recomb.covG.png"), plot = last_plot(), width = 200, height = 200, units = "mm")


ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=cv, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.7)) +
  ggtitle(label = "LDSC++ number of variants against h2 coefficient of variation (CV)") + 
  stat_cor(method="pearson", label.x.npc = 0.4,label.y.npc = 0.8, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_cv.recomb.h2.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

# ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=m, y=se, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) + 
#   coord_cartesian(ylim = c(0, 0.1)) +
#   ggtitle(label = "LDSC, number of variants against h2 S.E.") +
#   stat_cor(method="pearson", label.x.npc = 0.5,label.y.npc = 0.4, aes(group=g, color=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_m_se.recomb.h2.png"), plot = last_plot(), width = 200, height = 200, units = "mm")




```


# Overview of LD score regression h2 per INFO-bin
```{r overview regression info bins, fig.width=9, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}

# jn
p$infobin_coverage.8_10 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_8_10, labelString = "8_10", groupString = "no weighting")
p$infobin_coverage.8_10$minfo<-0.9

p$infobin_coverage.6_8 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_6_8, labelString = "6_8", groupString = "no weighting")
p$infobin_coverage.6_8$minfo<-0.7

p$infobin_coverage.4_6 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_4_6, labelString = "4_6", groupString = "no weighting")
p$infobin_coverage.4_6$minfo<-0.5

p$infobin_coverage.0_4 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_0_4, labelString = "0_4", groupString = "no weighting")
p$infobin_coverage.0_4$minfo<-0.25

p$infobin_coverage.8_10w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_8_10, labelString = "8_10", groupString = "INFO weighting")
p$infobin_coverage.8_10w$minfo<-0.9

p$infobin_coverage.6_8w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_6_8, labelString = "6_8", groupString = "INFO weighting")
p$infobin_coverage.6_8w$minfo<-0.7

p$infobin_coverage.4_6w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_4_6, labelString = "4_6", groupString = "INFO weighting")
p$infobin_coverage.4_6w$minfo<-0.5

p$infobin_coverage.0_4w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_0_4, labelString = "0_4", groupString = "INFO weighting")
p$infobin_coverage.0_4w$minfo<-0.25



p$cv_coverage <- rbind(
  p$infobin_coverage.8_10,
  p$infobin_coverage.6_8,
  p$infobin_coverage.4_6,
  p$infobin_coverage.0_4,
  p$infobin_coverage.8_10w,
  p$infobin_coverage.6_8w,
  p$infobin_coverage.4_6w,
  p$infobin_coverage.0_4w
  )

library(ggpubr)
# ggplot(data = p$cv_coverage, aes(x=minfo, y=s, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) +
#   #coord_cartesian(ylim = c(0, 0.9)) +
#   ggtitle(label = "LDSC++ mINFO vs covG") + 
#   stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_covg.jn.png"), plot = last_plot(), width = 200, height = 200, units = "mm")
# 
# ggplot(data = p$cv_coverage, aes(x=minfo, y=m, shape=l, color=g)) +
#   geom_point(size=3) +
#   #geom_smooth(method="glm", aes(group=g, color=g), level=0) +
#   #coord_cartesian(ylim = c(0, 0.9)) +
#   ggtitle(label = "LDSC++ mINFO vs #variants")
#   #stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_m.jn.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=minfo, y=s, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ block jackknife resampling, mINFO vs covG, only h2") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_covg.jn.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=minfo, y=m, shape=l, color=g)) +
  geom_point(size=3) +
  #geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ block jackknife resampling, mINFO vs #variants, only h2")
  #stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_m.jn.png"), plot = last_plot(), width = 200, height = 200, units = "mm")



# vbcs
p$infobin_coverage.8_10 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_8_10, labelString = "8_10", groupString = "no weighting")
p$infobin_coverage.8_10$minfo<-0.9

p$infobin_coverage.6_8 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_6_8, labelString = "6_8", groupString = "no weighting")
p$infobin_coverage.6_8$minfo<-0.7

p$infobin_coverage.4_6 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_4_6, labelString = "4_6", groupString = "no weighting")
p$infobin_coverage.4_6$minfo<-0.5

p$infobin_coverage.0_4 <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_0_4, labelString = "0_4", groupString = "no weighting")
p$infobin_coverage.0_4$minfo<-0.25

p$infobin_coverage.8_10w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_8_10, labelString = "8_10", groupString = "INFO weighting")
p$infobin_coverage.8_10w$minfo<-0.9

p$infobin_coverage.6_8w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_6_8, labelString = "6_8", groupString = "INFO weighting")
p$infobin_coverage.6_8w$minfo<-0.7

p$infobin_coverage.4_6w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_4_6, labelString = "4_6", groupString = "INFO weighting")
p$infobin_coverage.4_6w$minfo<-0.5

p$infobin_coverage.0_4w <- p$getCVDataframe(cCovstruct = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_0_4, labelString = "0_4", groupString = "INFO weighting")
p$infobin_coverage.0_4w$minfo<-0.25



p$cv_coverage <- rbind(
  p$infobin_coverage.8_10,
  p$infobin_coverage.6_8,
  p$infobin_coverage.4_6,
  p$infobin_coverage.0_4,
  p$infobin_coverage.8_10w,
  p$infobin_coverage.6_8w,
  p$infobin_coverage.4_6w,
  p$infobin_coverage.0_4w
  )

library(ggpubr)
# ggplot(data = p$cv_coverage, aes(x=minfo, y=s, shape=l, color=g)) +
#   geom_point(size=3) +
#   geom_smooth(method="glm", aes(group=g, color=g), level=0) +
#   #coord_cartesian(ylim = c(0, 0.9)) +
#   ggtitle(label = "LDSC++ mINFO vs covG") + 
#   stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_covg.vbcs.png"), plot = last_plot(), width = 200, height = 200, units = "mm")
# 
# ggplot(data = p$cv_coverage, aes(x=minfo, y=m, shape=l, color=g)) +
#   geom_point(size=3) +
#   #geom_smooth(method="glm", aes(group=g, color=g), level=0) +
#   #coord_cartesian(ylim = c(0, 0.9)) +
#   ggtitle(label = "LDSC++ mINFO vs #variants")
#   #stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
#   #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
# ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_m.vbcs.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=minfo, y=s, shape=l, color=g)) +
  geom_point(size=3) +
  geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ variable block-count sampling, mINFO vs covG, only h2") + 
  stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_covg.vbcs.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

ggplot(data = p$cv_coverage[p$cv_coverage$isDiag==T,], aes(x=minfo, y=m, shape=l, color=g)) +
  geom_point(size=3) +
  #geom_smooth(method="glm", aes(group=g, color=g), level=0) +
  #coord_cartesian(ylim = c(0, 0.9)) +
  ggtitle(label = "LDSC++ variable block-count sampling, mINFO vs #variants, only h2")
  #stat_cor(method="pearson", label.x.npc = 0.3,label.y.npc = 0.9, aes(group=g))
  #stat_cor(method="pearson", label.x.npc = "middle",label.y.npc = "bottom")
ggsave(filename=file.path(p$folderpath.plots,"ldsc_info_m.vbcs.png"), plot = last_plot(), width = 200, height = 200, units = "mm")

```


# Overview of rg between selected traits
```{r overview pairwise rg, fig.width=9, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}


#View(p$sumstats.sel.table)
# 
# library(heatmaply)
# #ggheatmap() #static version
# #heatmaply()
# heatmaply(
#     p$mvLD$covstruct.mvLDSC.1kg$S.forPlot,
#     cellnote = round(p$mvLD$covstruct.mvLDSC.1kg$S,digits = 3),
#     cellnote_size = 16,
#     cellnote_textposition = "middle center",
#     show_dendrogram = c(T,F),
#     xlab = "Trait GWAS",
#     ylab = "Trait GWAS",
#     main = "cov<sub>g</sub>",
#     scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
#         low = theme.color$contrastDark3, 
#         mid = "black",
#         high = theme.color$contrastLight3,
#         breaks = c(-0.5,0,0.5)
#       ),
#     file=file.path(p$folderpath.plots,paste0("cov_cluster.noimp.png")),
#     width = 1200, height = 1200,
#     hclust_method = "ward.D2"
#     #scale = "row"
#   )
# 
# heatmaply(
#     clipValues(p$mvLD$covstruct.mvLDSC.1kg$S_Stand.forPlot,-1,1),
#     cellnote = round(clipValues(p$mvLD$covstruct.mvLDSC.1kg$S_Stand,-1,1),digits = 3),
#     cellnote_size = 16,
#     cellnote_textposition = "middle center",
#     show_dendrogram = c(T,F),
#     xlab = "Trait GWAS",
#     ylab = "Trait GWAS",
#     main = "r<sub>g</sub>",
#     scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
#         low = theme.color$contrastDark3, 
#         mid = "black",
#         high = theme.color$contrastLight3,
#         breaks = c(-1,0,1)
#       ),
#     file=file.path(p$folderpath.plots,paste0("rg_cluster.noimp.png")),
#     width = 1200, height = 1200,
#     hclust_method = "ward.D2"
#     #scale = "row"
#   )
# 
# heatmaply(
#     p$mvLD$covstruct.mvLDSC.1kg.500$S.forPlot,
#     cellnote = round(p$mvLD$covstruct.mvLDSC.1kg.500$S,digits = 3),
#     cellnote_size = 16,
#     cellnote_textposition = "middle center",
#     show_dendrogram = c(T,F),
#     xlab = "Trait GWAS",
#     ylab = "Trait GWAS",
#     main = "cov<sub>g</sub>",
#     scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
#         low = theme.color$contrastDark3, 
#         mid = "black",
#         high = theme.color$contrastLight3,
#         breaks = c(-0.5,0,0.5)
#       ),
#     file=file.path(p$folderpath.plots,paste0("cov_cluster.500.png")),
#     width = 1200, height = 1200,
#     hclust_method = "ward.D2"
#     #scale = "row"
#   )
# 
# heatmaply(
#     clipValues(p$mvLD$covstruct.mvLDSC.1kg.500$S_Stand.forPlot,-1,1),
#     cellnote = round(clipValues(p$mvLD$covstruct.mvLDSC.1kg.500$S_Stand,-1,1),digits = 3),
#     cellnote_size = 16,
#     cellnote_textposition = "middle center",
#     show_dendrogram = c(T,F),
#     xlab = "Trait GWAS",
#     ylab = "Trait GWAS",
#     main = "r<sub>g</sub>",
#     scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
#         low = theme.color$contrastDark3, 
#         mid = "black",
#         high = theme.color$contrastLight3,
#         breaks = c(-1,0,1)
#       ),
#     file=file.path(p$folderpath.plots,paste0("rg_cluster.500.png")),
#     width = 1200, height = 1200,
#     hclust_method = "ward.D2"
#     #scale = "row"
#   )
# 
# heatmaply(
#     p$mvLD$covstruct.mvLDSC.1kg.ssimp.small$S.forPlot,
#     cellnote = round(p$mvLD$covstruct.mvLDSC.1kg.ssimp.small$S.forPlot,digits = 3),
#     cellnote_size = 16,
#     cellnote_textposition = "middle center",
#     show_dendrogram = c(T,F),
#     xlab = "Trait GWAS",
#     ylab = "Trait GWAS",
#     main = "cov<sub>g</sub>",
#     scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
#         low = theme.color$contrastDark3,
#         mid = "black",
#         high = theme.color$contrastLight3,
#         breaks = c(-1,0,1)
#       ),
#     file=file.path(p$folderpath.plots,paste0("cov_cluster.ssimp.png")),
#     width = 1200, height = 1200,
#     hclust_method = "ward.D2"
#     #scale = "row"
#   )


# p$printCorrSimplified(
#   corr = clipValues(p$mvLD$covstruct.mvLDSC.1kg$S_Stand.forPlot,-1,1),
#   #SE = p$mvLD$covstruct.mvLDSC.1kg$S_Stand.SE,
#   filename = file.path(p$folderpath.plots,"rg_simple.png"),
#   is.corr = T
#   )

# p$printCorrSimplified(
#   corr = clipValues(abs(p$mvLD$covstruct.mvLDSC$S_Stand.forPlot),-1,1),
#   #SE = p$mvLD$covstruct.mvLDSC.1kg$S_Stand.SE,
#   filename = file.path(p$folderpath.plots,"rg_abs_simple.png"),
#   is.corr = F,
#   absScale=T
#   )


#new main correlation plot 1A
# cPlot<-ggcorr(data = NULL, cor_matrix = clipValues(p$mvLD$covstruct.mvLDSC.1kg$S_Stand,-1,1), geom = "tile")
# ggsave(filename=file.path(p$folderpath.plots, paste0("rg.large.png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 3)


corrForPlot <- clipValues(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw$S_Stand,-1,1)
pForPlot <- p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw$cov.p.liab
colnames(corrForPlot)<-p$sumstats.sel.ssimp$code.nice
rownames(corrForPlot)<-p$sumstats.sel.ssimp$code.nice
colnames(pForPlot)<-p$sumstats.sel.ssimp$code.nice
rownames(pForPlot)<-p$sumstats.sel.ssimp$code.nice
corrForPlot<-corrForPlot[order(colnames(corrForPlot)),order(colnames(corrForPlot))]
#diag(corrForPlot) <- NA_real_
pForPlot<-pForPlot[order(colnames(pForPlot)),order(colnames(pForPlot))]
png(file.path(p$folderpath.plots,"rg.large.A.png"), width = 9, height = 9, units = 'in', res = 300, family = "Helvetica")
par(xpd=TRUE) #keep labels inside margins
pal<-colorRampPalette(c("#FF6666","#EEEEEE","#6666FF"))
corrplot(corr = corrForPlot, method = "circle", insig='blank', order = "original", p.mat = pForPlot, sig.level = c(0.05), pch.cex = 1.5, tl.cex = 0.8, full_col=FALSE, na.label = "square", na.label.col = "grey30", title = "A: LDSC++ genetic correlations, 1kGP3, number-of-variants defined blocks", tl.pos = 'd', mar=c(0,0,4,0), type = "upper", cl.pos = "r", col = pal(200))$corrPos -> p1 #type = "lower", diag=F, addCoef.col = 'black'
#text(p1$x, p1$y, round(p1$corr, 2))
p1$corr <- ifelse(0==-p1$y + ncol(corrForPlot)-p1$x + 1, NA_real_, p1$corr)
text(p1$x, p1$y, round(p1$corr, 2))
dev.off()

corrForPlot <- clipValues(p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw$S_Stand,-1,1)
pForPlot <- p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw$cov.p.liab
colnames(corrForPlot)<-p$sumstats.sel.ssimp$code.nice
rownames(corrForPlot)<-p$sumstats.sel.ssimp$code.nice
colnames(pForPlot)<-p$sumstats.sel.ssimp$code.nice
rownames(pForPlot)<-p$sumstats.sel.ssimp$code.nice
corrForPlot<-corrForPlot[order(colnames(corrForPlot)),order(colnames(corrForPlot))]
#diag(corrForPlot) <- NA_real_
pForPlot<-pForPlot[order(colnames(pForPlot)),order(colnames(pForPlot))]
png(file.path(p$folderpath.plots,"rg.large.B.png"), width = 9, height = 9, units = 'in', res = 300, family = "Helvetica")
par(xpd=TRUE) #keep labels inside margins
pal<-colorRampPalette(c("#FF6666","#EEEEEE","#6666FF"))
corrplot(corr = corrForPlot, method = "circle", insig='blank', order = "original", p.mat = pForPlot, sig.level = c(0.05), pch.cex = 1.5, tl.cex = 0.8, full_col=FALSE, na.label = "square", na.label.col = "grey30", title = "B: LDSC++ genetic correlations, 1kGP3, recombination distance defined blocks", tl.pos = 'd', mar=c(0,0,4,0), type = "upper", cl.pos = "r", col = pal(200))$corrPos -> p1 #type = "lower", diag=F, addCoef.col = 'black'
#text(p1$x, p1$y, round(p1$corr, 2))
p1$corr <- ifelse(0==-p1$y + ncol(corrForPlot)-p1$x + 1, NA_real_, p1$corr)
text(p1$x, p1$y, round(p1$corr, 2))
dev.off()


newnames <- p$sumstats.sel.ssimp$code.nice
newnamesInfoBin<-p$sumstats.sel.ssimp[p$sumstats.sel.infobin.code,]$code.nice



#covariance plot
batRes<-plotAndTestBatteryForMVLDSC(p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw,doPlotting = F)
corrForPlot <- p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw$S
pForPlot <- batRes$cov.p.fdr2
plot.corr(corr = corrForPlot, pmat = pForPlot, filename = file.path(p$folderpath.plots,"covG.large.png"), is.corr = F, is.full = F, newnames = p$sumstats.sel.ssimp$code.nice, par.col.lim.arg = c(-0.8,0.8))

#covariance CV plot
corrForPlot <- batRes$S.CV
pForPlot <- batRes$cov.p.fdr2
plot.corr(corr = corrForPlot, pmat = pForPlot, filename = file.path(p$folderpath.plots,"covGCV.large.png"), is.corr = F, is.full = F, newnames = p$sumstats.sel.ssimp$code.nice, par.col.lim.arg = c(0,0.8))


#simulation plots
batRes<-plotAndTestBatteryForMVLDSC(p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3.sim,folderpath.plots = p$folderpath.plots, code = "original.hm3.sim", titleTemplate = "original LDSC (Genomic SEM emulated), HM3 reference", titleAddition = "\nBlock Jackknife Resampling")

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01.sim, folderpath.plots = p$folderpath.plots, code = "original.1kg.sim", titleTemplate = "original LDSC (Genomic SEM emulated), 1kGP3 reference", titleAddition = "\nBlock Jackknife Resampling", df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw.sim, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.sim", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01.sim, df.summary=batRes$df.summary)


#Panel comparisons

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.GSEMemulation.hm3, folderpath.plots = p$folderpath.plots, code = "original.hm3", titleTemplate = "original LDSC (Genomic SEM emulated), HM3 reference", titleAddition = "\nBlock Jackknife Resampling", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, folderpath.plots = p$folderpath.plots, code = "original.1kg", titleTemplate = "original LDSC (Genomic SEM emulated), 1kGP3 reference", titleAddition = "\nBlock Jackknife Resampling", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.hm3res, folderpath.plots = p$folderpath.plots, code = "original.1kg.hm3res", titleTemplate = "Original LDSC (Genomic SEM emulated), 1kGP3 reference restricted to HM3 variants", titleAddition = "\nBlock Jackknife Resampling", newnames = newnames, df.summary=batRes$df.summary)

#explicit test of LDSC++ GSEMemulation vs original LDSC - only diagonal LDSC results
mS_LDSC.1kg <- matrix(data=NA_real_,ncol = ncol(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S),nrow = nrow(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S))
colnames(mS_LDSC.1kg)<-colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)
rownames(mS_LDSC.1kg)<-colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)

mS.SE_LDSC.1kg <- matrix(data=NA_real_,ncol = ncol(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S.SE),nrow = nrow(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S.SE))
colnames(mS.SE_LDSC.1kg)<-colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)
rownames(mS.SE_LDSC.1kg)<-colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)

diag(mS_LDSC.1kg)<-p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.LDSC.1kg.maf0_01")]
diag(mS.SE_LDSC.1kg)<-p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S),c("h2.se.LDSC.1kg.maf0_01")]
test.GSEMemulation_vs_originalLDSC<-difftest.matrix(
  mValues1 = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S,
  mStandard_errors1 = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S.SE,
  df1 = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$cov.blocks,
  mValues2 = mS_LDSC.1kg,
  mStandard_errors2 = mS.SE_LDSC.1kg,
  df2 = matrix(200,nrow(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S),ncol(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)),
  effectiveNumberOfTests = getEffectiveNumberOfTests(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$V_Stand)
  )

#average differences
mean((diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)-p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.LDSC.1kg.maf0_01")])/abs(p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.LDSC.1kg.maf0_01")]))

cv.S_StandLimit <- 0.05
S_ForCV<-abs(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)
g_se <- sqrt(diag(S_ForCV))
S_se2 <- g_se %*% t(g_se)
S_ForCV[S_ForCV < S_se2 * cv.S_StandLimit] <- (S_se2 * cv.S_StandLimit)[S_ForCV < S_se2 * cv.S_StandLimit] #cap at min rg 0.05 to avoid extreme values
S.CV1<-(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S.SE/S_ForCV)
h.CV1<-diag(S.CV1)

S_ForCV<-abs(p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.LDSC.1kg.maf0_01")])
g_se <- sqrt(S_ForCV)
S_se2 <- diag(g_se %*% t(g_se))
S_ForCV[S_ForCV < S_se2 * cv.S_StandLimit] <- (S_se2 * cv.S_StandLimit)[S_ForCV < S_se2 * cv.S_StandLimit] #cap at min rg 0.05 to avoid extreme values
h.CV2<-(abs(p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.se.LDSC.1kg.maf0_01")])/abs(p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.LDSC.1kg.maf0_01")]))

mean((h.CV1-h.CV2)/abs(h.CV2))


#average delta covG in percent
mean((diag(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S)-p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.LDSC.1kg.maf0_01")])/p$sumstats[p$sumstats$code %in% colnames(p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01$S) ,c("h2.LDSC.1kg.maf0_01")])

#comparison with ORIGINAL LDSC

#pretend this is a covstruct from ldsc
pretendLDSCcovstruct <- p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01
pretendLDSCcovstruct$S<-mS_LDSC.1kg
pretendLDSCcovstruct<-semplate$amendGsemLDSC(pretendLDSCcovstruct)
pretendLDSCcovstruct$S.SE<-mS.SE_LDSC.1kg
batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, folderpath.plots = p$folderpath.plots, code = "original.1kg.ldsc", titleTemplate = "LDSC++ (Genomic SEM emulated), 1kGP3 reference", titleAddition = "\nBlock Jackknife Resampling", mvldscComparison =pretendLDSCcovstruct,titleAdditionComparison = " vs original LDSC", newnames = newnames, df.summary=batRes$df.summary)


#comparison with Genomic SEM LDSC
batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01, folderpath.plots = p$folderpath.plots, code = "original.1kg.gsem", titleTemplate = "LDSC++ (Genomic SEM emulated), 1kGP3 reference", titleAddition = "\nBlock Jackknife Resampling", mvldscComparison =semplate$amendGsemLDSC(  p$mvLD$covstruct.GSEMmvLDSC.1kg),titleAdditionComparison = " vs original Genomic SEM LDSC", newnames = newnames, df.summary=batRes$df.summary)


#empty comparisons (only variable # blocks) with (emulated) original LDSC (200 fixed blocks)
batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.now", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs LDSC++ (Genomic SEM emulated) 200 blocks", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.r.now", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs LDSC++ (Genomic SEM emulated) 200 blocks", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.now", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs LDSC++ (Genomic SEM emulated) 200 blocks", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock, folderpath.plots = p$folderpath.plots, code = "1kg.jn.r.now", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs LDSC++ (Genomic SEM emulated) 200 blocks", newnames = newnames, df.summary=batRes$df.summary)

#Full comparisons with (emulated) original LDSC

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs original LDSC (Genomic SEM emulated)", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.r", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs original LDSC (Genomic SEM emulated)", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nBlock Jackknife Resampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs original LDSC (Genomic SEM emulated)", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.jn.r", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nBlock Jackknife Resampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.GSEMemulation.1kg.maf0_01,titleAdditionComparison = " vs original LDSC (Genomic SEM emulated)", newnames = newnames, df.summary=batRes$df.summary)

plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs_jn", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw,titleAdditionComparison = " vs Block Jackknife Resampling", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.r_jn", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw,titleAdditionComparison = " vs Block Jackknife Resampling", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v_r", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw,titleAdditionComparison = " vs recombination distance blocks", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v_r", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw,titleAdditionComparison = " vs recombination distance blocks", newnames = newnames, df.summary=batRes$df.summary)

#weighting extensions comparisons, incrementally added extensions

#full
batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.winfo.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.r.winfo.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.winfo.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock,titleAdditionComparison = "vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.jn.r.winfo.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

#winfo only
batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.winfo", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock,titleAdditionComparison = "\ncomparing with not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.winfo, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.r.winfo", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.winfo", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.winfo, folderpath.plots = p$folderpath.plots, code = "1kg.jn.r.winfo", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

#altcw only
batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.r.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.rblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)

batRes <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock.altcw, folderpath.plots = p$folderpath.plots, code = "1kg.jn.r.altcw", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, recombination distance blocks",mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.rblock,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnames, df.summary=batRes$df.summary)


#INFO-bin results

##jn
batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_8_10, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.info_8_10", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, 0.8<INFO<1.0", newnames = newnamesInfoBin)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_6_8, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.info_6_8", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, 0.6<INFO<0.8", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_4_6, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.info_4_6", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, 0.4<INFO<0.6", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_0_4, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.info_0_4", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, 0<INFO<0.4", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_8_10, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.winfo.info_8_10", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, INFO weighting, 0.8<INFO<1.0", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_8_10,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_6_8, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.winfo.info_6_8", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, INFO weighting, 0.6<INFO<0.8", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_6_8,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_4_6, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.winfo.info_4_6", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, INFO weighting, 0.4<INFO<0.6", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_4_6,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock.winfo_info_0_4, folderpath.plots = p$folderpath.plots, code = "1kg.jn.v.winfo.info_0_4", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nblock jackknife resampling, #variants blocks, INFO weighting, 0<INFO<0.4", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_0_4,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

##vbcs
batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_8_10, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.info_8_10", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, 0.8<INFO<1.0", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_6_8, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.info_6_8", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, 0.6<INFO<0.8", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_4_6, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.info_4_6", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, 0.4<INFO<0.6", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_0_4, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.info_0_4", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, 0<INFO<0.4", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_8_10, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.winfo.info_8_10", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, INFO weighting, 0.8<INFO<1.0", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.jn.varblock_info_8_10,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_6_8, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.winfo.info_6_8", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, INFO weighting, 0.6<INFO<0.8", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_6_8,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_4_6, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.winfo.info_4_6", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, INFO weighting, 0.4<INFO<0.6", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_4_6,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)

batResINFO <- plotAndTestBatteryForMVLDSC(mvldsc = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo_info_0_4, folderpath.plots = p$folderpath.plots, code = "1kg.vbcs.v.winfo.info_0_4", titleTemplate = "LDSC++, 1kGP3 reference", titleAddition = "\nVariable Block-Count Sampling, #variants blocks, INFO weighting, 0<INFO<0.4", mvldscComparison = p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock_info_0_4,titleAdditionComparison = " vs not applying weighting extensions", newnames = newnamesInfoBin, df.summary=batResINFO$df.summary)


fwrite(
  x = batRes$df.summary[,c(
     "mDeltaI",
    "mDeltaI.h2",
    "mAbsCovG",
    "mAbsCovG.h2",
   "mDeltaCovG",
   "mDeltaCovG.h2",
   "mRelDeltaCovG",
   "mRelDeltaCovG.h2",
   "maxRelSigDeltaCovG",
   "minRelSigDeltaCovG",
   "numSigDeltaCovG",
   "pWTCovG",
   "mAbsCovGCV",
   "mAbsCovGCV.h2",
   "mDeltaCovGCV",
   "mDeltaCovGCV.h2",
   "mRelDeltaCovGCV",
   "mRelDeltaCovGCV.h2",
   "maxRelSigDeltaCovGCV",
   "minRelSigDeltaCovGCV",
   "numNegDeltaCovGCV",
   "numSigDeltaCovGCV",
   "pWTCovGVar"
              )],
  file = file.path(p$folderpath.plots,"batterySummary.csv"),
  quote = "auto",row.names = TRUE, col.names = TRUE, nThread = p$nThread
  ) 

fwrite(
  x = batResINFO$df.summary[,c(
     "mDeltaI",
    "mDeltaI.h2",
    "mAbsCovG",
    "mAbsCovG.h2",
   "mDeltaCovG",
   "mDeltaCovG.h2",
   "mRelDeltaCovG",
   "mRelDeltaCovG.h2",
   "maxRelSigDeltaCovG",
   "minRelSigDeltaCovG",
   "numSigDeltaCovG",
   "pWTCovG",
   "mAbsCovGCV",
   "mAbsCovGCV.h2",
   "mDeltaCovGCV",
   "mDeltaCovGCV.h2",
   "mRelDeltaCovGCV",
   "mRelDeltaCovGCV.h2",
   "maxRelSigDeltaCovGCV",
   "minRelSigDeltaCovGCV",
   "numNegDeltaCovGCV",
   "numSigDeltaCovGCV",
   "pWTCovGVar"
              )],
  file = file.path(p$folderpath.plots,"batterySummaryINFObin.csv"),
  quote = "auto",row.names = TRUE, col.names = TRUE, nThread = p$nThread
  ) 

#gt-table: Table 2

p$df.summary.table2<-batRes$df.summary[
  c(
    "1kg.vbcs.v",
    "1kg.vbcs.r",
    "1kg.jn.v",
    "1kg.jn.r"
    ),
  c(
    "mDeltaCovG",
    "mDeltaCovG.h2",
   "mRelDeltaCovG",
   "mRelDeltaCovG.h2",
   "maxRelSigDeltaCovG",
   "minRelSigDeltaCovG",
   "mDeltaCovGCV",
   "mDeltaCovGCV.h2",
   "mRelDeltaCovGCV",
   "mRelDeltaCovGCV.h2",
   "maxRelSigDeltaCovGCV",
   "minRelSigDeltaCovGCV",
   "numNegDeltaCovGCV",
   "numSigDeltaCovG",
   "numSigDeltaCovGCV",
   "pWTCovG",
   "pWTCovGVar"
              )] 

p$df.summary.table2$code<-rownames(p$df.summary.table2)
p$df.summary.table2["1kg.vbcs.v","code"]<-"vbcs.v"
p$df.summary.table2["1kg.vbcs.r","code"]<-"vbcs.r"
p$df.summary.table2["1kg.jn.v","code"]<-"jn.v"
p$df.summary.table2["1kg.jn.r","code"]<-"jn.r"

p$df.summary.table2 %>% 
  dplyr::select(
    code,
    #mDeltaCovG,
    #mDeltaCovG.h2,
    mRelDeltaCovG,
    mRelDeltaCovG.h2,
    maxRelSigDeltaCovG,
    minRelSigDeltaCovG,
    numSigDeltaCovG,
    pWTCovG,
    #mDeltaCovGCV,
    #mDeltaCovGCV.h2,
    mRelDeltaCovGCV,
    mRelDeltaCovGCV.h2,
    maxRelSigDeltaCovGCV,
    minRelSigDeltaCovGCV,
    numNegDeltaCovGCV,
    numSigDeltaCovGCV,
    pWTCovGVar
  ) %>%
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    #title = "LDSC++ trait heritabilities (coefficient of variation): method comparison"
    title = ""
  ) %>% cols_label(
    code = "Code",
    #mDeltaCovG = "avg.ΔcovG",
    #mDeltaCovG.h2 = "avg.ΔcovG,h² only",
    mRelDeltaCovG = "avg.ΔcovG\nrel(%)",
    mRelDeltaCovG.h2 = "avg.ΔcovG\nrel(%),h² only",
    maxRelSigDeltaCovG = "max.sig.ΔcovG\nrel(%)",
    minRelSigDeltaCovG = "min.sig.ΔcovG\nrel(%)",
    #mDeltaCovGCV = "avg.ΔCV(covG)",
    #mDeltaCovGCV.h2 = "avg.ΔCV(covG)\nh² only",
    mRelDeltaCovGCV = "avg.ΔCV(covG)\nrel(%)",
    mRelDeltaCovGCV.h2 = "avg.ΔCV(covG)\nrel(%),h² only",
    maxRelSigDeltaCovGCV = "max.sig.ΔCV(covG)\nrel(%)",
    minRelSigDeltaCovGCV = "min.sig.ΔCV(covG)\nrel(%)",
    numNegDeltaCovGCV = "#neg.ΔCV(covG)",
    numSigDeltaCovG = "#sig.ΔcovG", 
    numSigDeltaCovGCV = "#sig.Δvar(covG)",
    pWTCovG = "p(W.T.)\nΔcovG",
    pWTCovGVar = "p(W.T.)\nΔvar(covG)"
  ) %>%
  fmt_number(
    columns = c(
   
   "mRelDeltaCovG",
   "mRelDeltaCovG.h2",
   "maxRelSigDeltaCovG",
   "minRelSigDeltaCovG",
   "mRelDeltaCovGCV",
   "mRelDeltaCovGCV.h2",
   "maxRelSigDeltaCovGCV",
   "minRelSigDeltaCovGCV"
              ),
    decimals = 2
  ) %>%
  fmt_number(
    columns = c(
   #"mDeltaCovG",
   #"mDeltaCovG.h2",
   #"mDeltaCovGCV",
   #"mDeltaCovGCV.h2",
   "pWTCovG",
   "pWTCovGVar"
              ),
    decimals = 3
  ) %>%
  tab_spanner("covG",c(mRelDeltaCovG,
    mRelDeltaCovG.h2,
    maxRelSigDeltaCovG,
    minRelSigDeltaCovG,
    numSigDeltaCovG,
    pWTCovG)) |>
  tab_spanner("covG variance",c(mRelDeltaCovGCV,
    mRelDeltaCovGCV.h2,
    maxRelSigDeltaCovGCV,
    minRelSigDeltaCovGCV,
    numNegDeltaCovGCV,
    numSigDeltaCovGCV,
    pWTCovGVar)) |>
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  ) |>
  tab_source_note("avg.ΔcovG\nrel(%): Average relative difference in covG (%)") |>
  tab_source_note("avg.ΔcovG\nrel(%),h² only: Average relative difference in covG (%), only considering diagonal covG elements (h²)") |>
  tab_source_note("max.sig.ΔcovG\nrel(%): Maximum relative difference in covG (%), only considering statistically significant (q<0.05) differences") |>
  tab_source_note("min.sig.ΔcovG\nrel(%): Minimum relative difference in covG (%), only considering statistically significant (q<0.05) differences") |>
  tab_source_note("#sig.ΔcovG: Number of statistically significant (q<0.05) differences in covG") |>
  tab_source_note("p(W.T.)\nΔcovG: Wilcoxon signed-rank test p-value for test of differences in covG compared to the reference") |>
  tab_source_note("avg.ΔCV(covG)\nrel(%): Average relative difference in covG CV, in percent") |>
  tab_source_note("avg.ΔCV(covG)\nrel(%),h² only: Average relative difference in covG CV (%), only considering diagonal covG elements (h²)") |>
  tab_source_note("max.sig.ΔCV(covG)\nrel(%): Maximum relative difference in covG CV (%), only considering statistically significant (q<0.05) differences") |>
  tab_source_note("min.sig.ΔCV(covG)\nrel(%): Minimum relative difference in covG CV (%), only considering statistically significant (q<0.05) differences") |>
  tab_source_note("#neg.ΔCV(covG): Number of negative differences in covG CV") |>
  tab_source_note("#sig.Δvar(covG): Number of statistically significant (q<0.05) differences in covG variance") |>
  tab_source_note("p(W.T.)\nΔvar(covG): Wilcoxon signed-rank test p-value for test of differences in covG variance compared to the reference")
  

```

# Compute post-analysis statistics and plots per-trait
```{r post-analysis per-trait statistics, fig.width=9, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}

#library(ROCR)

#TODO Add (adjusted? - see the SSimp publication) correlation measures for imputed-genotyped effects, as in the SSimp publication 

#imputed ld-imp
#limits<-c(4,5,6,7)
#ratios<-as.data.frame(matrix(data = NA, nrow = 0, ncol = 0))
for(iTrait in 1:nrow(p$sumstats.sel)){
  #iTrait<-1
  
  if(!file.exists(p$sumstats.sel[iTrait,]$imputedpath.05cm)) next
  
  cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$imputedpath.05cm, nThreads = 5)
  nrow(cSumstats)
  cond.invertedMAF<-cSumstats$FRQ > .5
  cSumstats$MAF<-ifelse(cond.invertedMAF,1-cSumstats$FRQ,cSumstats$FRQ)
  cSumstats <- cSumstats[MAF>0.01,]
  nrow(cSumstats)
  
  if(!any("BETA.I"==colnames(cSumstats))){
    cSumstats[,BETA.I:=BETA][,SE.I:=SE]
  }
  
  if(!any("LDIMP.L2.SUM"==colnames(cSumstats))){
    cSumstats[,LDIMP.L2.SUM:=NA_real_]
  }
  
  cSumstats[,ZZ:=Z*Z][,Z.I:=(BETA.I/SE.I)][,ZZ.I:=(BETA.I/SE.I)^2][,AC:=FRQ*N][,MAC:=MAF*N][,VSNP:=2*FRQ*(1-FRQ)][,NEF:=round((((BETA.I/SE.I)/BETA.I)^2)/VSNP,digits = 0)][,IMPUTED:=((is.finite(LDIMP.L2.SUM) | is.finite(SINFO)) & BETA.I==BETA)]
  cSumstats[is.na(ZZ.I),ZZ.I:=ZZ]
  quantile(cSumstats$ZZ)
  setkeyv(cSumstats,cols = c("CHR","BP"))
  
  #fix faulty SINFO - remove when it has been fixed in supermunge
  if(any("SINFO"==colnames(cSumstats)) & any("LDIMP.K"==colnames(cSumstats))){
    if(nrow(cSumstats[abs(SINFO)>0.1,])>0.01*nrow(cSumstats)){
      #re-compute SINFO
      cSumstats[!(is.na(LDIMP.L2.SUM) & is.na(SINFO)),LDIMP.Q:=LDIMP.K*L2_REF/(LDIMP.L2.SUM)][,SINFO:=as.double(pnorm(q = LDIMP.Q, mean = 1, sd = 1))]
    }
  }
  
  cSumstats[IMPUTED==F & (is.na(LDIMP.L2.SUM) & is.na(SINFO)),GROUP:=1]
  cSumstats[IMPUTED==F & !(is.na(LDIMP.L2.SUM) & is.na(SINFO)),GROUP:=2]
  cSumstats[IMPUTED==T, GROUP:=3]
  cSumstats$GROUP<-as.factor(cSumstats$GROUP)
  nrow(cSumstats[GROUP==1,])
  nrow(cSumstats[GROUP==2,])
  nrow(cSumstats[GROUP==3,])
  nrow(cSumstats[!is.finite(GROUP),])
  cSumstats[GROUP==2,c('d.z','d.beta','d.se') :=list((BETA.I/SE.I)-(BETA/SE),BETA.I-BETA,SE.I-SE)]
  cSumstats[GROUP==2,c('d.z.abs','d.beta.abs','d.se.abs') :=list(abs(d.z),abs(d.beta),abs(d.se))]
  
  cSumstatsPlot <- cSumstats[is.finite(ZZ),] #CHR == 1 #c("SNP","CHR", "IMPUTED","NAL","Z", "ZZ","ZZ.I","SINFO","LDIMP.L2.SUM","LDIMP.K","GROUP",'d.z','d.beta','d.se','d.z.abs','d.beta.abs','d.se.abs')
  cSumstatsPlot$index<-1:nrow(cSumstatsPlot)
  cSumstatsPlot$bucket<-ntile(cSumstatsPlot$index,n = nrow(cSumstatsPlot)/50)
  cSumstatsPlot[, c('iLDIMP.L2.SUM') := list(mean(LDIMP.L2.SUM,na.rm=T)), by = bucket]
  cSumstatsPlot[GROUP==1,c('LDIMP.L2.SUM'):=list(iLDIMP.L2.SUM)] #non-imputed, non-validated
  
  
  #scaling parameter obtained from SCHI04 as
  # cSumstatsStats <- cSumstatsPlot[IMPUTED==F,]
  # scalingParameter.LD_IMP.SUM <- mean(cSumstatsStats$LD_IMP.SUM, na.rm=T) #1039080898
  # scalingParameter.ZZ <- mean(cSumstatsStats$ZZ, na.rm=T) #2.09409
  # scalingParameter<-scalingParameter.LD_IMP.SUM*scalingParameter.ZZ
  # scalingParameter<-1039080898*2.09409
  # 
  # cSumstatsPlot[,LDIMP.SUM.CORR1:=LDIMP.L2.SUM*ZZ]
  # cSumstatsPlot[,LDIMP.SUM.CORR1.scaled:=(LDIMP.L2.SUM*ZZ)/eval(scalingParameter)]
  #cSumstatsPlot$LD_IMP.SUM.CORR1.scaled<-scale(cSumstatsPlot$LD_IMP.SUM.CORR1,center = F)
  #cSumstatsPlot[,SINFO:=pnorm(q = log10(LD_IMP.SUM.CORR1), mean = 0, sd = sd(cSumstatsPlot$LD_IMP.SUM.CORR1,na.rm = T))]
  #cSumstatsPlot[,SINFO:=pnorm(q = LD_IMP.SUM.CORR1.scaled, mean = 1, sd = sd(cSumstatsPlot$LD_IMP.SUM.CORR1.scaled,na.rm = T))]
  #cSumstatsPlot[,SINFO:=exp(LD_IMP.SUM.CORR1.scaled)/(1+exp(LD_IMP.SUM.CORR1.scaled))]
  
 
  
  #new scatterplots of genoptyped vs imputed effects
  
  #total 
  mzz <- mean(cSumstatsPlot$ZZ,na.rm=T)
  mbeta <- mean(abs(cSumstatsPlot$BETA),na.rm=T)
  mse <- mean(cSumstatsPlot$SE,na.rm=T)
  if(any("L2_REF"==colnames(cSumstatsPlot))){
    cPlot <- ggplot(cSumstatsPlot[ZZ>1.5 & ZZ<mzz*20,], aes(x=L2_REF,y=ZZ, color=IMPUTED)) +
      geom_point(alpha = 0.2, size=1) +
      theme_light() +
      ggtitle(paste("Association against LD score: ",p$sumstats.sel[iTrait,]$code.nice)," , Z^2 > 1.5") +
      xlab("LD score") + ylab("Association (Chi-squared)") +
      guides(fill=guide_legend(title="Imputed or non-imputed")) #does not work :(
    ggsave(filename=file.path(p$folderpath.plots, paste0("ZZ_L2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
  }
  
  cPlot <- ggplot(cSumstatsPlot[GROUP==2 & ZZ<mzz*40,], aes(x=Z,y=Z.I)) +
      geom_point(alpha = 0.2, size=1, aes(colour=SINFO)) +
      theme_light() +
      ggtitle(paste("Association against imputed association: ",p$sumstats.sel[iTrait,]$code.nice)) +
      xlab("Association Z") + ylab("Imputed association Z")
  ggsave(filename=file.path(p$folderpath.plots, paste0("Z.I_Z.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
    
  cPlot <- ggplot(cSumstatsPlot[GROUP==2 & ZZ<mzz*40,], aes(x=BETA,y=BETA.I)) +
      geom_point(alpha = 0.2, size=1, aes(colour=SINFO)) +
      theme_light() +
      ggtitle(paste("Unstandardised GWAS regression coefficient against imputed coefficient: ",p$sumstats.sel[iTrait,]$code.nice)) +
      xlab("Regression coefficient") + ylab("Imputed regression coefficient")
    ggsave(filename=file.path(p$folderpath.plots, paste0("BETA.I_BETA",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
    
  cPlot <- ggplot(cSumstatsPlot[GROUP==2 & ZZ<mzz*20,], aes(x=SE,y=SE.I)) +
      geom_point(alpha = 0.2, size=1, aes(colour=SINFO)) +
      theme_light() +
       ggtitle(paste("Unstandardised GWAS regression coefficient standard error against imputed coefficient standard error: ",p$sumstats.sel[iTrait,]$code.nice)) +
      xlab("Regression coefficient s.e.") + ylab("Imputed regression coefficient s.e.")
    ggsave(filename=file.path(p$folderpath.plots, paste0("SE.I_SE",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
    
    cSumstatsPlot[,rel.d.z.abs:=d.z.abs/abs(Z)][!is.finite(rel.d.z.abs), rel.d.z.abs:=NA_real_]
    cVal <- cor(x = cSumstatsPlot$SINFO,y = cSumstatsPlot$rel.d.z.abs, use = "pairwise.complete.obs")
    cPlot <- ggplot(cSumstatsPlot[GROUP==2 & ZZ<mzz*40 & ZZ> 1.5,], aes(x=SINFO,y=rel.d.z.abs)) +
      geom_point(alpha = 0.2, size=1, aes(colour=L2_REF)) +
      theme_light() +
       ggtitle(paste("Relative absolute error in association statistic against imputation quality score: ",p$sumstats.sel[iTrait,]$code.nice," , cor=",round(cVal,4)," , Z^2 >1.5")) +
      xlab("Imputation quality score (SINFO)") + ylab("Absolute error in association statistic Z relative to original Z")
    ggsave(filename=file.path(p$folderpath.plots, paste0("RADZ_SINFO",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
  
  cPlot <- ggplot(cSumstatsPlot[ZZ<mzz*10,], aes(x=Z, group=IMPUTED, fill=IMPUTED)) +
    geom_density(adjust=1.5,alpha = 0.5, linewidth=1) +
    theme_light() +
    ggtitle(paste("Density of GWAS association statistic: ",p$sumstats.sel[iTrait,]$code.nice)) +
      xlab("Association Z") + ylab("Kernel density")
  ggsave(filename=file.path(p$folderpath.plots, paste0("Z.density.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
  
  cPlot <- ggplot(cSumstatsPlot[ZZ<mzz*30 & ZZ>2,], aes(x=abs(Z), group=IMPUTED, fill=IMPUTED)) +
    geom_density(adjust=1.5,alpha = 0.5, size=1) +
    theme_light() +
    ggtitle(paste("Density of GWAS association statistic, high (Z^2>2) absolute associations: ",p$sumstats.sel[iTrait,]$code.nice)) +
    xlab("Association Z") + ylab("Kernel density")
  ggsave(filename=file.path(p$folderpath.plots, paste0("absZ.density.high.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
  
  cPlot <- ggplot(cSumstatsPlot[abs(BETA)<mbeta*20,], aes(x=BETA, group=IMPUTED, fill=IMPUTED)) +
    geom_density(adjust=1.5,alpha = 0.5, size=1) +
    theme_light() +
    ggtitle(paste("Density of GWAS regression coefficient: ",p$sumstats.sel[iTrait,]$code.nice)) +
      xlab("Regression coefficient") + ylab("Kernel density")
  ggsave(filename=file.path(p$folderpath.plots, paste0("BETA.density.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
  
  cPlot <- ggplot(cSumstatsPlot[SE<mse*5,], aes(x=SE, group=IMPUTED, fill=IMPUTED)) +
    geom_density(adjust=1.5,alpha = 0.5, size=1) +
    theme_light() + ggtitle(paste("Density of GWAS regression coefficient standard error: ",p$sumstats.sel[iTrait,]$code.nice)) +
      xlab("Regression coefficient standard error") + ylab("Kernel density")
  ggsave(filename=file.path(p$folderpath.plots, paste0("SE.density.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
  
  
  
  # if(nrow(cSumstats[GROUP==2,])>0){
  #   #construct set with double entries for validated variants
  #   cSumstatsPlot2 <- cSumstats[GROUP==2,.(SNP,ZZ,BETA,SE,IMPUTED=F)]
  #   cSumstatsPlot2 <- rbind(cSumstatsPlot2,cSumstats[GROUP==2,.(SNP,ZZ=ZZ.I,BETA=BETA.I,SE=SE.I,IMPUTED=T)])
  }
  
#stat_summary_bin(fun.y='mean', bins=1000, color='black', size=2, geom='point'))
  
  # cPlot <- ggplot(cSumstatsPlot, aes(x=LDIMP.L2.SUM,y=ZZ, color=GROUP)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #cPlot
  # ggsave(filename=file.path(p$folderpath.plots, paste0("ldsum.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # cPlot <- ggplot(cSumstatsPlot, aes(x=LDIMP.L2.SUM,y=(d.z^2)/ZZ, color=GROUP)) +
  # geom_point(alpha = 0.2, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #cPlot
  # ggsave(filename=file.path(p$folderpath.plots, paste0("ldsum.val.dzz.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # 
  # cPlot <- ggplot(cSumstatsPlot, aes(x=LDIMP.L2.SUM*ZZ,y=ZZ, color=GROUP)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #cPlot
  # ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # cPlot <- ggplot(cSumstatsPlot, aes(x= LDIMP.L2.SUM*ZZ,y=((d.z)^2)/ZZ, color=GROUP)) +
  # geom_point(alpha = 0.2, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #cPlot
  # ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.val.dzz.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  #  cPlot <- ggplot(cSumstatsPlot, aes(x= log10(LDIMP.L2.SUM*ZZ),y=log10(ZZ), color=GROUP)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #cPlot
  # ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.log.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # cPlot <- ggplot(cSumstatsPlot, aes(x= log10(LDIMP.L2.SUM*ZZ),y=((d.z)^2)/ZZ, color=GROUP)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #cPlot
  # ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.val.dzz.log.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # cPlot <- ggplot(cSumstatsPlot, aes(x= log10(LDIMP.L2.SUM*ZZ),y=log10(((d.z)^2)/ZZ), color=GROUP)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #cPlot
  # ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.val.dzz.log.2.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  
  # #add N*VSNP ?
  # cPlot <- ggplot(cSumstatsPlot, aes(x=log10(LDIMP.L2.SUM*L2_REF*sqrt(ZZ.I)),y=log10(d.z.abs/sqrt(ZZ)), color=L2_REF)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # cPlot
  # #ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.separation.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # cPlot <- ggplot(cSumstatsPlot, aes(x= log10(N*VSNP*LDIMP.L2.SUM*L2_REF*sqrt(ZZ.I)),y=log10(sqrt(ZZ)), color=GROUP)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  #   #geom_smooth(method='lm')
  # cPlot
  # #ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.log.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # #for testing
  # cSumstatsPlot<-cSumstatsPlot[CHR==1,]
  # cPlot <- ggplot(cSumstatsPlot, aes(x=L2_REF/VSNP,y=d.z.abs/sqrt(ZZ), color=d.z)) +
  # geom_point(alpha = 0.02, size=3) +
  # stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point')
  # #geom_smooth(method='lm') #formula= y~x
  # cPlot
  # #ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.experimental.",p$sumstats.sel[iTrait,]$code,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 2)
  # 
  # 
  # 
  # 
  # #for testing
  # #cSumstatsPlot<-cSumstatsPlot[CHR==1,]
  # 
  # #comparison between methods
  # cSumstatsPlot[,y:=d.z.abs/sqrt(ZZ)]
  # #setup measurements
  # cSumstatsPlot[,x1:=L2_REF/VSNP] #INFO 1
  # # cSumstatsPlot[,x2:=LDIMP.L2.SUM/(VSNP*LDIMP.K)] #INFO 2
  # # cSumstatsPlot[,x3:=(SE.I^2)/VSNP] #INFO 3
  # cSumstatsPlot[,x2:=LDIMP.L2.SUM*L2_REF] #custom 2
  # cSumstatsPlot[,x3:=LDIMP.L2.SUM*L2_REF*sqrt(ZZ.I)] #custom 3
  # cSumstatsPlot[,x4:=N*VSNP*LDIMP.L2.SUM*L2_REF] #custom 4
  # cSumstatsPlot[,x5:=N*VSNP*LDIMP.L2.SUM*L2_REF*sqrt(ZZ.I)] #custom 5
  # cSumstatsPlot[,x6:=L2_REF/(LDIMP.L2.SUM)] #custom 6
  # cSumstatsPlot[,x7:=LDIMP.K*L2_REF/(LDIMP.L2.SUM)] #custom 7
  # cSumstatsPlot[,x8:=pnorm(q = LDIMP.K*L2_REF/(LDIMP.L2.SUM), mean = 1, sd = 1)] #scaled and 0 -1 transformed custom 7
  # 
  # 
  # ratios.measurement<-as.data.frame(matrix(data = NA, nrow = 0, ncol = 0))
  # for(iMeasurement in 1:8){
  #   #iMeasurement<-8
  #   cCol<-paste0("x",iMeasurement)
  #   cSumstatsPlot$x<-cSumstatsPlot[,..cCol]
  #   
  #   x.cutoff<-mean(cSumstatsPlot$x,na.rm = T)
  #   #x.cutoff<-quantile(cSumstatsPlot$x,c(.75),na.rm = T)
  #   
  #   cSumstatsPlot[GROUP==2 & d.z/sqrt(ZZ) < 1,x.true:=T][GROUP==2 & d.z/sqrt(ZZ) > 1,x.true:=F]
  #   cSumstatsPlot[GROUP==2 & x>eval(x.cutoff),x.positive:=T][GROUP==2 & x<eval(x.cutoff),x.positive:=F]
  #   
  #   #ROC here!!
  #   dfROC<-as.data.frame(cSumstatsPlot[GROUP==2 & ZZ>0.1,.(predictions=x,labels=x.true)])
  #   pred <- prediction(predictions = dfROC$predictions, labels = dfROC$labels)
  #   perf <- performance(pred,"tpr","fpr")
  #   # plot(perf,main = paste0("ROC ",p$sumstats.sel[iTrait,]$code," model ",cCol), colorize=TRUE)
  #   # abline(a = 0, b = 1)
  #   perf <- performance(pred, measure = "auc")
  #   perf@y.values
  #   
  #   #correlation estimate
  #   dfCorr<-as.data.frame(cSumstatsPlot[GROUP==2 & ZZ>0.1,.(x,y)])
  #   estCor<- cor(x = dfCorr$x, y = dfCorr$y)
  #   
  #   nTot <- nrow(cSumstatsPlot[GROUP==2,])
  #   nPositive <- nrow(cSumstatsPlot[GROUP==2 & x.positive,])
  #   nTrue <- nrow(cSumstatsPlot[GROUP==2 & x.true,])
  #   nTruePositive <- nrow(cSumstatsPlot[GROUP==2 & x.true & x.positive,])
  #   nFalsePositive <- nrow(cSumstatsPlot[GROUP==2 & !x.true & x.positive,])
  #   nTrueNegative <- nrow(cSumstatsPlot[GROUP==2 & !x.true & !x.positive,])
  #   nFalseNegative <- nrow(cSumstatsPlot[GROUP==2 & x.true & !x.positive,])
  #   
  #   ratios.measurement[iMeasurement,c("measurement","m","cutoff", "PR","NR","TPR","TNR","FPR","FNR","PPV","NPV","AUC","cor")] <-c(
  #     cCol,
  #     nTot,
  #     x.cutoff,
  #     round(nPositive/nTot,digits = 3),
  #     round((nTot-nPositive)/nTot,digits = 3),
  #     round(nTruePositive/nTrue,digits = 3), #TPR
  #     round(nTrueNegative/(nTot-nTrue),digits = 3), #TNR
  #     round(nFalsePositive/(nTot-nTrue),digits = 3), #FPR
  #     round(nFalseNegative/nTrue,digits = 3), #FNR
  #     round(nTruePositive/(nTruePositive+nFalsePositive),digits = 3), #PPV
  #     round(nTrueNegative/(nTrueNegative+nFalseNegative),digits = 3), #NPV
  #     round(perf@y.values[[1]], digits = 4),
  #     round(estCor, digits = 4)
  #   )
  #   
  #   cPlot <- ggplot(cSumstatsPlot, aes(x=x,y=log10(y), color=d.z)) +
  #   geom_point(alpha = 0.02, size=3) +
  #   stat_summary_bin(fun='mean', bins=100,color='black', size=2, geom='point') +
  #   geom_vline(xintercept = x.cutoff, color="red") +
  #   ylab("log10(Absolute difference in Z/(Z^2))") + 
  #   theme(text = element_text(size=20)) 
  #   #geom_smooth(method='lm') #formula= y~x
  #   #cPlot
  #   ggsave(filename=file.path(p$folderpath.plots, paste0("impqual.measurement.",p$sumstats.sel[iTrait,]$code,".",cCol,".png")), plot = cPlot, width = 297, height = 210, units = "mm", dpi = 320, scale = 1)
  # }
  # 
  # setDT(ratios.measurement)
  # ratios.measurement[ ,x.eval:=round(as.numeric(PPV)*as.numeric(NPV),digits = 4)]
  # 
  # #ratios.measurement[ ,x.eval:=round(as.numeric(PPV)*as.numeric(NPV)*as.numeric(TPR)*as.numeric(TNR)/(as.numeric(FPR)*as.numeric(FNR)),digits = 3)]
  # 
  # #View(ratios.measurement)
  # ratios.measurement
  # fwrite(x = ratios.measurement, file = file.path(p$folderpath.plots,paste0("pnrates.",p$sumstats.sel[iTrait,]$code,".validation.pnrates.tsv")))
  # 
  # for(cLim in limits){
  #   #cLim<-4
  #   nTot <- nrow(cSumstatsPlot[GROUP==2,])
  #   nPositive <- nrow(cSumstatsPlot[GROUP==2 & log10(LDIMP.L2.SUM*ZZ)>cLim,])
  #   nTrue <- nrow(cSumstatsPlot[GROUP==2 & ((d.z)^2)/ZZ < 1,])
  #   nTruePositive <- nrow(cSumstatsPlot[GROUP==2 & log10(LDIMP.L2.SUM*ZZ)>cLim & ((d.z)^2)/ZZ < 1,])
  #   nFalsePositive <- nrow(cSumstatsPlot[GROUP==2 & log10(LDIMP.L2.SUM*ZZ)>cLim & ((d.z)^2)/ZZ > 1,])
  #   nTrueNegative <- nrow(cSumstatsPlot[GROUP==2 & log10(LDIMP.L2.SUM*ZZ)<cLim & ((d.z)^2)/ZZ > 1,])
  #   nFalseNegative <- nrow(cSumstatsPlot[GROUP==2 & log10(LDIMP.L2.SUM*ZZ)<cLim & ((d.z)^2)/ZZ < 1,])
  #   
  #   ratios[iTrait,c("trait","m", paste0("PR.",cLim), paste0("NR.",cLim), paste0("TPR.",cLim), paste0("TNR.",cLim), paste0("FPR.",cLim), paste0("FNR.",cLim))] <-c(
  #     p$sumstats.sel[iTrait,]$code,
  #     nTot,
  #     nPositive/nTot,
  #     (nTot-nPositive)/nTot,
  #     nTruePositive/nTrue, #TPR
  #     nTrueNegative/(nTot-nTrue), #TNR
  #     nFalsePositive/(nTot-nTrue), #FPR
  #     nFalseNegative/nTrue #FNR
  #   )
  # }
 # (ggplot(cSumstatsPlot, aes(x=ZZ,y=NAL,color=IMPUTED)) + geom_point(alpha = 0.4))
  
  
  
  
  #nrow(cSumstatsPlot[ZZ>5,])
  
}

ratios

#fwrite(x = ratios, file = paste0(p$folderpath.plots,"/imputation.validation.pnrates.tsv"))

```




# Other genetic covariancve and imputation results plots
```{r other plots, fig.width=9, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}
require(ROCR)



p$sumstats.sel.table<-p$sumstats.sel[,c("name.nice.and_code",
                                        "chisquare_mean",
                                        "cSumstats_impval_rmse_z",
                                        "cSumstats_impval_bias_z",
                                        "cSumstats_impval_rmse_z_01_05",
                                        "cSumstats_impval_bias_z_01_05",
                                        "cSumstats_impval_rmse_z_05_50",
                                        "cSumstats_impval_bias_z_05_50",
                                        "mean.S_Stand.deltaimpnoimp.500",
                                        "mean.S_Stand.SE.deltaimpnoimp.500",
                                        "variants_ldimp_mean_cell_partial",
                                        "variants_nonimputed_mean_cell"
                                        )]

p$sumstats.sel.table$ratio.ldimp.partial<-p$sumstats.sel.table$variants_ldimp_mean_cell_partial/p$sumstats.sel.table$variants_nonimputed_mean_cell


ggplot(p$sumstats.sel.table, aes(x=chisquare_mean, y=cSumstats_impval_rmse_z)) +
  geom_line(color=theme.color$contrastLight1, size=2, alpha=0.9) +
  geom_label_repel(aes(label = name.nice.and_code), size = 7, nudge_x = 0.3) +
  ggtitle("RMSE Z")

ggplot(p$sumstats.sel.table, aes(x=chisquare_mean, y=mean.S_Stand.deltaimpnoimp.500)) +
  geom_line(color=theme.color$contrastLight1, size=2, alpha=0.9) +
  geom_label_repel(aes(label = name.nice.and_code), size = 7, nudge_x = 0.3) +
  ggtitle("Mean absolute standardised genetic covariance difference, LDimp vs genotyped")


ggplot(p$sumstats.sel.table, aes(x=ratio.ldimp.partial, y=mean.S_Stand.deltaimpnoimp.500)) +
  geom_line(color=theme.color$contrastLight1, size=2, alpha=0.9) +
  geom_label_repel(aes(label = name.nice.and_code), size = 7, nudge_x = 0.3) +
  ggtitle("Mean absolute standardised genetic covariance difference, LDimp vs genotyped, 2")


ggplot(p$sumstats.sel.table, aes(x=chisquare_mean, y=mean.S_Stand.SE.deltaimpnoimp.500)) +
  geom_line(color=theme.color$contrastLight1, size=2, alpha=0.9) +
  geom_label_repel(aes(label = name.nice.and_code), size = 7, nudge_x = 0.3) +
  ggtitle("Mean standardised genetic covariance S.E. difference, LDimp vs genotyped")

#This does not work properly - really needed?
for(iTrait in 1:nrow(p$sumstats.sel)){
  #iTrait<-10
  if(!file.exists(p$sumstats.sel[iTrait,]$imputedpath.500)) next
  cSumstats <- shru::readFile(filePath = p$sumstats.sel[iTrait,]$imputedpath.500, nThreads = 5)

  nrow(cSumstats)
  cond.invertedMAF<-cSumstats$FRQ > .5
  cSumstats$MAF<-ifelse(cond.invertedMAF,1-cSumstats$FRQ,cSumstats$FRQ)
  cSumstats <- cSumstats[MAF>0.01,]
  nrow(cSumstats)
  cSumstats[,ZZ:=Z*Z][,ZZ.I:=(BETA.I/SE.I)^2][,AC:=FRQ*N][,MAC:=MAF*N][,VSNP:=2*FRQ*(1-FRQ)][,NEF:=round((((BETA.I/SE.I)/BETA.I)^2)/VSNP,digits = 0)][,IMPUTED:=(is.finite(LDIMP.L2.SUM) & BETA.I==BETA)]

  cSumstats[,IMPUTED:=(is.finite(LDIMP.L2.SUM) & BETA.I==BETA)][,SINFO:=NA_real_]
  cSumstats$SINFO<-as.double(cSumstats$SINFO)
  cSumstats[IMPUTED==T & is.finite(L2_REF),SINFO:=pnorm(q = LDIMP.K*L2_REF/(LDIMP.L2.SUM), mean = 1, sd = 1)]

  setkeyv(cSumstats,cols = c("CHR","BP"))
  cSumstats$d.z<-NA_real_
  cSumstats[,d.z :=(BETA.I/SE.I)-(BETA/SE)]
  cSumstats[,d.z.abs:=abs(d.z)]

  cSumstats$y<-NA_real_
  cSumstats[,y:=d.z.abs/sqrt(ZZ)][,x:=SINFO]
  cSumstats[,y.true:=ifelse(y<0.5,T,F)]

  #ROC here!!
  dfROC<-as.data.frame(cSumstats[ZZ>0.01579077 & is.finite(x) & is.finite(y),.(predictions=x,labels=y.true)])
  pred <- prediction(predictions = dfROC$predictions, labels = dfROC$labels)
  perf <- performance(pred,"tpr","fpr")
  plot(perf,main = paste0("ROC ",p$sumstats.sel[iTrait,]$code), colorize=TRUE)
  # abline(a = 0, b = 1)
  perf <- performance(pred, measure = "auc")
  perf@y.values
}

```



# Improved annotation of chosen datasets, HM3 validation version

```{r improved annotation of chosen datasets HM3 version, fig.width=10, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}
#View(p$sumstats[p$sumstats.sel.ssimp.code,])


p$sumstats.varcodes.h2<-c(
                                        "h2.LDSC.hm3.plainN",
                                        "h2.LDSC.hm3",
                                        "h2.GSEMmvLDSC.hm3",
                                        "h2.mvLDSC.GSEMemulation.hm3",
                                        "h2.mvLDSC.hm3.jn.varblock",
                                        "h2.mvLDSC.hm3.jn.rblock",
                                        "h2.mvLDSC.hm3.jn.varblock.altcw",
                                        "h2.mvLDSC.hm3.jn.rblock.altcw",
                                        "h2.mvLDSC.hm3.jn.varblock.winfo",
                                        "h2.mvLDSC.hm3.jn.rblock.winfo",
                                        "h2.mvLDSC.hm3.jn.varblock.winfo.altcw",
                                        "h2.mvLDSC.hm3.jn.rblock.winfo.altcw",
                                        "h2.mvLDSC.hm3.vbcs.varblock",
                                        "h2.mvLDSC.hm3.vbcs.rblock",
                                        "h2.mvLDSC.hm3.vbcs.varblock.altcw",
                                        "h2.mvLDSC.hm3.vbcs.rblock.altcw",
                                        "h2.mvLDSC.hm3.vbcs.varblock.winfo",
                                        "h2.mvLDSC.hm3.vbcs.rblock.winfo",
                                        "h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                        "h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                        )
p$sumstats.varcodes.se<-c(
                                        "h2.se.LDSC.hm3.plainN",
                                        "h2.se.LDSC.hm3",
                                        "h2.se.GSEMmvLDSC.hm3",
                                        "h2.se.mvLDSC.GSEMemulation.hm3",
                                        "h2.se.mvLDSC.hm3.jn.varblock",
                                        "h2.se.mvLDSC.hm3.jn.rblock",
                                        "h2.se.mvLDSC.hm3.jn.varblock.altcw",
                                        "h2.se.mvLDSC.hm3.jn.rblock.altcw",
                                        "h2.se.mvLDSC.hm3.jn.varblock.winfo",
                                        "h2.se.mvLDSC.hm3.jn.rblock.winfo",
                                        "h2.se.mvLDSC.hm3.jn.varblock.winfo.altcw",
                                        "h2.se.mvLDSC.hm3.jn.rblock.winfo.altcw",
                                        "h2.se.mvLDSC.hm3.vbcs.varblock",
                                        "h2.se.mvLDSC.hm3.vbcs.rblock",
                                        "h2.se.mvLDSC.hm3.vbcs.varblock.altcw",
                                        "h2.se.mvLDSC.hm3.vbcs.rblock.altcw",
                                        "h2.se.mvLDSC.hm3.vbcs.varblock.winfo",
                                        "h2.se.mvLDSC.hm3.vbcs.rblock.winfo",
                                        "h2.se.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                        "h2.se.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                        )
p$sumstats.varcodes.cv<-c(
                                       "h2.cv.LDSC.hm3.plainN",
                                        "h2.cv.LDSC.hm3",
                                        "h2.cv.GSEMmvLDSC.hm3",
                                        "h2.cv.mvLDSC.GSEMemulation.hm3",
                                        "h2.cv.mvLDSC.hm3.jn.varblock",
                                        "h2.cv.mvLDSC.hm3.jn.rblock",
                                        "h2.cv.mvLDSC.hm3.jn.varblock.altcw",
                                        "h2.cv.mvLDSC.hm3.jn.rblock.altcw",
                                        "h2.cv.mvLDSC.hm3.jn.varblock.winfo",
                                        "h2.cv.mvLDSC.hm3.jn.rblock.winfo",
                                        "h2.cv.mvLDSC.hm3.jn.varblock.winfo.altcw",
                                        "h2.cv.mvLDSC.hm3.jn.rblock.winfo.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock.winfo",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock.winfo",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                        )

p$sumstats.varcodes.combined <- paste0(p$sumstats.varcodes.h2,".combined")

p$sumstats[,p$sumstats.varcodes.cv]<-p$sumstats[,p$sumstats.varcodes.se]/p$sumstats[,p$sumstats.varcodes.h2]

p$sumstats[,p$sumstats.varcodes.combined]<-paste0(unlist(round( p$sumstats[,p$sumstats.varcodes.h2],digits = 2))," (",unlist(round(p$sumstats[,p$sumstats.varcodes.cv],digits = 3)),")")

#select only combined columns
#p$sumstats.sel.table<-p$sumstats.sel.table[,unlist(c("name.nice.and_code",p$sumstats.sel.table.varcodes.combined))]

p$sumstats.sel.table<-p$sumstats[p$sumstats.sel.ssimp.code,]
p$sumstats.sel.table<-p$sumstats.sel.table[order(p$sumstats.sel.table$code.nice),]

#View(p$sumstats.sel.table)
#p$sumstats.sel.table

#selected sumstats, heritabilities combined
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.LDSC.hm3.plainN.combined",
                                                    "h2.LDSC.hm3.combined",
                                                    "h2.GSEMmvLDSC.hm3.combined",
                                                    "h2.mvLDSC.GSEMemulation.hm3.combined"
                                                    
                                                    )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities"
  ) %>% cols_label(
    name.nice.and_code = "Trait"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.hm3.rtf"))

#selected sumstats, method heritabilities
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.LDSC.hm3.plainN",
                                                    "h2.LDSC.hm3",
                                                    "h2.GSEMmvLDSC.hm3",
                                                    "h2.mvLDSC.GSEMemulation.hm3"
                                                    
                                                    )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities"
  ) %>% cols_label(
    name.nice.and_code = "Trait"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.hm3.h2.rtf"))


#selected sumstats, method heritability CVs
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.cv.LDSC.hm3.plainN",
                                        "h2.cv.LDSC.hm3",
                                        "h2.cv.GSEMmvLDSC.hm3",
                                        "h2.cv.mvLDSC.GSEMemulation.hm3"
                                                    
                                                    )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities"
  ) %>% cols_label(
    name.nice.and_code = "Trait"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.hm3.h2cv.rtf"))



#selected sumstats, LDSC++ additions, varblocks heritabilities
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                "h2.mvLDSC.hm3.jn.varblock",
                                        "h2.mvLDSC.hm3.jn.varblock.altcw",
                                        "h2.mvLDSC.hm3.jn.varblock.winfo",
                                        "h2.mvLDSC.hm3.jn.varblock.winfo.altcw",
                                        "h2.mvLDSC.hm3.vbcs.varblock",
                                        "h2.mvLDSC.hm3.vbcs.varblock.altcw",
                                        "h2.mvLDSC.hm3.vbcs.varblock.winfo",
                                        "h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, HM3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.mvLDSC.hm3.jn.varblock = "LDSC++ jn|varblocks",
    h2.mvLDSC.hm3.jn.varblock.altcw = "LDSC++ jn|varblocks|altcw",
    h2.mvLDSC.hm3.jn.varblock.winfo = "LDSC++ jn|varblocks|infow",
    h2.mvLDSC.hm3.vbcs.varblock = "LDSC++ vbcs|varblocks",
    h2.mvLDSC.hm3.vbcs.varblock.altcw = "LDSC++ vbcs|varblocks|altcw",
    h2.mvLDSC.hm3.vbcs.varblock.winfo = "LDSC++ vbcs|varblocks|infow",
    h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw = "LDSC++ vbcs|varblocks|altcw|infow"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.varblock.hm3.h2.rtf"))

p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                "h2.cv.mvLDSC.hm3.jn.varblock",
                                        "h2.cv.mvLDSC.hm3.jn.varblock.altcw",
                                        "h2.cv.mvLDSC.hm3.jn.varblock.winfo",
                                        "h2.cv.mvLDSC.hm3.jn.varblock.winfo.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock.winfo",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, HM3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.cv.mvLDSC.hm3.jn.varblock = "LDSC++ jn|varblocks",
    h2.cv.mvLDSC.hm3.jn.varblock.altcw = "LDSC++ jn|varblocks|altcw",
    h2.cv.mvLDSC.hm3.jn.varblock.winfo = "LDSC++ jn|varblocks|infow",
    h2.cv.mvLDSC.hm3.vbcs.varblock = "LDSC++ vbcs|varblocks",
    h2.cv.mvLDSC.hm3.vbcs.varblock.altcw = "LDSC++ vbcs|varblocks|altcw",
    h2.cv.mvLDSC.hm3.vbcs.varblock.winfo = "LDSC++ vbcs|varblocks|infow",
    h2.cv.mvLDSC.hm3.vbcs.varblock.winfo.altcw = "LDSC++ vbcs|varblocks|altcw|infow"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.varblock.hm3.h2cv.rtf"))



#selected sumstats, LDSC++ additions, rblocks heritabilities
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                "h2.mvLDSC.hm3.jn.rblock",
                                        "h2.mvLDSC.hm3.jn.rblock.altcw",
                                        "h2.mvLDSC.hm3.jn.rblock.winfo",
                                        "h2.mvLDSC.hm3.jn.rblock.winfo.altcw",
                                        "h2.mvLDSC.hm3.vbcs.rblock",
                                        "h2.mvLDSC.hm3.vbcs.rblock.altcw",
                                        "h2.mvLDSC.hm3.vbcs.rblock.winfo",
                                        "h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, HM3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.mvLDSC.hm3.jn.rblock = "LDSC++ jn|rblocks",
    h2.mvLDSC.hm3.jn.rblock.altcw = "LDSC++ jn|rblocks|altcw",
    h2.mvLDSC.hm3.jn.rblock.winfo = "LDSC++ jn|rblocks|infow",
    h2.mvLDSC.hm3.vbcs.rblock = "LDSC++ vbcs|rblocks",
    h2.mvLDSC.hm3.vbcs.rblock.altcw = "LDSC++ vbcs|rblocks|altcw",
    h2.mvLDSC.hm3.vbcs.rblock.winfo = "LDSC++ vbcs|rblocks|infow",
    h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw = "LDSC++ vbcs|rblocks|altcw|infow"
    
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.rblock.hm3.h2.rtf"))

#selected sumstats, LDSC++ additions, rblocks heritability CVs
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                "h2.cv.mvLDSC.hm3.jn.rblock",
                                        "h2.cv.mvLDSC.hm3.jn.rblock.altcw",
                                        "h2.cv.mvLDSC.hm3.jn.rblock.winfo",
                                        "h2.cv.mvLDSC.hm3.jn.rblock.winfo.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock.winfo",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, HM3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.cv.mvLDSC.hm3.jn.rblock = "LDSC++ jn|rblocks",
    h2.cv.mvLDSC.hm3.jn.rblock.altcw = "LDSC++ jn|rblocks|altcw",
    h2.cv.mvLDSC.hm3.jn.rblock.winfo = "LDSC++ jn|rblocks|infow",
    h2.cv.mvLDSC.hm3.vbcs.rblock = "LDSC++ vbcs|rblocks",
    h2.cv.mvLDSC.hm3.vbcs.rblock.altcw = "LDSC++ vbcs|rblocks|altcw",
    h2.cv.mvLDSC.hm3.vbcs.rblock.winfo = "LDSC++ vbcs|rblocks|infow",
    h2.cv.mvLDSC.hm3.vbcs.rblock.winfo.altcw = "LDSC++ vbcs|rblocks|altcw|infow"
    
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.rblock.hm3.h2cv.rtf"))


```


# Improved annotation of chosen datasets, 1kGP3

```{r improved annotation of chosen datasets, fig.width=10, fig.height=6, out.width="1600px", out.height="1000px", eval=FALSE, purl=FALSE}
#View(p$sumstats[p$sumstats.sel.ssimp.code,])


p$sumstats.varcodes.h2<-c(
                                        "h2.LDSC.hm3.plainN",
                                        "h2.LDSC.hm3",
                                        "h2.GSEMmvLDSC.hm3",
                                        "h2.mvLDSC.GSEMemulation.hm3",
                                        "h2.mvLDSC.GSEMemulation.1kg.hm3res",
                                        "h2.mvLDSC.GSEMemulation.1kg.maf0_01",
                                        "h2.mvLDSC.GSEMemulation.1kg.maf0_05",
                                        "h2.mvLDSC.GSEMemulation.1kg.maf0_1",
                                        "h2.LDSC.1kg.maf0_01",
                                        "h2.LDSC.1kg.maf0_05",
                                        "h2.LDSC.1kg.maf0_1",
                                        "h2.mvLDSC.1kg.vbcs.varblock",
                                        "h2.mvLDSC.1kg.vbcs.rblock",
                                        "h2.mvLDSC.1kg.vbcs.varblock.altcw",
                                        "h2.mvLDSC.1kg.vbcs.rblock.altcw",
                                        "h2.mvLDSC.1kg.vbcs.varblock.winfo",
                                        "h2.mvLDSC.1kg.vbcs.rblock.winfo",
                                        "h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw",
                                        "h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw",
                                        "h2.mvLDSC.1kg.jn.varblock",
                                        "h2.mvLDSC.1kg.jn.rblock",
                                        "h2.mvLDSC.1kg.jn.varblock.altcw",
                                        "h2.mvLDSC.1kg.jn.rblock.altcw",
                                        "h2.mvLDSC.1kg.jn.varblock.winfo",
                                        "h2.mvLDSC.1kg.jn.rblock.winfo",
                                        "h2.mvLDSC.1kg.jn.varblock.winfo.altcw",
                                        "h2.mvLDSC.1kg.jn.rblock.winfo.altcw",
                                        "h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                        "h2.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                        )
p$sumstats.varcodes.se<-c(
                                        "h2.se.LDSC.hm3.plainN",
                                        "h2.se.LDSC.hm3",
                                        "h2.se.GSEMmvLDSC.hm3",
                                        "h2.se.mvLDSC.GSEMemulation.hm3",
                                        "h2.se.mvLDSC.GSEMemulation.1kg.hm3res",
                                        "h2.se.mvLDSC.GSEMemulation.1kg.maf0_01",
                                        "h2.se.mvLDSC.GSEMemulation.1kg.maf0_05",
                                        "h2.se.mvLDSC.GSEMemulation.1kg.maf0_1",
                                        "h2.se.LDSC.1kg.maf0_01",
                                        "h2.se.LDSC.1kg.maf0_05",
                                        "h2.se.LDSC.1kg.maf0_1",
                                        "h2.se.mvLDSC.1kg.vbcs.varblock",
                                        "h2.se.mvLDSC.1kg.vbcs.rblock",
                                        "h2.se.mvLDSC.1kg.vbcs.varblock.altcw",
                                        "h2.se.mvLDSC.1kg.vbcs.rblock.altcw",
                                        "h2.se.mvLDSC.1kg.vbcs.varblock.winfo",
                                        "h2.se.mvLDSC.1kg.vbcs.rblock.winfo",
                                        "h2.se.mvLDSC.1kg.vbcs.varblock.winfo.altcw",
                                        "h2.se.mvLDSC.1kg.vbcs.rblock.winfo.altcw",
                                        "h2.se.mvLDSC.1kg.jn.varblock",
                                        "h2.se.mvLDSC.1kg.jn.rblock",
                                        "h2.se.mvLDSC.1kg.jn.varblock.altcw",
                                        "h2.se.mvLDSC.1kg.jn.rblock.altcw",
                                        "h2.se.mvLDSC.1kg.jn.varblock.winfo",
                                        "h2.se.mvLDSC.1kg.jn.rblock.winfo",
                                        "h2.se.mvLDSC.1kg.jn.varblock.winfo.altcw",
                                        "h2.se.mvLDSC.1kg.jn.rblock.winfo.altcw",
                                        "h2.se.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                        "h2.se.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                        )
p$sumstats.varcodes.cv<-c(
                                        "h2.cv.LDSC.hm3.plainN",
                                        "h2.cv.LDSC.hm3",
                                        "h2.cv.GSEMmvLDSC.hm3",
                                        "h2.cv.mvLDSC.GSEMemulation.hm3",
                                        "h2.cv.mvLDSC.GSEMemulation.1kg.hm3res",
                                        "h2.cv.mvLDSC.GSEMemulation.1kg.maf0_01",
                                        "h2.cv.mvLDSC.GSEMemulation.1kg.maf0_05",
                                        "h2.cv.mvLDSC.GSEMemulation.1kg.maf0_1",
                                        "h2.cv.LDSC.1kg.maf0_01",
                                        "h2.cv.LDSC.1kg.maf0_05",
                                        "h2.cv.LDSC.1kg.maf0_1",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock.altcw",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock.altcw",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock.winfo",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock.winfo",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock.winfo.altcw",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock.winfo.altcw",
                                        "h2.cv.mvLDSC.1kg.jn.varblock",
                                        "h2.cv.mvLDSC.1kg.jn.rblock",
                                        "h2.cv.mvLDSC.1kg.jn.varblock.altcw",
                                        "h2.cv.mvLDSC.1kg.jn.rblock.altcw",
                                        "h2.cv.mvLDSC.1kg.jn.varblock.winfo",
                                        "h2.cv.mvLDSC.1kg.jn.rblock.winfo",
                                        "h2.cv.mvLDSC.1kg.jn.varblock.winfo.altcw",
                                        "h2.cv.mvLDSC.1kg.jn.rblock.winfo.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                        "h2.cv.mvLDSC.hm3.vbcs.rblock.winfo.altcw"
                                        )

p$sumstats.varcodes.combined <- paste0(p$sumstats.varcodes.h2,".combined")

p$sumstats[,p$sumstats.varcodes.cv]<-p$sumstats[,p$sumstats.varcodes.se]/p$sumstats[,p$sumstats.varcodes.h2]

p$sumstats[,p$sumstats.varcodes.combined]<-paste0(unlist(round( p$sumstats[,p$sumstats.varcodes.h2],digits = 2))," (",unlist(round(p$sumstats[,p$sumstats.varcodes.cv],digits = 3)),")")

#select only combined columns
#p$sumstats.sel.table<-p$sumstats.sel.table[,unlist(c("name.nice.and_code",p$sumstats.sel.table.varcodes.combined))]

p$sumstats.sel.table<-p$sumstats[p$sumstats.sel.ssimp.code,]
p$sumstats.sel.table<-p$sumstats.sel.table[order(p$sumstats.sel.table$code.nice),]

#View(p$sumstats.sel.table)
#p$sumstats.sel.table

#selected sumstats, method heritabilities combined

p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.LDSC.hm3.plainN.combined",
                                                    "h2.LDSC.hm3.combined",
                                                    "h2.GSEMmvLDSC.hm3.combined",
                                                    "h2.mvLDSC.GSEMemulation.hm3.combined",
                                                    "h2.mvLDSC.GSEMemulation.1kg.hm3res.combined",
                                                    "h2.mvLDSC.GSEMemulation.1kg.maf0_01.combined",
                                                    "h2.mvLDSC.GSEMemulation.1kg.maf0_05.combined",
                                                    "h2.mvLDSC.GSEMemulation.1kg.maf0_1.combined",
                                                    "h2.LDSC.1kg.maf0_01.combined",
                                                    "h2.LDSC.1kg.maf0_05.combined",
                                                    "h2.LDSC.1kg.maf0_1.combined"
                                                    
                                                    )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): method comparison"
  ) %>% cols_label(
    name.nice.and_code = "Trait"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.method.rtf"))

#selected sumstats, method heritabilities
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.LDSC.hm3.plainN",
                                                    "h2.LDSC.hm3",
                                                    "h2.GSEMmvLDSC.hm3",
                                                    "h2.mvLDSC.GSEMemulation.hm3",
                                                    "h2.mvLDSC.GSEMemulation.1kg.hm3res",
                                                    "h2.mvLDSC.GSEMemulation.1kg.maf0_01",
                                                    "h2.mvLDSC.GSEMemulation.1kg.maf0_05",
                                                    "h2.mvLDSC.GSEMemulation.1kg.maf0_1",
                                                    "h2.LDSC.1kg.maf0_01",
                                                    "h2.LDSC.1kg.maf0_05",
                                                    "h2.LDSC.1kg.maf0_1"
                                                    )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): method comparison"
  ) %>% cols_label(
    name.nice.and_code = "Trait"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.method.h2.rtf"))


#selected sumstats, method heritability CVs
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.cv.LDSC.hm3.plainN",
                                                    "h2.cv.LDSC.hm3",
                                                    "h2.cv.GSEMmvLDSC.hm3",
                                                    "h2.cv.mvLDSC.GSEMemulation.hm3",
                                                    "h2.cv.mvLDSC.GSEMemulation.1kg.hm3res",
                                                    "h2.cv.mvLDSC.GSEMemulation.1kg.maf0_01",
                                                    "h2.cv.mvLDSC.GSEMemulation.1kg.maf0_05",
                                                    "h2.cv.mvLDSC.GSEMemulation.1kg.maf0_1",
                                                    "h2.cv.LDSC.1kg.maf0_01",
                                                    "h2.cv.LDSC.1kg.maf0_05",
                                                    "h2.cv.LDSC.1kg.maf0_1"
                                                    
                                                    )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): method comparison"
  ) %>% cols_label(
    name.nice.and_code = "Trait"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.method.h2cv.rtf"))


#selected sumstats, LDSC++ additions, varblocks combined

p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                            #"h2.mvLDSC.GSEMemulation.1kg.maf0_01.combined",
                                                "h2.mvLDSC.1kg.jn.varblock.combined",
                                        "h2.mvLDSC.1kg.jn.varblock.altcw.combined",
                                        "h2.mvLDSC.1kg.jn.varblock.winfo.combined",
                                        "h2.mvLDSC.1kg.jn.varblock.winfo.altcw.combined",
                                        "h2.mvLDSC.1kg.vbcs.varblock.combined",
                                        "h2.mvLDSC.1kg.vbcs.varblock.altcw.combined",
                                        "h2.mvLDSC.1kg.vbcs.varblock.winfo.combined",
                                        "h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw.combined"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, 1kGP3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.mvLDSC.1kg.jn.varblock.combined = "LDSC++ jn|varblocks",
    h2.mvLDSC.1kg.jn.varblock.altcw.combined = "LDSC++ jn|varblocks|altcw",
    h2.mvLDSC.1kg.jn.varblock.winfo.combined = "LDSC++ jn|varblocks|infow",
    h2.mvLDSC.1kg.vbcs.varblock.combined = "LDSC++ vbcs|varblocks",
    h2.mvLDSC.1kg.vbcs.varblock.altcw.combined = "LDSC++ vbcs|varblocks|altcw",
    h2.mvLDSC.1kg.vbcs.varblock.winfo.combined = "LDSC++ vbcs|varblocks|infow",
    h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw.combined = "LDSC++ vbcs|varblocks|altcw|infow"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.varblock.rtf"))

#selected sumstats, LDSC++ additions, varblocks heritabilities
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                            #"h2.mvLDSC.GSEMemulation.1kg.maf0_01",
                                                "h2.mvLDSC.1kg.jn.varblock",
                                        "h2.mvLDSC.1kg.jn.varblock.altcw",
                                        "h2.mvLDSC.1kg.jn.varblock.winfo",
                                        "h2.mvLDSC.1kg.jn.varblock.winfo.altcw",
                                        "h2.mvLDSC.1kg.vbcs.varblock",
                                        "h2.mvLDSC.1kg.vbcs.varblock.altcw",
                                        "h2.mvLDSC.1kg.vbcs.varblock.winfo",
                                        "h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, 1kGP3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.mvLDSC.1kg.jn.varblock = "LDSC++ jn|varblocks",
    h2.mvLDSC.1kg.jn.varblock.altcw = "LDSC++ jn|varblocks|altcw",
    h2.mvLDSC.1kg.jn.varblock.winfo = "LDSC++ jn|varblocks|infow",
    h2.mvLDSC.1kg.vbcs.varblock = "LDSC++ vbcs|varblocks",
    h2.mvLDSC.1kg.vbcs.varblock.altcw = "LDSC++ vbcs|varblocks|altcw",
    h2.mvLDSC.1kg.vbcs.varblock.winfo = "LDSC++ vbcs|varblocks|infow",
    h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw = "LDSC++ vbcs|varblocks|altcw|infow"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.varblock.h2.rtf"))

p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                            #"h2.cv.mvLDSC.GSEMemulation.1kg.maf0_01",
                                                "h2.cv.mvLDSC.1kg.jn.varblock",
                                        "h2.cv.mvLDSC.1kg.jn.varblock.altcw",
                                        "h2.cv.mvLDSC.1kg.jn.varblock.winfo",
                                        "h2.cv.mvLDSC.1kg.jn.varblock.winfo.altcw",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock.altcw",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock.winfo",
                                        "h2.cv.mvLDSC.1kg.vbcs.varblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, 1kGP3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.cv.mvLDSC.1kg.jn.varblock = "LDSC++ jn|varblocks",
    h2.cv.mvLDSC.1kg.jn.varblock.altcw = "LDSC++ jn|varblocks|altcw",
    h2.cv.mvLDSC.1kg.jn.varblock.winfo = "LDSC++ jn|varblocks|infow",
    h2.cv.mvLDSC.1kg.vbcs.varblock = "LDSC++ vbcs|varblocks",
    h2.cv.mvLDSC.1kg.vbcs.varblock.altcw = "LDSC++ vbcs|varblocks|altcw",
    h2.cv.mvLDSC.1kg.vbcs.varblock.winfo = "LDSC++ vbcs|varblocks|infow",
    h2.cv.mvLDSC.1kg.vbcs.varblock.winfo.altcw = "LDSC++ vbcs|varblocks|altcw|infow"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.varblock.h2cv.rtf"))

#selected sumstats, LDSC++ additions, rblocks combined
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                            "h2.mvLDSC.GSEMemulation.1kg.maf0_01.combined",
                                                "h2.mvLDSC.1kg.jn.rblock.combined",
                                        "h2.mvLDSC.1kg.jn.rblock.altcw.combined",
                                        "h2.mvLDSC.1kg.jn.rblock.winfo.combined",
                                        "h2.mvLDSC.1kg.jn.rblock.winfo.altcw.combined",
                                        "h2.mvLDSC.1kg.vbcs.rblock.combined",
                                        "h2.mvLDSC.1kg.vbcs.rblock.altcw.combined",
                                        "h2.mvLDSC.1kg.vbcs.rblock.winfo.combined",
                                        "h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw.combined"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, 1kGP3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.mvLDSC.1kg.jn.rblock.combined = "LDSC++ jn|rblocks",
    h2.mvLDSC.1kg.jn.rblock.altcw.combined = "LDSC++ jn|rblocks|altcw",
    h2.mvLDSC.1kg.jn.rblock.winfo.combined = "LDSC++ jn|rblocks|infow",
    h2.mvLDSC.1kg.vbcs.rblock.combined = "LDSC++ vbcs|rblocks",
    h2.mvLDSC.1kg.vbcs.rblock.altcw.combined = "LDSC++ vbcs|rblocks|altcw",
    h2.mvLDSC.1kg.vbcs.rblock.winfo.combined = "LDSC++ vbcs|rblocks|infow",
    h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw.combined = "LDSC++ vbcs|rblocks|altcw|infow"
    
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.rblock.rtf"))

#selected sumstats, LDSC++ additions, rblocks heritabilities
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                            #"h2.mvLDSC.GSEMemulation.1kg.maf0_01",
                                                "h2.mvLDSC.1kg.jn.rblock",
                                        "h2.mvLDSC.1kg.jn.rblock.altcw",
                                        "h2.mvLDSC.1kg.jn.rblock.winfo",
                                        "h2.mvLDSC.1kg.jn.rblock.winfo.altcw",
                                        "h2.mvLDSC.1kg.vbcs.rblock",
                                        "h2.mvLDSC.1kg.vbcs.rblock.altcw",
                                        "h2.mvLDSC.1kg.vbcs.rblock.winfo",
                                        "h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, 1kGP3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.mvLDSC.1kg.jn.rblock = "LDSC++ jn|rblocks",
    h2.mvLDSC.1kg.jn.rblock.altcw = "LDSC++ jn|rblocks|altcw",
    h2.mvLDSC.1kg.jn.rblock.winfo = "LDSC++ jn|rblocks|infow",
    h2.mvLDSC.1kg.vbcs.rblock = "LDSC++ vbcs|rblocks",
    h2.mvLDSC.1kg.vbcs.rblock.altcw = "LDSC++ vbcs|rblocks|altcw",
    h2.mvLDSC.1kg.vbcs.rblock.winfo = "LDSC++ vbcs|rblocks|infow",
    h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw = "LDSC++ vbcs|rblocks|altcw|infow"
    
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.rblock.h2.rtf"))

#selected sumstats, LDSC++ additions, rblocks heritability CVs
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                            #"h2.cv.mvLDSC.GSEMemulation.1kg.maf0_01",
                                                "h2.cv.mvLDSC.1kg.jn.rblock",
                                        "h2.cv.mvLDSC.1kg.jn.rblock.altcw",
                                        "h2.cv.mvLDSC.1kg.jn.rblock.winfo",
                                        "h2.cv.mvLDSC.1kg.jn.rblock.winfo.altcw",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock.altcw",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock.winfo",
                                        "h2.cv.mvLDSC.1kg.vbcs.rblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods, 1kGP3 panel MAF > 0.01"
  ) %>% cols_label(
     name.nice.and_code = "Trait",
    h2.cv.mvLDSC.1kg.jn.rblock = "LDSC++ jn|rblocks",
    h2.cv.mvLDSC.1kg.jn.rblock.altcw = "LDSC++ jn|rblocks|altcw",
    h2.cv.mvLDSC.1kg.jn.rblock.winfo = "LDSC++ jn|rblocks|infow",
    h2.cv.mvLDSC.1kg.vbcs.rblock = "LDSC++ vbcs|rblocks",
    h2.cv.mvLDSC.1kg.vbcs.rblock.altcw = "LDSC++ vbcs|rblocks|altcw",
    h2.cv.mvLDSC.1kg.vbcs.rblock.winfo = "LDSC++ vbcs|rblocks|infow",
    h2.cv.mvLDSC.1kg.vbcs.rblock.winfo.altcw = "LDSC++ vbcs|rblocks|altcw|infow"
    
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.rblock.h2cv.rtf"))



#final panel comparison with selected extensions, heritabilities
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.LDSC.hm3",
 "h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                                "h2.LDSC.1kg.maf0_01",
 
 "h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ panels, MAF > 0.01"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.panels.h2.rtf"))

#final panel comparison with selected extensions, heritability CVs
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.cv.LDSC.hm3",
 "h2.cv.mvLDSC.hm3.vbcs.varblock.winfo.altcw",
                                                "h2.cv.LDSC.1kg.maf0_01",
 
 "h2.cv.mvLDSC.1kg.vbcs.varblock.winfo.altcw"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ panels, MAF > 0.01"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.panels.h2cv.rtf"))


#final panel comparison with selected extensions, combined
#varblock
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.LDSC.hm3.combined",
 "h2.mvLDSC.hm3.vbcs.varblock.winfo.altcw.combined",
                                                "h2.LDSC.1kg.maf0_01.combined",
 
 "h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw.combined"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ panels"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.varblock.panels.rtf"))


#final method comparison with selected extensions, combined
#varblock
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                    "h2.LDSC.1kg.maf0_01.combined",
                                                    "h2.mvLDSC.GSEMemulation.1kg.maf0_01.combined",
 "h2.mvLDSC.1kg.vbcs.varblock.winfo.altcw.combined",
 "h2.mvLDSC.1kg.vbcs.rblock.winfo.altcw.combined"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "LDSC++ trait heritabilities (coefficient of variation): LDSC++ methods"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.heritabilities.ldscpp.varblock.methods.rtf"))

# 
# #Statistical tests of differences in heritabilities and their standard errors of selected situations
# #heritability differences
# p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
#                                                 "h2.pd.mvLDSC.1kg.vbcs.varblock.winfo.altcw",
#                                                 "h2.pd.mvLDSC.1kg.jn.varblock.winfo.altcw",
#                                                 "h2.pd.mvLDSC.1kg.vbcs.rblock.winfo.altcw",
#                                                 "h2.pd.mvLDSC.1kg.jn.rblock.winfo.altcw",
#                                                 "h2.pd.mvLDSC.1kg.vbcs.varblock.winfo.altcw.vsLDSC",
#                                                 "h2.pd.mvLDSC.1kg.jn.varblock.winfo.altcw.vsLDSC",
#                                                 "h2.pd.mvLDSC.1kg.vbcs.rblock.winfo.altcw.vsLDSC",
#                                                 "h2.pd.mvLDSC.1kg.jn.rblock.winfo.altcw.vsLDSC",
#                                                 "h2.pd.mvLDSC.1kg.jn.varblock.winfo.altcw.vsrmethod",
#                                                 "h2.pd.mvLDSC.1kg.jn.rblock.winfo.altcw.vsrmethod"
#                                                 )] %>% 
#   gt() %>% 
#   #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
#   tab_header(
#     title = "P-values, difference in LDSC++ trait heritabilities, 1kGP3 panel MAF > 0.01"
#   ) %>%
#   tab_style(
#     style = cell_text(size = px(14)),
#     locations = cells_column_labels(everything())       
#   ) %>%
#   tab_style(
#     style = cell_text(size = px(16),weight = "bold"),
#     locations = cells_body(everything())        
#   )
# 
# p$plots.sumstats.sel.table
# 
# gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.dtests.h2.rtf"))

#heritability S.E. differences
p$plots.sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code",
                                                "h2.se.pd.mvLDSC.1kg.vbcs.varblock.winfo.altcw",
                                                "h2.se.pd.mvLDSC.1kg.jn.varblock.winfo.altcw",
                                                "h2.se.pd.mvLDSC.1kg.vbcs.rblock.winfo.altcw",
                                                "h2.se.pd.mvLDSC.1kg.jn.rblock.winfo.altcw",
                                                "h2.se.pd.mvLDSC.1kg.vbcs.varblock.winfo.altcw.vsLDSC",
                                                "h2.se.pd.mvLDSC.1kg.jn.varblock.winfo.altcw.vsLDSC",
                                                "h2.se.pd.mvLDSC.1kg.vbcs.rblock.winfo.altcw.vsLDSC",
                                                "h2.se.pd.mvLDSC.1kg.jn.rblock.winfo.altcw.vsLDSC",
                                                "h2.se.pd.mvLDSC.1kg.jn.varblock.winfo.altcw.vsrmethod",
                                                "h2.se.pd.mvLDSC.1kg.jn.rblock.winfo.altcw.vsrmethod"
                                                )] %>% 
  gt() %>% 
  #fmt_number(columns = vars(chisquare_mean), decimals = 2) %>%
  tab_header(
    title = "P-values, difference in LDSC++ trait heritability S.E., 1kGP3 panel MAF > 0.01"
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.dtests.h2.se.rtf"))

#number of variants
for(cPanel in c("hm3","1kg","hc1kg")){ #"ihc1kg"
  #cPanel<-"hm3"
  cImpstats<-NULL
  if(cPanel == "hm3") {
    cImpstats <- p$mvLD$covstruct.mvLDSC.hm3.vbcs.varblock.winfo.altcw$impstats 
  } else if(cPanel == "1kg") {
    cImpstats <- p$mvLD$covstruct.mvLDSC.1kg.vbcs.varblock.winfo.altcw$impstats
  } else if(cPanel == "hc1kg"){
     cImpstats <- p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw$impstats 
  } else if(cPanel == "ihc1kg") {
    cImpstats <- p$mvLD$covstruct.mvLDSC.hc1kg.vbcs.varblock.winfo.altcw.ldimp$impstats
  }
        
  
  #general variant count: ldimp only
  p$sumstats.sel.table<-cImpstats[,c(
                                          "trait.name",
                                          "medianChiSquare.tot",
                                          "medianChiSquare.imputed",
                                          "medianChiSquare.nonimputed",
                                          "m",
                                          "m.imputed",
                                          "m.merged",
                                          "m.merged.imputed",
                                          "m.cell.mean.nonimputed",
                                          "m.cell.mean.imputed.full",
                                          "m.cell.mean.imputed.partial"
                                          )]
  setDT(p$sumstats.sel.table)
  p$sumstats.sel.table<-p$sumstats.sel.table[,.(
    code=trait.name,
    medianChiSquare.nonimputed=ifelse(is.na(medianChiSquare.nonimputed),
    medianChiSquare.tot,medianChiSquare.nonimputed),
    medianChiSquare.imputed,
    m.merged.nonimputed=ifelse(is.na(m.merged.imputed),m.merged,m.merged-m.merged.imputed),
    m.merged.imputed,
    m.cell.mean.nonimputed,
    m.cell.mean.imputed.full,
    m.cell.mean.imputed.partial)]
  
  cat("\nCurrent panel: ",cPanel)
  
  #medianMerged.m<-5931905 #hard coded - from LDSC++
  medianMerged.m<-median(p$sumstats.sel.table$m.merged.nonimputed,na.rm = T)
  cat("\nMedian merged variant count (non-imputed): ",medianMerged.m)
  cat("\nMean LDSC cell overlap (non-imputed) with the median merged variant count (non-imputed): ",mean(p$sumstats.sel.table$m.cell.mean.nonimputed,na.rm=T)/medianMerged.m)
  cat("\nMean LDSC cell overlap (non-imputed+imputed) with the median merged variant count (non-imputed):",
      mean(p$sumstats.sel.table$m.cell.mean.nonimputed+p$sumstats.sel.table$m.cell.mean.imputed.full+p$sumstats.sel.table$m.cell.mean.imputed.partial,na.rm=T)/medianMerged.m)
  cat("\nRatio of mean imputed variant cell count to nonimputed: ",
      mean(p$sumstats.sel.table$m.cell.mean.imputed.full+p$sumstats.sel.table$m.cell.mean.imputed.partial,na.rm=T)/mean(p$sumstats.sel.table$m.cell.mean.nonimputed,na.rm=T))
  
  
  
  
  p$sumstats.sel.table[p$dsStat$sumstats.meta,on=c("code"),c("variants_raw"):=i.m.genotyped]
  p$sumstats.sel.table[code=="BODYT2MI",variants_raw:=1973473] #i.e. m of WEIG04M
  
  p$sumstats.sel.table[p$sumstats.sel,on=c("code"),c("name.nice.and_code"):=i.name.nice.and_code]
  
  p$sumstats.sel.table[,code:=NULL]
  
  p$sumstats.sel.table<-p$sumstats.sel.table[,c("name.nice.and_code","medianChiSquare.nonimputed","medianChiSquare.imputed","variants_raw","m.merged.nonimputed","m.merged.imputed","m.cell.mean.nonimputed","m.cell.mean.imputed.full","m.cell.mean.imputed.partial")]
  
  p$plots.sumstats.sel.table<-p$sumstats.sel.table %>% 
    gt() %>% 
    fmt_number(columns = vars(
      variants_raw,
      m.merged.nonimputed,
      m.merged.imputed,
      m.cell.mean.nonimputed,
      m.cell.mean.imputed.full,
      m.cell.mean.imputed.partial
      ), decimals = 0) %>%
    fmt_number(columns = vars(
      medianChiSquare.imputed,
      medianChiSquare.nonimputed
      ), decimals = 3) %>%
    tab_header(
      title = paste("GWAS summary statistics genetic variant count: ", cPanel)
    ) %>% cols_label(
      name.nice.and_code  = "Trait",
      medianChiSquare.nonimputed = md("Median Chi<sup>2</sup>, genotyped"),
      medianChiSquare.imputed = md("Median Chi<sup>2</sup>, imputed"),
      variants_raw = md("Original no. variants"),
      m.merged.nonimputed = md("Merged no. variants, genotyped"),
      m.merged.imputed = md("Merged no. variants, imputed"),
      m.cell.mean.nonimputed = md("Mean no. variants, LDSC, genotyped"),
      m.cell.mean.imputed.full = md("Mean no. variants, LDSC, imputed*2"),
      m.cell.mean.imputed.partial = md("Mean no. variants, LDSC, imputed*1")
    ) %>%
    tab_style(
      style = cell_text(size = px(14)),
      locations = cells_column_labels(everything())       
    ) %>%
    tab_style(
      style = cell_text(size = px(16),weight = "bold"),
      locations = cells_body(everything())        
    )
  
  p$plots.sumstats.sel.table
  
  gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/sumstats.sel.table.variants.ldimp.",cPanel,".rtf"))

}


#panel variant counts
p$sumstats.sel.table<-p$dsStat$panel.meta[,c("code","m","m.0_01","m.01_05","m.05_50")]

p$plots.sumstats.sel.table<-p$sumstats.sel.table %>% 
  gt() %>% 
  fmt_number(columns = vars(
    m,
    m.0_01,
    m.01_05,
    m.05_50
    ), decimals = 0) %>%
  tab_header(
    title = "Reference panel genetic variant count"
  ) %>% cols_label(
    code  = "Panel",
    m = md("Total"),
    m.0_01 = md("MAF < 0.01"),
    m.01_05 = md("0.01 < MAF < 0.05"),
    m.05_50 = md("0.05 < MAF < 0.5"),
  ) %>%
  tab_style(
    style = cell_text(size = px(14)),
    locations = cells_column_labels(everything())       
  ) %>%
  tab_style(
    style = cell_text(size = px(16),weight = "bold"),
    locations = cells_body(everything())        
  )

p$plots.sumstats.sel.table

gtsave(data = p$plots.sumstats.sel.table, filename = paste0(p$folderpath.plots,"/panel.table.variants.ldimp.rtf"))

#weighted means scratch for h2 comparisons
#BIPD
weighted.mean(
  c(0.28,
    0.31,
    0.06,
    0.24
    ),
  c(
    (2257+5782),
    (2595+5384),
    (1449+1193),
    (6286+12350)
  )
  )


weighted.mean(
  c(0.03,
    0.03,
    0.07,
    0.01
    ),
  c(
    (2257+5782),
    (2595+5384),
    (1449+1193),
    (6286+12350)
  )
  )

sum(c(
    (2257+5782),
    (2595+5384),
    (1449+1193),
    (6286+12350)
  ))

#NEUR
(0.29-0.147)/qnorm(0.025,lower.tail = F)


```

#TESTs
```{r TEST, eval=FALSE, purl=FALSE}
superm <- supermunge(
  filePaths = file.path(p$folderpath.data.sumstats.cleaned,"SCHI06.gz"),
                     refFilePath = p$filepath.SNPReference.1kg,
                     traitNames = "TEST", writeOutput = F,
                     process=TRUE)
```

