#!/bin/bash
#Load the R module
#module add apps/R/3.6.0
#module add apps/R/3.6.3

srun -p cpu --ntasks 1 --cpus-per-task 6 --mem 32G --pty /bin/bash
#srun -p cpu --ntasks 1 --cpus-per-task 7 --mem 160G --pty /bin/bash #for massive ref-panel and dbSNP testing

#misc
#merge plink datasets
plink --merge-list ~/project/JZ_GED_PHD_C1/working_directory/plink.merge.txt --make-bed --out ~/project/JZ_GED_PHD_C1/working_directory/ref4validation

rsync -avzhpt --progress /mnt/lustre/groups/ukbiobank/sumstats/ /scratch/groups/gwas_sumstats/
#sbatch --time 2-00:00:00 --partition brc,shared --job-name="rsync" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="rsync -avzhpt /mnt/lustre/groups/ukbiobank/sumstats/ /scratch/groups/gwas_sumstats/" --output "rsync.$(date +%Y%m%d).out.txt"
#This command may have added duplicate subfolders under each folder
ls /mnt/lustre/groups/ukbiobank/sumstats | xargs -n 1 -P 4 -I % rsync -avzhpt /mnt/lustre/groups/ukbiobank/sumstats/% /scratch/groups/gwas_sumstats/%
rsync -avzhpt --progress /rosalind_ro/scratch/groups/gwas_sumstats/ /scratch/prj/gwas_sumstats/
sbatch --time 2-00:00:00 --partition cpu --job-name="rsync" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="rsync -avzhpt /rosalind_ro/scratch/groups/gwas_sumstats/ /scratch/prj/gwas_sumstats/" --output "rsync.$(date +%Y%m%d).out.txt"

#Thibault's gcta command
../../JZ_GED_PHD_ADMIN_GENERAL/software/gcta/gcta_v1.94.0Beta_linux_kernel_3_x86_64/gcta --bfile ../data/reference_panel/1KG_eur.plink/g1000_eur --mtcojo-file mtcojo_summary_data.list --ref-ld-chr ../data/ld_scores/eur_w_ld_chr.1KG_Phase3/ --w-ld-chr ../data/ld_scores/eur_w_ld_chr.1KG_Phase3/ --out test_mtcojo_result


#work

sbatch --time 23:59:00 --partition brc,shared --job-name="GSEMGWAS" --ntasks 1 --cpus-per-task 6 --mem-per-cpu 8G --wrap="Rscript setup1.R -l cluster" --output "setup1_$(date +%Y%m%d).out.txt" --error "setup1_$(date +%Y%m%d).err.txt"

sbatch --time 23:59:00 --partition brc,shared --job-name="mvLD.mvLDSC" --ntasks 1 --cpus-per-task 3 --mem-per-cpu 12G --wrap="Rscript setup2.R -t mvLD.mvLDSC -l cluster" --output "setup2.mvLD.mvLDSC.out" --error "setup2.mvLD.mvLDSC.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="mvLD.HDL.piecewise" --ntasks 1 --cpus-per-task 3 --mem 64G --wrap="Rscript setup2.R -t mvLD.HDL.piecewise -l cluster" --output "setup2.mvLD.HDL.piecewise.out" --error "setup2.mvLD.HDL.piecewise.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="mvLD.HDL.jackknife" --ntasks 1 --cpus-per-task 3 --mem 80G --wrap="Rscript setup2.R -t mvLD.HDL.jackknife -l cluster" --output "setup2.mvLD.HDL.jackknife.out" --error "setup2.mvLD.HDL.jackknife.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="mvLD.origHDL" --ntasks 1 --cpus-per-task 3 --mem 80G --wrap="Rscript setup2.R -t mvLD.origHDL -l cluster" --output "setup2.mvLD.origHDL.out" --error "setup2.mvLD.origHDL.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="mvLD.origHDL.liabilityScale" --ntasks 1 --cpus-per-task 3 --mem 90G --wrap="Rscript setup2.R -t mvLD.origHDL.liabilityScale -l cluster" --output "setup2.mvLD.origHDL.liabilityScale.out" --error "setup2.mvLD.origHDL.liabilityScale.err"

sbatch --time 23:59:00 --partition brc,shared --job-name="munge" --ntasks 1 --cpus-per-task 3 --mem-per-cpu 10G --wrap="Rscript setup2.R -t munge -l cluster" --output "setup2.munge.$(date +%Y%m%d).out.txt" --error "setup2.munge.$(date +%Y%m%d).err.txt"

#raiss --gwas 'ALCD03.chr' --ref-folder '/users/k1204688/brc_scratch/Public/1KG_Phase3/All' --ld-folder '../data/eur_w_ld_chr' --zscore-folder '../data/gwas_sumstats/munged' --output-folder '../data/gwas_sumstats/imputed'
#source ../python-venv/bin/activate
raiss --chrom 1 --gwas 'ALCD03' --ref-folder '../data/reference.panel.1KG_Phase3' --ld-folder '../data/eur_w_ld_chr' --zscore-folder '../data/gwas_sumstats/munged/ALCD03.chr' --output-folder '../data/gwas_sumstats/imputed' --l2-regularization 0.01 --eigen-threshold 0.05 --R2-threshold 0.3

#python ../../software/ldsc/ldsc.py --bfile 22 --l2 --ld-wind-cm 1 --out 22
for chr in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23; do sbatch --time 23:59:00 --partition brc,shared --job-name="LDSC" --ntasks 1 --cpus-per-task 3 --mem 15G --wrap="../../software/ldsc/ldsc.py --bfile $chr --l2 --ld-wind-cm 1 --out $chr" --output "ldsc_${chr}_$(date +%Y%m%d).out.txt" --error "ldsc_${chr}_$(date +%Y%m%d).err.txt"; done

#raiss --chrom 1 --gwas 'ALCD03' --ref-folder '../data/reference.panel.1KG_Phase3.CLEANED.EUR.cM' --ld-folder '../data/reference.panel.1KG_Phase3.CLEANED.EUR.cM' --zscore-folder '../data/gwas_sumstats/munged/ALCD03.chr' --output-folder '../data/gwas_sumstats/imputed' --l2-regularization 0.01 --eigen-threshold 0.05 --R2-threshold 0.3
mkdir -p ../data/gwas_sumstats/imputed/ALCD03
for chr in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23; do raiss --chrom $chr --gwas 'ALCD03' --ref-folder '../data/reference.panel.1KG_Phase3.CLEANED.EUR.cM' --ld-folder '../data/reference.panel.1KG_Phase3.CLEANED.EUR.cM' --zscore-folder '../data/gwas_sumstats/munged/ALCD03.chr' --output-folder '../data/gwas_sumstats/imputed/ALCD03' --l2-regularization 0.01 --eigen-threshold 0.05 --R2-threshold 0.3; done

sbatch --time 23:59:00 --partition brc,shared --job-name="impALCD03" --ntasks 1 --cpus-per-task 3 --mem 20G --wrap="source ../python-venv/bin/activate; for chr in {1..23}; do raiss --chrom \$chr --gwas ALCD03 --ref-folder ../data/reference.panel.1KG_Phase3.CLEANED.EUR.cM --ld-folder ../data/reference.panel.1KG_Phase3.CLEANED.EUR.cM --zscore-folder ../data/gwas_sumstats/munged/ALCD03.chr --output-folder ../data/gwas_sumstats/imputed/ALCD03 --l2-regularization 0.01 --eigen-threshold 0.05 --R2-threshold 0.3; done" --output impute.ALCD03.$(date +%Y%m%d).out.txt --error impute.ALCD03.$(date +%Y%m%d).err.txt


#awk -f ../../../../scripts/merge.imputed.dataset.awk z_${dscode}_1.txt z_${dscode}_2.txt z_${dscode}_3.txt z_${dscode}_4.txt z_${dscode}_5.txt z_${dscode}_6.txt z_${dscode}_7.txt z_${dscode}_8.txt z_${dscode}_9.txt z_${dscode}_10.txt z_${dscode}_11.txt z_${dscode}_12.txt z_${dscode}_13.txt z_${dscode}_14.txt z_${dscode}_15.txt z_${dscode}_1.txt z_${dscode}_1.txt z_${dscode}_1.txt z_${dscode}_1.txt z_${dscode}_1.txt z_${dscode}_1.txt z_${dscode}_1.txt
dscode=ALCD03; awk -v prefix="z_${dscode}_" -v suffix=".txt" -f ../../../../scripts/merge.imputed.dataset.awk > $dscode;

sbatch --time 23:59:00 --partition brc,shared --job-name="gsem1" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 0:9999 -l cluster" --output "setup2.cfa.1.out" --error "setup2.cfa.1.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem2" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 10000:19999 -l cluster" --output "setup2.cfa.2.out" --error "setup2.cfa.2.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem3" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 20000:29999 -l cluster" --output "setup2.cfa.3.out" --error "setup2.cfa.3.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem4" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 30000:39999 -l cluster" --output "setup2.cfa.4.out" --error "setup2.cfa.4.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem5" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 40000:49999 -l cluster" --output "setup2.cfa.5.out" --error "setup2.cfa.5.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem6" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 50000:59999 -l cluster" --output "setup2.cfa.6.out" --error "setup2.cfa.6.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem7" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 60000:69999 -l cluster" --output "setup2.cfa.7.out" --error "setup2.cfa.7.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem8" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 70000:79999 -l cluster" --output "setup2.cfa.8.out" --error "setup2.cfa.8.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem9" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 80000:89999 -l cluster" --output "setup2.cfa.9.out" --error "setup2.cfa.9.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem10" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 90000:99999 -l cluster" --output "setup2.cfa.10.out" --error "setup2.cfa.10.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem11" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 100000:109999 -l cluster" --output "setup2.cfa.11.out" --error "setup2.cfa.11.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem12" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 110000:119999 -l cluster" --output "setup2.cfa.12.out" --error "setup2.cfa.12.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="gsem13" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a 120000:131072 -l cluster" --output "setup2.cfa.13.out" --error "setup2.cfa.13.err"

#first run
#for gsemi in `seq 0 999 131072`; do gsemi2=$((gsemi+999)); sbatch --time 23:59:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a $gsemi:$gsemi2 -l cluster" --output "setup2.cfa.$gsemi-$gsemi2.out" --error "setup2.cfa.$gsemi-$gsemi2.err"; done
#second run
for gsemi in `seq 0 1000 32768`; do gsemi2=$((gsemi+1000)); sbatch --time 23:59:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a $gsemi:$gsemi2 -l cluster" --output "setup2.cfa.$gsemi-$gsemi2.out" --error "setup2.cfa.$gsemi-$gsemi2.err"; done
#third run
for gsemi in `seq 0 500 4096`; do gsemi2=$((gsemi+500)); sbatch --time 23:59:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup2.R -t cfa -a $gsemi:$gsemi2 -l cluster" --output "setup2.cfa.$gsemi-$gsemi2.out" --error "setup2.cfa.$gsemi-$gsemi2.err"; done

sbatch --time 23:59:00 --partition brc,shared --job-name="prepgwas" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup2.R -l cluster" --output "setup2.prepgwas.out" --error "setup2.prepgwas.err"

sbatch --time 2-00:00:00 --partition brc,shared --job-name="gsemgwas" --ntasks 1 --cpus-per-task 24 --mem-per-cpu 6G --oversubscribe --wrap="Rscript setup2.R -l cluster" --output "setup2.gsemgwas.$(date +%Y%m%d).out.txt" --error "setup2.gsemgwas.$(date +%Y%m%d).err.txt"

#setup3

sbatch --time 2-00:00:00 --partition brc,shared --job-name="mvLD.mvLDSC" --ntasks 1 --cpus-per-task 3 --mem 64G --wrap="Rscript setup3.R -t mvLD.mvLDSC -l cluster" --output "setup3.mvLD.mvLDSC.out" --error "setup3.mvLD.mvLDSC.err"
sbatch --time 23:59:00 --partition brc,shared --job-name="LDSC" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="sh setup3.ldsc.sh" --output "setup3LDSC.out" --error "setup3LDSC.err"

sbatch --time 23:59:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 3 --mem 40G --wrap="Rscript setup3.R -t munge -l cluster" --output "setup3.munge.$(date +%Y%m%d).out.txt"
sbatch --time 23:00:00 --partition brc,shared --job-name="preplfgwas" --ntasks 1 --cpus-per-task 2 --mem 64G --wrap="Rscript setup3.R -t preplfgwas -l cluster" --output "setup3.lfGWAS.sumstats.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition brc,shared --job-name="mvLD.mvLDSC" --ntasks 1 --cpus-per-task 3 --mem 64G --wrap="Rscript setup3.R -t mvLD.mvLDSC -l cluster" --output "setup3.mvLD.mvLDSC.$(date +%Y%m%d).out.txt"
gsemi=0; gsemi2=$((gsemi+250-1)); sbatch --time 2-00:00:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t cfa -a $gsemi:$gsemi2 -l cluster" --output "setup3.cfa.$gsemi-$gsemi2.out.txt";
for gsemi in `seq 250 250 $((4096+1))`; do gsemi2=$((gsemi+250-1)); sbatch --time 2-00:00:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t cfa -a $gsemi:$gsemi2 -l cluster" --output "setup3.cfa.$gsemi-$gsemi2.out.txt"; done
sbatch --time 23:00:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t cfa -l cluster" --output "setup3.cfa.$(date +%Y%m%d).out.txt"

sbatch --time 2-00:00:00 --partition brc,shared --job-name="lfgwas" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t lfgwas -a 1 -l cluster" --output "setup3.lfgwas.chr1.$(date +%Y%m%d).out.txt"
for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lfgwas" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t lfgwas -a $chr -l cluster" --output "setup3.lfgwas.chr$chr.$(date +%Y%m%d).out.txt"; done

#run lf1 chr1 first as a test
lf=1;for chr in 1; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$lf:$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t lfgwas -a $lf:$chr -l cluster" --output "setup3.lfgwas.F$lf.chr$chr.$(date +%Y%m%d).out.txt"; done
lf=1;for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$lf:$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t lfgwas -a $lf:$chr -l cluster" --output "setup3.lfgwas.F$lf.chr$chr.$(date +%Y%m%d).out.txt"; done
#lf=1;for chr in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do Rscript setup3.R -t lfgwas -a $lf:$chr > setup3.lfgwas.chr$chr.$(date +%Y%m%d).out.txt; done

for chr in 1; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t lfgwas -a $chr -l cluster" --output "setup3.lfgwas.F_ALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup3.R -t lfgwas -a $chr -l cluster" --output "setup3.lfgwas.F_ALL.chr$chr.$(date +%Y%m%d).out.txt"; done

#setup4
sbatch --time 23:59:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup4.R -t munge -l cluster" --output "setup4.munge.$(date +%Y%m%d).out.txt"
Rscript setup4.R -t munge > "setup4.munge.$(date +%Y%m%d).out.txt" #alt
sbatch --time 2-00:00:00 --partition brc,shared --job-name="mvLD" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup4.R -t mvLD -l cluster" --output "setup4.mvLD.$(date +%Y%m%d).out.txt"
sbatch --time 4-00:00:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 3 --mem 32G --wrap="Rscript setup4.R -t cfa -l cluster" --output "setup4.cfa.$(date +%Y%m%d).out.txt"
sbatch --time 4-00:00:00 --partition brc,shared --job-name="preplfgwas" --ntasks 1 --cpus-per-task 3 --mem 64G --wrap="Rscript setup4.R -t preplfgwas -l cluster" --output "setup4.lfGWAS.sumstats.$(date +%Y%m%d).out.txt"
for chr in 1; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup4.R -t lfgwas -a M25-4.74.ML:$chr -l cluster" --output "setup4.lfgwas.M25-4.74.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup4.R -t lfgwas -a M25-4.74.ML:$chr -l cluster" --output "setup4.lfgwas.M25-4.74.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 1; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup4.R -t lfgwas -a M25-7.29.ML:$chr -l cluster" --output "setup4.lfgwas.M25-7.29.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup4.R -t lfgwas -a M25-7.29.ML:$chr -l cluster" --output "setup4.lfgwas.M25-7.29.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 1; do sbatch --time 4-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup4.R -t lfgwas -a M25-10.55.ML:$chr -l cluster" --output "setup4.lfgwas.M25-10.55.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 4-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup4.R -t lfgwas -a M25-10.55.ML:$chr -l cluster" --output "setup4.lfgwas.M25-10.55.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
sbatch --time 2-00:00:00 --partition brc,shared --job-name="mvLD2" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup4.R -t mvLD2 -l cluster" --output "setup4.mvLD2.$(date +%Y%m%d).out.txt"

#setup5
sbatch --time 2-00:00:00 --partition brc,shared --job-name="mvLD" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup5.R -t mvLD -l cluster" --output "setup5.mvLD.$(date +%Y%m%d).out.txt"
sbatch --time 4-00:00:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 3 --mem 32G --wrap="Rscript setup5.R -t cfa -l cluster" --output "setup5.cfa.$(date +%Y%m%d).out.txt"
sbatch --time 4-00:00:00 --partition brc,shared --job-name="lga1" --ntasks 1 --cpus-per-task 3 --mem 24G --wrap="Rscript setup5.R -t lfgwas -a M18-4.64.ML -l cluster" --output "setup5.lfgwas.M18-4.64.ML.FALL.chrALL.$(date +%Y%m%d).out.txt"
sbatch --time 4-00:00:00 --partition brc,shared --job-name="preplfgwas" --ntasks 1 --cpus-per-task 3 --mem 64G --wrap="Rscript setup5.R -t preplfgwas -l cluster" --output "setup5.lfGWAS.sumstats.$(date +%Y%m%d).out.txt"
for chr in 1; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup5.R -t lfgwas -a M18_4_14.C.DWLS:$chr -l cluster" --output "setup5.lfgwas.M18_4_14.C.DWLS.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup5.R -t lfgwas -a M18_4_14.C.DWLS:$chr -l cluster" --output "setup5.lfgwas.M18_4_14.C.DWLS.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
#sbatch --time 4-00:00:00 --partition brc,shared --job-name="lga1" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup5.R -t lfgwas -a M18_4_7.O.DWLS -l cluster" --output "setup5.lfgwas.M18_4_7.O.DWLS.FALL.chr$chr.$(date +%Y%m%d).out.txt" #ran out of memory?

#setup6
sbatch --time 8:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup6.R -t munge -l cluster" --output "setup6.munge.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition brc,shared --job-name="gsem1" --ntasks 1 --cpus-per-task 3 --mem 32G --wrap="Rscript setup6.R -t cfa -a 1 -l cluster" --output "setup6.cfa1.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition brc,shared --job-name="gsem2" --ntasks 1 --cpus-per-task 3 --mem 32G --wrap="Rscript setup6.R -t cfa -a 2 -l cluster" --output "setup6.cfa2.$(date +%Y%m%d).out.txt"
sbatch --time 4-00:00:00 --partition brc,shared --job-name="gsem3" --ntasks 1 --cpus-per-task 3 --mem 32G --wrap="Rscript setup6.R -t cfa -a 3 -l cluster" --output "setup6.cfa3.$(date +%Y%m%d).out.txt"
for chr in 1; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup6.R -t lfgwas -a M25_4_27.COR.ML:$chr -l cluster" --output "setup6.lfgwas.M25_4_27.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup6.R -t lfgwas -a M25_4_27.COR.ML:$chr -l cluster" --output "setup6.lfgwas.M25_4_27.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 1; do sbatch --time 4-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup6.R -t lfgwas -a M25_7_195.COR.ML:$chr -l cluster" --output "setup6.lfgwas.M25_7_195.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22; do sbatch --time 4-00:00:00 --partition brc,shared --job-name="lga1_$chr" --ntasks 1 --cpus-per-task 2 --mem 16G --wrap="Rscript setup6.R -t lfgwas -a M25_7_195.COR.ML:$chr -l cluster" --output "setup6.lfgwas.M25_7_195.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done

sbatch --time 2-00:00:00 --partition brc,shared --job-name="nmeta" --ntasks 1 --cpus-per-task 4 --mem 16G --wrap="Rscript setup6.R -t nmeta -a M25_7_26.ORT.ML:5 -l cluster" --output "setup6.nmeta.M25_7_26.ORT.ML:5.$(date +%Y%m%d).out.txt"
sbatch --time 3-00:00:00 --partition brc,shared --job-name="nmeta" --ntasks 1 --cpus-per-task 4 --mem 16G --wrap="Rscript setup6.R -t nmeta -a M25_7_26.ORT.ML:6 -l cluster" --output "setup6.nmeta.M25_7_26.ORT.ML:6.$(date +%Y%m%d).out.txt"
sbatch --time 3-00:00:00 --partition brc,shared --job-name="nmeta" --ntasks 1 --cpus-per-task 4 --mem 16G --wrap="Rscript setup6.R -t nmeta -a M25_10_244.ORT.ML:1 -l cluster" --output "setup6.nmeta.M25_10_244.ORT.ML:1.$(date +%Y%m%d).out.txt"

#setup7
#munging with the new basic ld-imputation procedure. slow, running datasets in parallell.
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a ADHD05 -l cluster" --output "setup7.munge.ADHD05.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a ALCD03 -l cluster" --output "setup7.munge.ALCD03.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a ANOR02 -l cluster" --output "setup7.munge.ANOR02.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a ANXI03 -l cluster" --output "setup7.munge.ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a ANXI04 -l cluster" --output "setup7.munge.ANXI04.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a AUTI07 -l cluster" --output "setup7.munge.AUTI07.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a BIPO02 -l cluster" --output "setup7.munge.BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a DEPR05 -l cluster" --output "setup7.munge.DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a DEPR08 -l cluster" --output "setup7.munge.DEPR08.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a EDUC03 -l cluster" --output "setup7.munge.EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a HEAL01 -l cluster" --output "setup7.munge.HEAL01.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a INCO03 -l cluster" --output "setup7.munge.INCO03.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a INSO02 -l cluster" --output "setup7.munge.INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a NEUR02 -l cluster" --output "setup7.munge.NEUR02.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a PTSD04 -l cluster" --output "setup7.munge.PTSD04.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a RISK02 -l cluster" --output "setup7.munge.RISK02.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a RISK03 -l cluster" --output "setup7.munge.RISK03.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a SCHI04 -l cluster" --output "setup7.munge.SCHI04.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a SUBJ01 -l cluster" --output "setup7.munge.SUBJ01.$(date +%Y%m%d).out.txt"
sbatch --time 24:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 24G --wrap="Rscript setup7.R -t munge -a TIRE01 -l cluster" --output "setup7.munge.TIRE01.$(date +%Y%m%d).out.txt"
#produce summary table when all the above have been munged and imputed
sbatch --time 9:00:00 --partition brc,shared --job-name="smunge" --ntasks 1 --cpus-per-task 4 --mem 48G --wrap="Rscript setup7.R -t munge -l cluster" --output "setup7.munge.$(date +%Y%m%d).out.txt"
sbatch --time 10:00:00 --partition brc,shared --job-name="pre" --ntasks 1 --cpus-per-task 4 --mem 32G --wrap="Rscript setup7.R -t pre -l cluster" --output "setup7.pre.$(date +%Y%m%d).out.txt"
for chr in 22; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t pre -a MPRE_2_1.COR.ML:$chr -l cluster" --output "setup7.lfgwas.MPRE_2_1.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 21; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t pre -a MPRE_2_1.COR.ML:$chr -l cluster" --output "setup7.lfgwas.MPRE_2_1.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 20; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t pre -a MPRE_2_1.COR.ML:$chr -l cluster" --output "setup7.lfgwas.MPRE_2_1.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 15 16 17 18 19; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t pre -a MPRE_2_1.COR.ML:$chr -l cluster" --output "setup7.lfgwas.MPRE_2_1.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 1 2 3 4 5 6 7 8 9 10 11 12 13 14; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t pre -a MPRE_2_1.COR.ML:$chr -l cluster" --output "setup7.lfgwas.MPRE_2_1.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done

sbatch --time 12:00:00 --partition brc,shared --job-name="mvLD" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup7.R -t mvLD -l cluster" --output "setup7.mvLD.$(date +%Y%m%d).out.txt"
sbatch --time 1-12:00:00 --partition brc,shared --job-name="gsem" --ntasks 1 --cpus-per-task 4 --mem 32G --wrap="Rscript setup7.R -t cfa -a 1 -l cluster" --output "setup7.cfa.$(date +%Y%m%d).out.txt"
#on activating implicit multithreading
#https://hpc.nih.gov/apps/R.html#threading
#for chr in 1; do sbatch --time 4-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="OMP_NUM_THREADS=3; Rscript setup7.R -t lfgwas -a M20_6_198.COR.ML:$chr:1 -l cluster" --output "setup7.lfgwas.M20_6_198.COR.ML.F1.chr$chr.$(date +%Y%m%d).out.txt"; done
#test
#for chr in 22; do sbatch --time 1:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t lfgwas -a M20_6_198.COR.ML:$chr -l cluster" --output "setup7.lfgwas.M20_6_198.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done

for chr in 20 21 22; do sbatch --time 6-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t lfgwas -a M20_6_172.COR.ML:$chr -l cluster" --output "setup7.lfgwas.M20_6_172.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19; do sbatch --time 6-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t lfgwas -a M20_6_172.COR.ML:$chr -l cluster" --output "setup7.lfgwas.M20_6_172.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for chr in 1 2 3 4; do sbatch --time 6-00:00:00 --partition brc,shared --job-name="lga_$chr" --ntasks 1 --cpus-per-task 10 --mem 32G --wrap="Rscript setup7.R -t lfgwas -a M20_6_172.COR.ML:$chr -l cluster" --output "setup7.lfgwas.M20_6_172.COR.ML.FALL.chr$chr.$(date +%Y%m%d).out.txt"; done
for fac in 1 2 3 4 5 6; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="nm$fac" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup7.R -t nmeta -a M20_6_172.COR.ML:$fac -l cluster" --output "setup7.nmeta.M20_6_172.COR.ML:$fac.$(date +%Y%m%d).out.txt"; done
for fac in 1 2 3 4 5 6; do sbatch --time 2-00:00:00 --partition brc,shared --job-name="nm$fac" --ntasks 1 --cpus-per-task 3 --mem 16G --wrap="Rscript setup7.R -t nmeta -a M20_6_30.ORT.ML:$fac -l cluster" --output "setup7.nmeta.M20_6_30.ORT.ML:$fac.$(date +%Y%m%d).out.txt"; done

sbatch --time 2-00:00:00 --partition brc,shared --job-name="mvLD2" --ntasks 1 --cpus-per-task 4 --mem 64G --wrap="Rscript setup7.R -t mvLD2 -l cluster" --output "setup7.mvLD2.$(date +%Y%m%d).out.txt"

#MAGMA workshop:
#https://ibg.colorado.edu/cdrom2021/Day10-posthuma/magma_session/instructions.txt

while read code; do
  magma --annotate window=35,10 --snp-loc ../data/gwas_sumstats/export/"$code".SNPLOC --gene-loc ../data/gene_mapping/NCBI37.3.gene.loc --out setup7.annotate."$code";
  magma --bfile ../data/reference_panel/1KG_eur.plink/g1000_eur --gene-annot setup7.annotate."$code".genes.annot --pval ../data/gwas_sumstats/export/"$code".PVAL ncol=N --out setup7.geneanalysis."$code";
done < sumstats.sel.set1.code.txt

for code in SCHI04 SUBJ01 TIRE01; do
  magma --annotate window=35,10 --snp-loc ../data/gwas_sumstats/export/"$code".SNPLOC --gene-loc ../data/gene_mapping/NCBI37.3.gene.loc --out setup7.annotate."$code";
  magma --bfile ../data/reference_panel/1KG_eur.plink/g1000_eur --gene-annot setup7.annotate."$code".genes.annot --pval ../data/gwas_sumstats/export/"$code".PVAL ncol=N --out setup7.geneanalysis."$code";
done

for code in GSEM.F4 GSEM.F5 GSEM.F6; do
  magma --annotate window=35,10 --snp-loc ../data/gwas_sumstats/export/"$code".SNPLOC --gene-loc ../data/gene_mapping/NCBI37.3.gene.loc --out setup7.annotate."$code";
  magma --bfile ../data/reference_panel/1KG_eur.plink/g1000_eur --gene-annot setup7.annotate."$code".genes.annot --pval ../data/gwas_sumstats/export/"$code".PVAL ncol=N --out setup7.geneanalysis."$code";
done

for code in HEAL01 INCO03 INSO02 NEUR02 PTSD04 RISK02 RISK03 SCHI04 SUBJ01 TIRE01 GSEM.F1 GSEM.F2 GSEM.F3 GSEM.F4 GSEM.F5 GSEM.F6; do
  magma --annotate window=35,10 --snp-loc ../data/gwas_sumstats/export/"$code".SNPLOC --gene-loc ../data/gene_mapping/NCBI37.3.gene.loc --out setup7.annotate."$code";
  magma --bfile ../data/reference_panel/1KG_eur.plink/g1000_eur --gene-annot setup7.annotate."$code".genes.annot --pval ../data/gwas_sumstats/export/"$code".PVAL ncol=N --out setup7.geneanalysis."$code";
done

while read code; do
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --gene-covar geneset.gtex8_rnaseq_median.magma --out setup7.geneset.gtex8_rnaseq_median.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --gene-covar geneset.brainspan_rnaseq_mean.magma --out setup7.geneset.brainspan_rnaseq_mean.$code";
  #sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.v7.5.1.magma --out setup7.geneset.msigdb.v7.5.1.$code";
done < sumstats.sel.set1.code.txt

while read code; do
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.h.all.v7.5.1.magma --out setup7.geneset.msigdb.h.all.v7.5.1.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c3.all.v7.5.1.magma --out setup7.geneset.msigdb.c3.all.v7.5.1.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c5.all.v7.5.1.magma --out setup7.geneset.msigdb.c5.all.v7.5.1.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c7.all.v7.5.1.magma --out setup7.geneset.msigdb.c7.all.v7.5.1.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c8.all.v7.5.1.magma --out setup7.geneset.msigdb.c8.all.v7.5.1.$code";
done < sumstats.sel.set1.code.txt

while read code; do
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c2.cp.biocarta.v7.5.1.magma --out setup7.geneset.msigdb.c2.cp.biocarta.v7.5.1.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c2.cp.kegg.v7.5.1.magma --out setup7.geneset.msigdb.c2.cp.kegg.v7.5.1.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c3.mir.mirdb.v7.5.1.magma --out setup7.geneset.msigdb.c3.mir.mirdb.v7.5.1.$code";
  sbatch --time 20:00 --partition brc,shared --job-name="m$code" --ntasks 1 --cpus-per-task 2 --mem 4G --wrap="magma --gene-results setup7.geneanalysis.$code.genes.raw --set-annot geneset.msigdb.c3.tft.gtrd.v7.5.1.magma --out setup7.geneset.msigdb.c3.tft.gtrd.v7.5.1.$code";
done < sumstats.sel.set1.code.txt

#awk 'echo $0;' < sumstats.sel.set1.code.txt


#setup8

#Simulations
#simulated gwas - run in gwas_sumstats/raw

#ldak simulations - NOT USED!!
batch --time 2-00:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 512G --wrap="ldak --make-phenos bespoke --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --her 0.2 --power -1 --causals file1 --effects file2" --output "setup8.sim1.020.$(date +%Y%m%d).out.txt"

#gcta64  --bfile ../data/reference_panel/1kg_bed/1KG_Phase3.WG.CLEANED.EUR_MAF001.CM23  --simu-qt --simu-hsq 0.5 --simu-rep 1  --out test
#--simu-causal-loci causal.snplist
#--keep test.indi.list
#new - specify the h2 assuming GCTA keeps this promise also accounting for LD (papers says it does)
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4987770

sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects1.020.100 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait1.020.100 --thread-num 10" --output "setup8.sim1.020.100.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects2.020.100 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait2.020.100 --thread-num 10" --output "setup8.sim2.020.100.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects1.020.75 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait1.020.75 --thread-num 10" --output "setup8.sim1.020.75.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects2.020.75 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait2.020.75 --thread-num 10" --output "setup8.sim2.020.75.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects1.020.50 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait1.020.50 --thread-num 10" --output "setup8.sim1.020.50.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects2.020.50 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait2.020.50 --thread-num 10" --output "setup8.sim2.020.50.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects1.020.25 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait1.020.25 --thread-num 10" --output "setup8.sim1.020.25.$(date +%Y%m%d).out.txt"
sbatch --time 12:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 128G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects2.020.25 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait2.020.25 --thread-num 10" --output "setup8.sim2.020.25.$(date +%Y%m%d).out.txt"

#name test
#sbatch --time 2-00:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 10 --mem 512G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects1p020p40 --simu-qt --simu-hsq 0.2 --simu-rep 1 --out simTrait1.020.40 --thread-num 10" --output "setup8.sim1.020.40.$(date +%Y%m%d).out.txt"


#old - specifying full estimated h2 not accounting for LD
# sbatch --time 2-00:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 5 --mem 300G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects1.ld --simu-qt --simu-hsq 0.3855130 --simu-rep 1 --out simTrait1.020.ld" --output "setup8.sim1.020.$(date +%Y%m%d).out.txt"
# sbatch --time 2-00:00:00 --partition cpu --job-name="sim" --ntasks 1 --cpus-per-task 5 --mem 300G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --simu-causal-loci ~/project/JZ_GED_PHD_C1/working_directory/simEffects2.ld --simu-qt --simu-hsq 0.3852309 --simu-rep 1 --out simTrait2.020.ld" --output "setup8.sim2.020.$(date +%Y%m%d).out.txt"
#template
#sbatch --time 24:00:00 --partition cpu --job-name="sim010" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="module add gcta; gcta-1.94.1 --bfile ../../reference_panel/hc1kgp3.b38.plink/1kGP_high_coverage_Illumina.filtered.SNV_INDEL_SV_phased_panel.frq.CM.eur --simu-qt --simu-causal-loci neuro02.sim.varlist --simu-hsq 0.10 --out sim010;" --output "setup8.sim010.$(date +%Y%m%d).out.txt"

#test with gcta greml - redundant? - we get the h2 from the GWAS below - no these give different results (based on the full GRM rather than sparse?)
sbatch --time 10:00:00 --partition cpu --job-name="greml1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1.020.100.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim1.020.100" --output fastGWA.sim1.020.100.reml.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="greml2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait2.020.100.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim2.020.100" --output fastGWA.sim2.020.100.reml.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="greml1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1.020.75.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim1.020.75" --output fastGWA.sim1.020.75.reml.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="greml2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait2.020.75.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim2.020.75" --output fastGWA.sim2.020.75.reml.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="greml1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1.020.50.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim1.020.50" --output fastGWA.sim1.020.50.reml.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="greml2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait2.020.50.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim2.020.50" --output fastGWA.sim2.020.50.reml.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="greml1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1.020.25.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim1.020.25" --output fastGWA.sim1.020.25.reml.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="greml2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait2.020.25.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --reml --out sim2.020.25" --output fastGWA.sim2.020.25.reml.$(date +%Y%m%d).out.txt

#bivariate gcta greml - needs a lot of time to run 8-10h (more last time I tested)
sbatch --time 2-00:00:00 --partition cpu --job-name="greml12" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --reml-bivar --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1and2.020.100.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1and2.020.100" --output fastGWA.sim1and2.020.100.reml.$(date +%Y%m%d).out.txt
sbatch --time 2-00:00:00 --partition cpu --job-name="greml12" --ntasks 1 --cpus-per-task 10 --mem 128 --wrap="gcta64 --reml-bivar --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1and2.020.75.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1and2.020.75" --output fastGWA.sim1and2.020.75.reml.$(date +%Y%m%d).out.txt
sbatch --time 2-00:00:00 --partition cpu --job-name="greml12" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --reml-bivar --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1and2.020.50.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1and2.020.50" --output fastGWA.sim1and2.020.50.reml.$(date +%Y%m%d).out.txt
sbatch --time 2-00:00:00 --partition cpu --job-name="greml12" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --reml-bivar --grm /users/k19049801/project/personality-SEM-2021/work/UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm --pheno simTrait1and2.020.25.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1and2.020.25" --output fastGWA.sim1and2.020.25.reml.$(date +%Y%m%d).out.txt

#run the gwases
#generate the covar and keep files first
#--h2-limit 1.0
sbatch --time 10:00:00 --partition cpu --job-name="gas1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait1.020.100.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1.020.100.assoc" --output fastGWA.sim1.020.100.assoc.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="gas2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait2.020.100.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim2.020.100.assoc" --output fastGWA.sim2.020.100.assoc.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="gas1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait1.020.75.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1.020.75.assoc" --output fastGWA.sim1.020.75.assoc.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="gas2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait2.020.75.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim2.020.75.assoc" --output fastGWA.sim2.020.75.assoc.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="gas1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait1.020.50.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1.020.50.assoc" --output fastGWA.sim1.020.50.assoc.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="gas2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait2.020.50.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim2.020.50.assoc" --output fastGWA.sim2.020.50.assoc.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="gas1" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait1.020.25.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim1.020.25.assoc" --output fastGWA.sim1.020.25.assoc.$(date +%Y%m%d).out.txt
sbatch --time 10:00:00 --partition cpu --job-name="gas2" --ntasks 1 --cpus-per-task 10 --mem 64G --wrap="gcta64 --bfile /users/k19049801/project/JZ_GED_PHD_C1/data/geno/GLADv3_EDGIv1_NBRv2/imputed/bfiles/GLAD_EDGI_NBR --grm-sparse UKB_GLADv3_EDGIv1_NBRv2.LD_Pruned.grm.sparse --fastGWA-mlm --pheno simTrait2.020.25.phen --qcovar covar.q.txt --covar covar.c.txt --keep keep.all.txt --thread-num 10 --out sim2.020.25.assoc" --output fastGWA.sim2.020.25.assoc.$(date +%Y%m%d).out.txt

# #munge for LDSC++ etc - NEEDS UPDATING WITH NEW TRAIT CODES - OLD
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a ANXI03 -l cluster" --output "setup8.munge.ANXI03.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a BIPO02 -l cluster" --output "setup8.munge.BIPO02.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a BODY14 -l cluster" --output "setup8.munge.BODY14.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a BODYTM -l cluster" --output "setup8.munge.BODYTM.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a BODYTF -l cluster" --output "setup8.munge.BODYTF.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a DEPR05 -l cluster" --output "setup8.munge.DEPR05.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a DEPR5B -l cluster" --output "setup8.munge.DEPR5B.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a DEPR09 -l cluster" --output "setup8.munge.DEPR09.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a EDUC03 -l cluster" --output "setup8.munge.EDUC03.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a HEIG04M -l cluster" --output "setup8.munge.HEIG04M.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a HEIG04F -l cluster" --output "setup8.munge.HEIG04F.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a HEIG09 -l cluster" --output "setup8.munge.HEIG09.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a INSO02 -l cluster" --output "setup8.munge.INSO02.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a NEUR02 -l cluster" --output "setup8.munge.NEUR02.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a SCHI04 -l cluster" --output "setup8.munge.SCHI04.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a WEIG04M -l cluster" --output "setup8.munge.WEIG04M.$(date +%Y%m%d).out.txt"
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t munge -a WEIG04F -l cluster" --output "setup8.munge.WEIG04F.$(date +%Y%m%d).out.txt"

#munge for original Python LDSC
#test without slurm
#/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/BODY14.gz --out BODY14 --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 50000000

#cleaning of missing sumstats
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/scratch/prj/gwas_sumstats/new/GIANT_Randall2013PlosGenet_stage1_publicrelease_HapMapCeuFreq_BMI_MEN_N.txt.gz' -c 'BODYTM' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/scratch/prj/gwas_sumstats/original/PGC2_MDD_Wray/daner_ukb_170227_aligned.assoc.gz' -c 'DEPR12' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMOK10' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMOK10.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMOK11' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMOK11.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'AUTI09' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "AUTI09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'CANU03' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "CANU03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'ADHD06' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "ADHD06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'ANXI06' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "ANXI06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR14' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "DEPR14.$(date +%Y%m%d).out.txt"
#sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'NPAT12' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "NPAT12.$(date +%Y%m%d).out.txt" #this did not work because of new cluster update
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.100.assoc.fastGWA -c 'SMRV01' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV01.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.100.assoc.fastGWA -c 'SMRV02' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.75.assoc.fastGWA -c 'SMRV03' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.75.assoc.fastGWA -c 'SMRV04' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.50.assoc.fastGWA -c 'SMRV05' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.50.assoc.fastGWA -c 'SMRV06' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.25.assoc.fastGWA -c 'SMRV07' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV07.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f /users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.25.assoc.fastGWA -c 'SMRV08' -o '/scratch/prj/gwas_sumstats/cleaned' --filter.maf 0.001 --filter.info 0.6 --population EUR --process FALSE --doPipelineSpecific FALSE" --output "SMRV08.$(date +%Y%m%d).out.txt"


#cleaning of NtoNEFF sumstats for original LDSC
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'ANXI03' -r '/scratch/prj/gwas_sumstats/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.001 --filter.info 0.6 --population EUR --setntoneff TRUE --doPipelineSpecific FALSE --forLDSC FALSE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'BIPO02' -r '/scratch/prj/gwas_sumstats/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.001 --filter.info 0.6 --population EUR --setntoneff TRUE --doPipelineSpecific FALSE --forLDSC FALSE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR05' -r '/scratch/prj/gwas_sumstats/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.001 --filter.info 0.6 --population EUR --setntoneff TRUE --doPipelineSpecific FALSE --forLDSC FALSE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR09' -r '/scratch/prj/gwas_sumstats/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.001 --filter.info 0.6 --population EUR --setntoneff TRUE --doPipelineSpecific FALSE --forLDSC FALSE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR12' -r '/scratch/prj/gwas_sumstats/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.001 --filter.info 0.6 --population EUR --setntoneff TRUE --doPipelineSpecific FALSE --forLDSC FALSE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'INSO02' -r '/scratch/prj/gwas_sumstats/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.001 --filter.info 0.6 --population EUR --setntoneff TRUE --doPipelineSpecific FALSE --forLDSC FALSE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="clean" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SCHI04' -r '/scratch/prj/gwas_sumstats/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.001 --filter.info 0.6 --population EUR --setntoneff TRUE --doPipelineSpecific FALSE --forLDSC FALSE" --output "SCHI04.$(date +%Y%m%d).out.txt"

#hm3 variant list munge - should already be present in the repo munge
sbatch --time 24:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/ANXI03.gz --out ANXI03 --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 1000000;" --output=/dev/null --error=/dev/null
sbatch --time 24:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/BIPO02.gz --out BIPO02 --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 1000000;" --output=/dev/null --error=/dev/null
sbatch --time 24:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/BODY14.gz --out BODY14 --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 1000000;" --output=/dev/null --error=/dev/null
sbatch --time 24:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/BODYTM.gz --out BODYTM --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 1000000;" --output=/dev/null --error=/dev/null
sbatch --time 24:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/DEPR12.gz --out DEPR12 --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 1000000;" --output=/dev/null --error=/dev/null
sbatch --time 24:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/NEUR04.gz --out NEUR04 --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 1000000;" --output=/dev/null --error=/dev/null


#for Guiseppe
#sbatch --time 24:00:00 --partition cpu --job-name="mALCO04" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/munge_sumstats.py --sumstats /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/cleaned/ALCO04.gz --out ALCO04 --merge-alleles /users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018 --info-min 0.6 --maf-min 0.01 --ignore NEFF,CHR,BP,BETA,SE --chunksize 1000000;" --output=/dev/null --error=/dev/null
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'ALCO04' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "ALCO04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'SMOK10' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "SMOK10.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'SMOK11' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "SMOK11.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'AUTI09' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "AUTI09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'BIPO03' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "BIPO03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'CANU03' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "CANU03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'SCHI06' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "SCHI06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'ADHD06' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "ADHD06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'ANXI06' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "ANXI06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'DEPR13' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "DEPR13.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="munge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap "Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -r '/scratch/prj/gwas_sumstats/variant_lists/reference.1000G.maf.0.005.txt.gz' -c 'DEPR14' -o '/scratch/prj/gwas_sumstats/munged' --filter.maf 0.01 --filter.info 0.6 --population EUR --process TRUE --doPipelineSpecific FALSE --format ldsc" --output "DEPR14.$(date +%Y%m%d).out.txt"

#hm3 variant list munge, including NEFF
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SCHI04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV01" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV01' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV01.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV06" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV06' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV07" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV07' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV07.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV08" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SMRV08' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --format ldsc" --output "SMRV08.$(date +%Y%m%d).out.txt"

#hm3 variant list munge, including NEFF - 0 MAF, 0 INFO - supermunge standard format
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "SCHI04.$(date +%Y%m%d).out.txt"
#donot use these
#sbatch --time 01:00:00 --partition cpu --job-name="mSMRV01" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.ld.assoc.fastGWA' -c 'SMRV01' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "SMRV01.$(date +%Y%m%d).out.txt"
#sbatch --time 01:00:00 --partition cpu --job-name="mSMRV02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2ld..assoc.fastGWA' -c 'SMRV02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/w_hm3.snplist.flaskapp2018' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "SMRV02.$(date +%Y%m%d).out.txt"

#1kG variant list, 0.01 MAF
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.01 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "SCHI04.$(date +%Y%m%d).out.txt"

#1kG variant list, 0.05 MAF
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "SCHI04.$(date +%Y%m%d).out.txt"

#1kG variant list, 0.05 MAF, 0 INFO
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.05 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "SCHI04.$(date +%Y%m%d).out.txt"

#1kG variant list, 0.1 MAF
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "SCHI04.$(date +%Y%m%d).out.txt"

#new 1kG variant list, 0 MAF 0 INFO - apply filters later - for INFO weighting etc
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/1kgp3.bX.eur.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "SCHI04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV01" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.100.assoc.fastGWA' -c 'SMRV01' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV01.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.100.assoc.fastGWA' -c 'SMRV02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.75.assoc.fastGWA' -c 'SMRV03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.75.assoc.fastGWA' -c 'SMRV04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.50.assoc.fastGWA' -c 'SMRV05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV06" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.50.assoc.fastGWA' -c 'SMRV06' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV06.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV07" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim1.020.50.assoc.fastGWA' -c 'SMRV07' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV07.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSMRV08" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_C1/working_directory/sim2.020.50.assoc.fastGWA' -c 'SMRV08' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/GLAD_EDGI_NBR.mix.l2.jz2024.gz' --filter.maf 0 --filter.info 0.0 --population MIX --doPipelineSpecific FALSE" --output "SMRV08.$(date +%Y%m%d).out.txt"

#HC1kG variant list munge, 0.05 MAF
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.05 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "SCHI04.$(date +%Y%m%d).out.txt"

#HC1kG variant list munge, 0.1 MAF
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/ANXI03.gz' -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BIPO02.gz' -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODY14.gz' -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/BODYTM.gz' -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR05.gz' -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR09.gz' -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/DEPR12.gz' -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/EDUC03.gz' -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/INSO02.gz' -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/NEUR04.gz' -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -f '/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/data/gwas_sumstats/cleaned_ntoneff/SCHI04.gz' -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.1 --filter.info 0.6 --population EUR --doPipelineSpecific FALSE --forLDSC TRUE" --output "SCHI04.$(date +%Y%m%d).out.txt"

#new HC1kG variant list, 0 MAF 0 INFO - apply filters and apply NEFF later - for INFO weighting etc - NOT for original LDSC
#sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.mix.l2.jz2023.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/dbsnp151.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mANXI03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'ANXI03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBIPO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'BIPO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODY14" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'BODY14' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mBODYTM" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'BODYTM' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR05" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR05' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR09" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR09' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mDEPR12" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'DEPR12' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "DEPR12.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mEDUC03" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'EDUC03' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "EDUC03.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mINSO02" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'INSO02' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mNEUR04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'NEUR04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "NEUR04.$(date +%Y%m%d).out.txt"
sbatch --time 01:00:00 --partition cpu --job-name="mSCHI04" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript ~/project/genetic_correlations/scripts/cleaning/runStandardCleanAndMunge.R -c 'SCHI04' -r '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.eur.l2.jz2024.gz' -s '/users/k19049801/project/JZ_GED_PHD_C1/data/variant_lists/hc1kgp3.b38.jz2024.synonyms.gz' --filter.maf 0.0 --filter.info 0.0 --population EUR --doPipelineSpecific FALSE" --output "SCHI04.$(date +%Y%m%d).out.txt"


#imputation
sbatch --time 2-00:00:00 --partition cpu --job-name="iANXI03" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a ANXI03 -l cluster" --output "setup8.ldimp.ANXI03.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iBIPO02" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a BIPO02 -l cluster" --output "setup8.ldimp.BIPO02.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iBODY14" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a BODY14 -l cluster" --output "setup8.ldimp.BODY14.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iBODYTM" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a BODYTM -l cluster" --output "setup8.ldimp.BODYTM.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iDEPR05" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a DEPR05 -l cluster" --output "setup8.ldimp.DEPR05.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iDEPR5B" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a DEPR5B -l cluster" --output "setup8.ldimp.DEPR5B.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iDEPR09" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a DEPR09 -l cluster" --output "setup8.ldimp.DEPR09.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iEDUC03" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a EDUC03 -l cluster" --output "setup8.ldimp.EDUC03.$(date +%Y%m%d).out.txt"
#sbatch --time 2-00:00:00 --partition cpu --job-name="iGWISM" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a GWISM -l cluster" --output "setup8.ldimp.GWISM.$(date +%Y%m%d).out.txt"
#sbatch --time 2-00:00:00 --partition cpu --job-name="iGWISI" --ntasks 1 --cpus-per-task 4 --mem 80G --wrap="Rscript setup8.R -t impute -a GWISI -l cluster" --output "setup8.ldimp.GWISI.$(date +%Y%m%d).out.txt" #no, use the one also placed in the munged folder
sbatch --time 2-00:00:00 --partition cpu --job-name="iHEIG04M" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a HEIG04M -l cluster" --output "setup8.ldimp.HEIG04M.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iHEIG04F" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a HEIG04F -l cluster" --output "setup8.ldimp.HEIG04F.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iINSO02" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a INSO02 -l cluster" --output "setup8.ldimp.INSO02.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iNEUR02" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a NEUR02 -l cluster" --output "setup8.ldimp.NEUR02.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iSCHI04" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a SCHI04 -l cluster" --output "setup8.ldimp.SCHI04.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iWEIG04M" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a WEIG04M -l cluster" --output "setup8.ldimp.WEIG04M.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="iWEIG04F" --ntasks 1 --cpus-per-task 6 --mem 80G --wrap="Rscript setup8.R -t impute -a WEIG04F -l cluster" --output "setup8.ldimp.WEIG04F.$(date +%Y%m%d).out.txt"

#per dataset stats
sbatch --time 8:00:00 --partition cpu --job-name="dstat" --ntasks 1 --cpus-per-task 6 --mem 32G --wrap="Rscript setup8.R -t stat -l cluster" --output "setup8.stat.$(date +%Y%m%d).out.txt"

#GWIS
sbatch --time 2-00:00:00 --partition cpu --job-name="GWIS" --ntasks 1 --cpus-per-task 5 --mem 32G --wrap="Rscript setup8.R -t gwis -l cluster" --output "setup8.gwis.$(date +%Y%m%d).out.txt"
sbatch --time 2-00:00:00 --partition cpu --job-name="mvLD.gwis" --ntasks 1 --cpus-per-task 5 --mem 32G --wrap="Rscript setup8.R -t mvLD.gwis -l cluster" --output "setup8.mvLD.gwis.$(date +%Y%m%d).out.txt"

#ssimp imputation
sbatch --time 2-00:00:00 --partition cpu --job-name="ssimpformat" --ntasks 1 --cpus-per-task 5 --mem 8G --wrap="Rscript setup8.R -t ssimpformat -l cluster" --output "setup8.ssimpformat.$(date +%Y%m%d).out.txt"

#visualise gwas distributions
sbatch --time 2:00:00 --partition cpu --job-name="plotdistributions" --ntasks 1 --cpus-per-task 5 --mem 20G --wrap="Rscript setup8.R -t plotdistributions -l cluster" --output "setup8.plotdistributions.$(date +%Y%m%d).out.txt"

/scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --download.build.db
#test
sbatch --time 1:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 3 --mem 8G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge/ADHD05.gz --ref /scratch/prj/gwas_sumstats/ssimp_software-master/ref/small.vcf.sample.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/ADHD05.txt; date;" --output "SSIMP_evaluation.ADHD05.$(date +%Y%m%d).out.txt"


for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/SCHI04.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/SCHI04.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.SCHI04.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/BODY11.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/BODY11.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.BODY11.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/NEUR02.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/NEUR02.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.NEUR02.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/DEPR05.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/DEPR05.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.DEPR05.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/DEPR5B.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/DEPR5B.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.DEPR5B.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/DEPR09.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/DEPR09.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.DEPR09.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/BIPO02.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/BIPO02.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.BIPO02.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/ANXI03.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/ANXI03.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.ANXI03.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/INSO02.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/INSO02.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.INSO02.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/EDUC03.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/EDUC03.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.EDUC03.$chr.$(date +%Y%m%d).out.txt"; done

for chr in 1 2 3 4 5 6 7 8 9 10 11 "12-14" "15-17" "18-23"; do sbatch --time 2-00:00:00 --partition cpu --job-name="ssimp" --ntasks 1 --cpus-per-task 4 --mem 20G --wrap="export LC_ALL=C; date; /scratch/prj/gwas_sumstats/ssimp_software-master/ssimp --gwas ../data/gwas_sumstats/munged_1kg_eur_supermunge.4ssimp/BVOL01.gz --ref ../data/reference_panel/hc1kgp3.b38.plink/hc1kgp3.b38.eur.jz2022.vcf.gz --out ../data/gwas_sumstats/imputed_1kg_eur_ssimp/BVOL01.$chr --impute.maf 0.001 --tag.maf 0.01 --impute.range $chr; date;" --output "setup8.ssimp.BVOL01.$chr.$(date +%Y%m%d).out.txt"; done

sbatch --time 2-00:00:00 --partition cpu --job-name="merge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap="Rscript setup8.R -t merge -l cluster" --output "setup8.merge.$(date +%Y%m%d).out.txt"

# #extra munge to produce exported files for original ldsc - Imputed BODY11 needs lots of memory - DEPRECATED
# sbatch --time 24:00:00 --partition cpu --job-name="smunge" --ntasks 1 --cpus-per-task 5 --mem 64G --wrap="Rscript setup8.R -t munge2 -l cluster" --output "setup8.munge2.$(date +%Y%m%d).out.txt"

sbatch --time 2-00:00:00 --partition cpu --job-name="LDSC" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; Rscript setup8.R -t LD -l cluster;" --output "setup8.LD.$(date +%Y%m%d).out.txt"

sbatch --time 2-00:00:00 --partition cpu --job-name="mvLD" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript setup8.R -t mvLD -l cluster" --output "setup8.mvLD.$(date +%Y%m%d).out.txt"
#for running original ldsc - using the . sh equivalent to bash source to activate the virtual python envoronment
sbatch --time 2-00:00:00 --partition cpu --job-name="mvLD" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap=". /users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate; Rscript setup8.R -t mvLD -l cluster;" --output "setup8.mvLD.$(date +%Y%m%d).out.txt"
Rscript setup8.R -t mvLD -l cluster > "setup8.mvLD.$(date +%Y%m%d).out.txt" #if needed to run on interactive node beacuase of python venv activation issues
#run ldsc manually - test
#/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc/ldsc.py --n-blocks 250 --ref-ld /users/k19049801/project/JZ_GED_PHD_C1/data/ld_scores/hc1kgp3.b38.eur.l2.jz2023/hc1kgp3.b38.eur.jz2023.1cm.250blocks --rg /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/munged_1kg_eur_supermunge/BODY14.gz,/users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/munged_1kg_eur_supermunge/NEUR02.gz --samp-prev nan,nan --pop-prev nan,nan --no-check-alleles --out test.ldsc

#original ldsc validations - DO NOT FORGET TO CHANGE THE FOLDERNAME FOR DIFFERENT LDSC VERSIONS
#these worked - need LDSC munged sumstats!
source ~/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc-venv2.7/bin/activate
/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py --h2 /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/munged/SCHI05.sumstats.gz --ref-ld-chr /users/k19049801/project/JZ_GED_PHD_C1/data/ld_scores/eur_w_ld_chr/ --w-ld-chr /users/k19049801/project/JZ_GED_PHD_C1/data/ld_scores/eur_w_ld_chr/ --samp-prev 0.2395569 --pop-prev 0.0072 --no-check-alleles
/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py --rg /users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/munged/SCHI05.sumstats.gz,/users/k19049801/project/JZ_GED_PHD_C1/data/gwas_sumstats/munged/BODY11.sumstats.gz --ref-ld-chr /users/k19049801/project/JZ_GED_PHD_C1/data/ld_scores/eur_w_ld_chr/ --w-ld-chr /users/k19049801/project/JZ_GED_PHD_C1/data/ld_scores/eur_w_ld_chr/ --samp-prev 0.2395569,nan --pop-prev 0.0072,nan --no-check-alleles
#/users/k19049801/project/JZ_GED_PHD_ADMIN_GENERAL/software/ldsc.orig/ldsc.py --n-blocks 200 --rg /scratch/prj/gwas_sumstats/munged/ANXI03.sumstats.gz,/scratch/prj/gwas_sumstats/munged/ANXI03.sumstats.gz --ref-ld-chr /scratch/users/k19049801/project/JZ_GED_PHD_C1/data/ld_scores/eur_w_ld_chr/ --w-ld-chr /scratch/users/k19049801/project/JZ_GED_PHD_C1/data/ld_scores/eur_w_ld_chr/ --out ANXI03_h2_hm3.ldsc --no-check-alleles --samp-prev 0.5,0.5 --pop-prev 0.16,0.16


#fix duplicates in PLINK refpanel for Thibault
plink2 --bfile g1000_eur --rm-dup --make-pgen --out g1000_eur.iddup


sbatch --time 01:00:00 --partition cpu --job-name="Rtest" --ntasks 1 --cpus-per-task 6 --mem 64G --wrap="Rscript code.r" --output "rtest.$(date +%Y%m%d).out.txt"

